// @salla.sa/twilight-components/dist/components/salla-notification-item2.js
/*!
 * Crafted with ❤ by Salla
 */
import { proxyCustomElement, HTMLElement, h } from '@stencil/core/internal/client';

const sallaNotificationsCss = "";

const SallaNotificationItem = /*@__PURE__*/ proxyCustomElement(class SallaNotificationItem extends HTMLElement {
    constructor() {
        super();
        this.__registerHost();
    }
    render() {
        return h("a", { key: '63f74fac3cc9933dd0f4808ba55852dfd7bddb62', href: this.notification.url, class: { "s-notifications-item": true, "s-notifications-item-shadow": true, "s-notifications-item-read": !this.notification.is_new } }, h("span", { key: '29c5ed5f2c13e17ad27e534806e59f8c316f4543', class: "s-notifications-item-icon" }, h("i", { key: '2d3da9b6b8f06d0f9af0841bcf9fe3a6ad4ebca1', class: { 's-notifications-item-icon sicon-bell-ring': true, 's-notifications-item-icon-new': this.notification.is_new } })), h("div", { key: 'ccb51c53e1584e72eb69120b7e68322f1e1abdb3', class: "s-notifications-item-content" }, h("div", { key: '2ea640668fbd6616f7968e5585a3ae5b0d6fb17e', class: "s-notifications-item-content-leading" }, h("h4", { key: '00d91ee1b4d9aa6b3a51cb7a6dcada8c601017c5' }, this.notification.title), h("p", { key: '8ff6803d1df5128cf9d815e5645d739fcf11d975' }, this.notification.sub_title)), h("span", { key: '3f4e6b5b0bdc3e669307dc1ec6bba315e86e3fea', class: "s-notifications-item-content-trailing" }, h("i", { key: 'e2777150fad76083f8673457e4f2b90b23885974', class: "sicon-time s-notifications-item-content-trailing-icon" }), h("span", { key: '80fb3b5c1d28c4cfbd84d2354c10776ccc209883' }, this.notification.date))));
    }
    static get style() { return sallaNotificationsCss; }
}, [0, "salla-notification-item", {
        "notification": [16]
    }]);
function defineCustomElement() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["salla-notification-item"];
    components.forEach(tagName => { switch (tagName) {
        case "salla-notification-item":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, SallaNotificationItem);
            }
            break;
    } });
}
defineCustomElement();

export { SallaNotificationItem as S, defineCustomElement as d };
