// @salla.sa/twilight-components/dist/cjs/salla-loyalty-prize-item.cjs.entry.js
/*!
 * Crafted with ❤ by Salla
 */
'use strict';

var index = require('./index-Bxxuur88.js');

const sallaLoyaltyPrizeItemCss = "salla-loyalty-prize-item:hover{box-shadow:0 5px 15px 1px rgba(0, 0, 0, 0.04) !important}";

const SallaLoyaltyPrizeItem = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        this.prizeItemSelected = index.createEvent(this, "prizeItemSelected");
    }
    onPrizeItemClick() {
        this.prizeItemSelected.emit(this.item);
    }
    render() {
        return (index.h(index.Host, { key: '41df19345e2a4ed6a38eb7ee1fa7a409d19fa89d' }, index.h("div", { key: '2e8fadde2fe4aefbf640fd7025b4ff807739f547', onClick: this.onPrizeItemClick.bind(this) }, index.h("img", { key: '30342789e4dc4c65756205a928cb1011a6c3e9b3', class: "s-loyalty-prize-item-image", src: this.item.image, alt: this.item.name }), index.h("div", { key: 'afc2880c949c32b7acd03fafb7e67c913543e476', class: "s-loyalty-prize-item-title" }, this.item.name), index.h("div", { key: 'f59c157c72b18d5bc678882c0ba2619e15cea629', class: "s-loyalty-prize-item-subtitle" }, this.item.description), index.h("div", { key: 'a68e7ae92728ce8a17025e8ed462686c70653748', class: "s-loyalty-prize-item-points" }, this.item.cost_points, " ", salla.lang.get('pages.loyalty_program.point')))));
    }
};
SallaLoyaltyPrizeItem.style = sallaLoyaltyPrizeItemCss;

exports.salla_loyalty_prize_item = SallaLoyaltyPrizeItem;
