// @salla.sa/twilight-components/dist/cjs/salla-bullet-delivery.cjs.entry.js
/*!
 * Crafted with ❤ by Salla
 */
'use strict';

var index = require('./index-Bxxuur88.js');
var search = require('./search-c7Aa7lM9.js');
var location = require('./location-DIyESLuI.js');
var keyboard_arrow_right = require('./keyboard_arrow_right-BayM_Il2.js');

/**
 * API Service for salla-bullet-delivery component
 * Uses Salla's shipping API and scope API for delivery location management
 */
const API_LOG_PREFIX = 'BulletDeliveryAPI';
const validId = (id) => {
    const n = id == null ? Number.NaN : Number(id);
    return !Number.isNaN(n) && n !== 0;
};
async function withApiErrorHandling(fn, fallback, logLabel) {
    try {
        return await fn();
    }
    catch (error) {
        console.error(`${API_LOG_PREFIX}: ${logLabel}`, error);
        return fallback;
    }
}
/** Uses country code from API (e.g. 'SA'). */
const isSaudiArabia = (countryCode) => String(countryCode).toUpperCase() === 'SA';
/**
 * Fast Delivery API Service
 * Uses Salla's shipping API and scopes API. Branch and address data are used as returned by the API (no mapping).
 */
const bulletDeliveryAPI = {
    /**
     * Get available countries from Salla shipping API
     * @param forBranch - when true, use for_branch=1 (e.g. for pickup/branches tab)
     */
    async getCountries(forBranch = false) {
        return withApiErrorHandling(async () => {
            const data = (await salla.api.request("shipping/countries", { params: { for_branch: forBranch ? 1 : 0 } }))?.data ?? [];
            return data.map((c) => ({ id: c.id, code: c.code, name: c.name, has_regions: isSaudiArabia(c.code) }));
        }, [], 'Error getting countries');
    },
    /**
     * Get regions for a country from Salla shipping API
     * Endpoint: GET /shipping/countries/<COUNTRY_ID>/region
     */
    async getRegions(countryId) {
        if (!validId(countryId)) {
            console.warn(`${API_LOG_PREFIX}: getRegions called without valid country_id`);
            return [];
        }
        return withApiErrorHandling(async () => {
            const data = (await salla.api.request(`shipping/countries/${countryId}/regions`))?.data ?? [];
            return data.map((r) => ({ id: r.id, name: r.name, code: r.code, country_id: Number(countryId) }));
        }, [], 'Error fetching regions');
    },
    /**
     * Get cities from Salla shipping API
     * @param regionId - Optional; when provided (e.g. for SA), cities are filtered by region
     */
    async getCities(countryId, regionId) {
        if (!validId(countryId)) {
            console.warn(`${API_LOG_PREFIX}: getCities called without valid country_id`);
            return [];
        }
        return withApiErrorHandling(async () => {
            const params = { for_branch: 0, country_id: countryId };
            if (regionId)
                params.region_id = regionId;
            const data = (await salla.api.request("shipping/cities", { params }))?.data ?? [];
            return data.map((c) => ({ id: c.id, name: c.name, country_id: countryId, ...(c.region_id != null && { region_id: c.region_id }) }));
        }, [], 'Error fetching cities');
    },
    /**
     * Get districts from Salla shipping API
     */
    async getDistricts(cityId) {
        if (!validId(cityId)) {
            console.warn(`${API_LOG_PREFIX}: getDistricts called without valid city_id`);
            return [];
        }
        return withApiErrorHandling(async () => {
            const raw = (await salla.api.request("shipping/districts", { params: { for_branch: 0, city_id: cityId } }))?.data;
            const list = Array.isArray(raw) ? raw : (Array.isArray(raw?.districts) ? raw.districts : []);
            return list.map((d) => ({ id: d.id, name: d.name, name_en: d.name_en, city_id: cityId }));
        }, [], 'Error fetching districts');
    },
    async getSavedAddresses() {
        return withApiErrorHandling(async () => {
            const data = (await salla.api.request("address"))?.data;
            return Array.isArray(data) ? data : [];
        }, [], 'Error fetching user addresses');
    },
    async getBranches({ query, lat, lng, country_id, per_page = 20 } = {}) {
        return withApiErrorHandling(async () => {
            const params = { per_page };
            if (query?.trim().length >= 2)
                params.query = query;
            if (lat)
                params.lat = lat;
            if (lng)
                params.lng = lng;
            if (country_id)
                params.country_id = country_id;
            const res = await salla.api.request("branches", { params });
            return (Array.isArray(res?.data) ? res.data : []);
        }, [], 'Error fetching branches');
    },
    async saveAddressLocation(payload) {
        return withApiErrorHandling(async () => {
            const res = await salla.api.request('address/location', { ...payload, for_allocation: true }, 'post');
            const data = res?.data ?? res;
            const address = typeof data === 'object' && data != null && 'id' in data ? data : undefined;
            return { success: true, address };
        }, { success: false }, 'Error saving address/location');
    },
    async setDeliveryScope(scopeId) {
        return withApiErrorHandling(async () => {
            await salla.scope.change({ id: scopeId });
            salla.storage.set("scope", { ...(salla.storage.get("scope") || {}), id: scopeId });
            return true;
        }, false, 'Error setting delivery scope');
    },
    async allocateScope(payload) {
        const errMsg = (d) => d?.error?.message ?? d?.message ?? 'Failed to allocate scope';
        try {
            const response = await salla.api.request("scopes/allocation", payload, 'post');
            // SDK resolves with { data } and throws on non-2xx; any resolved value is success
            const data = response?.data ?? response;
            return { success: true, data: data };
        }
        catch (error) {
            console.error(`${API_LOG_PREFIX}: Error allocating scope`, error);
            const err = error;
            return { success: false, error: err?.response?.data != null ? errMsg(err.response.data) : (err?.message ?? 'Unknown error') };
        }
    },
};

/**
 * Helper functions for salla-bullet-delivery component
 */
const EARTH_RADIUS_KM = 6371;
const deg2rad = (deg) => deg * (Math.PI / 180);
/** Haversine distance in km */
function calculateDistance(lat1, lon1, lat2, lon2) {
    const dLat = deg2rad(lat2 - lat1);
    const dLon = deg2rad(lon2 - lon1);
    const a = Math.sin(dLat / 2) ** 2 +
        Math.cos(deg2rad(lat1)) * Math.cos(deg2rad(lat2)) * Math.sin(dLon / 2) ** 2;
    return EARTH_RADIUS_KM * 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
}
/** First working-hours slot from API branch.working_hours[].times[]. */
const getBranchFirstSlot = (workingHours) => workingHours?.[0]?.times?.[0] ?? null;
/** Nearest available branch by distance; single pass O(n). */
function findNearestBranch(branches, userLat, userLon) {
    let nearest = null;
    let minD = Number.POSITIVE_INFINITY;
    for (const b of branches) {
        const lat = b.location?.lat != null ? Number(b.location.lat) : Number.NaN;
        const lng = b.location?.lng != null ? Number(b.location.lng) : Number.NaN;
        if (Number.isNaN(lat) || Number.isNaN(lng) || b.is_open === false)
            continue;
        const d = calculateDistance(userLat, userLon, lat, lng);
        if (d < minD) {
            minD = d;
            nearest = b;
        }
    }
    return nearest;
}
/** 24h time string → locale 12h (e.g. "23:55" → "11:55 PM"). */
function formatTimeTo12h(time24, locale) {
    const [h, m] = String(time24).trim().split(':').map(Number);
    return Number.isNaN(h) || Number.isNaN(m)
        ? time24
        : new Date(2000, 0, 1, h, m).toLocaleTimeString(locale, { hour: 'numeric', minute: '2-digit', hour12: true });
}
/** Working hours display: "until X" when from is 00:00, else "from - to" in 12h. */
function formatWorkingHoursDisplay(from, to, locale, untilLabel) {
    const f = String(from).trim();
    const t = String(to).trim();
    return !f || !t ? '' : f === '00:00' ? `${untilLabel} ${formatTimeTo12h(t, locale)}` : `${formatTimeTo12h(f, locale)} - ${formatTimeTo12h(t, locale)}`;
}
const filterBranches = (branches, searchQuery) => {
    if (!searchQuery)
        return branches;
    const q = searchQuery.toLowerCase();
    return branches.filter((b) => b.name.toLowerCase().includes(q) || b.city?.name?.toLowerCase().includes(q));
};
/** Match saved address to guest intent by city/district (and optional street). */
function findMatchingAddress(guestIntent, userAddresses) {
    const cityId = getIntentCityId(guestIntent);
    const districtId = getIntentDistrictId(guestIntent);
    return userAddresses.find((addr) => {
        const addrCity = addr.city?.id ?? addr.city_id;
        const addrDistrict = addr.district?.id ?? addr.district_id;
        if (cityId == null || addrCity !== cityId)
            return false;
        if (districtId != null) {
            if (addrDistrict !== districtId)
                return false;
            if (guestIntent.street && addr.street)
                return guestIntent.street.toLowerCase() === addr.street.toLowerCase();
            return true;
        }
        return true;
    }) ?? null;
}
const getIntentSubtitle = (intent, toAddress, fromBranch) => {
    if (intent.type === 'branch') {
        const loc = intent.branch_details?.city ?? intent.branch_details?.name;
        return loc ? `${fromBranch} ${loc}`.trim() : '';
    }
    if (intent.type === 'address' && intent.address_details) {
        const addr = [intent.address_details.district?.name, intent.address_details.city?.name, intent.address_details.country?.name].filter(Boolean).join('، ') || intent.address_details.short_address || '';
        return addr ? `${toAddress} ${addr}`.trim() : '';
    }
    return '';
};
const idFrom = (val) => (val != null ? Number(val) : undefined);
const getIntentCountryId = (i) => idFrom(i?.address_details?.country?.id);
const getIntentCountryCode = (i) => i?.country_code ?? i?.address_details?.country?.code ?? '';
const getIntentRegionId = (i) => idFrom(i?.address_details?.region?.id);
const getIntentCityId = (i) => idFrom(i?.address_details?.city?.id);
const getIntentDistrictId = (i) => idFrom(i?.address_details?.district?.id);
const getIntentBranchId = (i) => i?.branch_details?.id;
/** Intent coordinates (branch tab only; address flow no longer uses coordinates). */
const getIntentLatitude = (i) => i?.branch_details?.latitude;
const getIntentLongitude = (i) => i?.branch_details?.longitude;
const hasSessionAddressIntent = (intent) => {
    if (!intent || intent.type !== 'address' || intent.address_id)
        return false;
    const c = getIntentCountryId(intent);
    const city = getIntentCityId(intent);
    const d = getIntentDistrictId(intent);
    const code = getIntentCountryCode(intent);
    return !!(c && city && (!isSaudiArabia(code) || d));
};
const requireRegionAndDistrictForSA = (countryCode, regionId, districtIdOrName) => !countryCode || !isSaudiArabia(countryCode)
    ? true
    : (regionId != null && Number(regionId) !== 0) && (districtIdOrName != null && (typeof districtIdOrName === 'number' ? districtIdOrName !== 0 : String(districtIdOrName).trim().length > 0));
function buildAddressLocationPayloadFromSelection(options) {
    const { countryId, countryCode, regionId, cityId, districtId, description = '' } = options;
    const desc = description || '';
    return {
        country_id: Number(countryId) || undefined,
        region_id: regionId != null ? Number(regionId) : undefined,
        city_id: cityId != null ? Number(cityId) : undefined,
        district_id: districtId != null ? Number(districtId) : undefined,
        description: desc,
        ...(countryCode && isSaudiArabia(countryCode) && { local: desc }),
    };
}
const GEO_ERROR_MESSAGES = {
    [GeolocationPositionError.PERMISSION_DENIED]: 'Location permission denied',
    [GeolocationPositionError.POSITION_UNAVAILABLE]: 'Location unavailable',
    [GeolocationPositionError.TIMEOUT]: 'Location timeout',
};
const getGeolocationErrorMessage = (errorCode) => GEO_ERROR_MESSAGES[errorCode] ?? 'An error occurred while detecting location';

var Store = `<svg width="15" height="14" viewBox="0 0 15 14" fill="none" xmlns="http://www.w3.org/2000/svg">
    <path
        d="M1.14453 5.83032L1.14453 9.16486C1.14453 11.0511 1.14453 11.9943 1.73032 12.5803C2.3161 13.1663 3.25891 13.1663 5.14453 13.1663L9.14453 13.1663C11.0301 13.1663 11.9729 13.1663 12.5587 12.5803C13.1445 11.9943 13.1445 11.0511 13.1445 9.16486V5.83032"
        stroke="currentcolor" stroke-linecap="round" />
    <path
        d="M9.14453 10.1619C8.68846 10.5667 7.9624 10.8285 7.14453 10.8285C6.32666 10.8285 5.60059 10.5667 5.14453 10.1619"
        stroke="currentcolor" stroke-linecap="round" />
    <path
        d="M5.90256 4.44533C5.71455 5.12426 5.03086 6.29542 3.73185 6.46516C2.58488 6.61502 1.71497 6.11437 1.49277 5.90504C1.24779 5.7353 0.689441 5.19216 0.552703 4.85269C0.415966 4.51322 0.575492 3.77771 0.689441 3.47785L1.14495 2.15892C1.25615 1.82765 1.51646 1.04411 1.78334 0.779094C2.05022 0.514073 2.59054 0.502543 2.81294 0.502543L7.48329 0.502543C8.68543 0.519525 11.3139 0.491824 11.8335 0.502545C12.3532 0.513266 12.6654 0.948828 12.7565 1.13553C13.5318 3.01341 13.8333 4.08889 13.8333 4.54717C13.7321 5.03605 13.3133 5.95787 11.8335 6.36332C10.2956 6.7847 9.42363 5.96509 9.15007 5.65043M5.27015 5.65044C5.48665 5.91635 6.16578 6.45158 7.15029 6.46516C8.1348 6.47873 8.98485 5.79188 9.28681 5.44676C9.37227 5.34492 9.55687 5.04279 9.7483 4.44533"
        stroke="currentcolor" stroke-linecap="round" stroke-linejoin="round" />
</svg>
    `;

var GPS = `<svg width="16" height="16" viewBox="0 0 16 16" fill="none" xmlns="http://www.w3.org/2000/svg">
    <path
        d="M13.3258 7.65006C13.3258 10.7797 10.7888 13.3167 7.65916 13.3167C4.52955 13.3167 1.99249 10.7797 1.99249 7.65006C1.99249 4.52045 4.52955 1.9834 7.65916 1.9834C10.7888 1.9834 13.3258 4.52045 13.3258 7.65006Z"
        stroke="#555555" stroke-width="1.3" />
    <path d="M14.65 7.65039H13.3167" stroke="#555555" stroke-width="1.3" stroke-linecap="round"
        stroke-linejoin="round" />
    <path d="M1.98336 7.65039L0.650024 7.65039" stroke="#555555" stroke-width="1.3"
        stroke-linecap="round" stroke-linejoin="round" />
    <path d="M7.65002 0.650065L7.65002 1.9834" stroke="#555555" stroke-width="1.3"
        stroke-linecap="round" stroke-linejoin="round" />
    <path d="M7.65002 13.3171L7.65002 14.6504" stroke="#555555" stroke-width="1.3"
        stroke-linecap="round" stroke-linejoin="round" />
    <path
        d="M9.64999 7.65039C9.64999 8.75496 8.75456 9.65039 7.64999 9.65039C6.54542 9.65039 5.64999 8.75496 5.64999 7.65039C5.64999 6.54582 6.54542 5.65039 7.64999 5.65039C8.75456 5.65039 9.64999 6.54582 9.64999 7.65039Z"
        stroke="#555555" stroke-width="1.3" />
</svg>
    `;

var MiniMap = `<svg width="35" height="35" viewBox="0 0 35 35" fill="none" xmlns="http://www.w3.org/2000/svg">
    <g filter="url(#filter0_d_24181_220540)">
    <mask id="mask0_24181_220540" style="mask-type:alpha" maskUnits="userSpaceOnUse" x="1" y="0" width="33" height="32">
    <rect x="1.90958" y="0.576577" width="30.8468" height="30.8468" rx="9.22523" fill="#D9D9D9" stroke="white" stroke-width="1.15315"/>
    </mask>
    <g mask="url(#mask0_24181_220540)">
    <rect x="1.90958" y="0.576577" width="30.8468" height="30.8468" rx="9.22523" fill="#E8E8E8" stroke="white" stroke-width="1.15315"/>
    <path d="M17.9096 6.77445L28.7204 5.33301H32.4681L32.7564 11.8195V23.6393L10.4141 23.9276V10.8105L17.9096 6.77445Z" fill="#97E7C8"/>
    <path d="M-2.70303 10.6664H10.2699M41.6934 5.47722C38.378 5.47722 34.7744 5.47722 27.8555 5.47722C20.9366 5.47722 18.4862 6.63037 15.4591 8.64839C12.4321 10.6664 10.2699 10.6664 10.2699 10.6664M10.2699 34.4502V23.6394M10.2699 10.6664V23.6394M-6.16249 23.6394H10.2699M10.2699 23.6394H21.0807M38.8105 23.6394H21.0807M21.0807 23.6394V34.4502" stroke="white" stroke-width="2.30631"/>
    <rect x="2.34202" y="1.00901" width="29.982" height="29.982" rx="8.79279" stroke="white" stroke-width="2.01802"/>
    </g>
    <g clip-path="url(#clip0_24181_220540)">
    <path fill-rule="evenodd" clip-rule="evenodd" d="M18.069 19.8621C19.4995 18.6902 22.171 16.0948 22.171 13.0548C22.171 11.7333 21.6461 10.466 20.7117 9.5316C19.7773 8.5972 18.51 8.07227 17.1885 8.07227C15.8671 8.07227 14.5998 8.5972 13.6654 9.5316C12.731 10.466 12.2061 11.7333 12.2061 13.0548C12.2061 16.0948 14.8776 18.6902 16.3081 19.8621C16.5553 20.0679 16.8669 20.1806 17.1885 20.1806C17.5102 20.1806 17.8218 20.0679 18.069 19.8621ZM17.1885 14.1618C17.517 14.1618 17.838 14.0644 18.1111 13.8819C18.3842 13.6995 18.597 13.4401 18.7227 13.1367C18.8484 12.8333 18.8812 12.4994 18.8172 12.1773C18.7531 11.8552 18.595 11.5593 18.3627 11.3271C18.1305 11.0948 17.8346 10.9367 17.5125 10.8726C17.1904 10.8085 16.8565 10.8414 16.5531 10.9671C16.2497 11.0928 15.9903 11.3056 15.8079 11.5787C15.6254 11.8518 15.528 12.1728 15.528 12.5012C15.528 12.9416 15.703 13.364 16.0144 13.6754C16.3258 13.9868 16.7481 14.1618 17.1885 14.1618Z" fill="#55575D"/>
    </g>
    </g>
    <defs>
    <filter id="filter0_d_24181_220540" x="-0.000325561" y="0" width="34.6667" height="34.6667" filterUnits="userSpaceOnUse" color-interpolation-filters="sRGB">
    <feFlood flood-opacity="0" result="BackgroundImageFix"/>
    <feColorMatrix in="SourceAlpha" type="matrix" values="0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 127 0" result="hardAlpha"/>
    <feOffset dy="1.33333"/>
    <feGaussianBlur stdDeviation="0.666667"/>
    <feColorMatrix type="matrix" values="0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0.04 0"/>
    <feBlend mode="normal" in2="BackgroundImageFix" result="effect1_dropShadow_24181_220540"/>
    <feBlend mode="normal" in="SourceGraphic" in2="effect1_dropShadow_24181_220540" result="shape"/>
    </filter>
    <clipPath id="clip0_24181_220540">
    <rect width="12.1081" height="12.1081" fill="white" transform="translate(11.1338 8.07227)"/>
    </clipPath>
    </defs>
    </svg>
    `;

var GetDirections = `<svg width="16" height="16" viewBox="0 0 16 16" fill="none" xmlns="http://www.w3.org/2000/svg">
    <path
        d="M7.94798 3.19336C11.1309 2.1083 12.7223 1.56578 13.5783 2.42174C14.4342 3.27771 13.8917 4.86914 12.8066 8.05201L12.0678 10.2195C11.2344 12.6639 10.8178 13.8861 10.131 13.9872C9.94631 14.0144 9.75517 13.9981 9.57248 13.9394C8.89297 13.7212 8.53377 12.4326 7.81538 9.8553C7.65604 9.28364 7.57637 8.99781 7.39494 8.7795C7.34228 8.71614 7.28386 8.65772 7.2205 8.60506C7.00219 8.42363 6.71636 8.34396 6.1447 8.18462C3.56741 7.46623 2.27876 7.10703 2.0606 6.42752C2.00195 6.24483 1.98558 6.05369 2.01277 5.86903C2.11392 5.18221 3.33613 4.76556 5.78054 3.93225L7.94798 3.19336Z"
        stroke="#555555" stroke-width="1.2" />
</svg>
    `;

const sallaBulletDeliveryCss = "@supports (interpolate-size: allow-keywords) and (block-size: calc-size(auto, size)) {\n  .s-bullet-delivery .s-bullet-delivery-modal .s-modal-body, .s-bullet-delivery-inner {\n    @apply [interpolate-size:allow-keywords];\n  }\n  .s-bullet-delivery .s-bullet-delivery-modal .s-modal-body {\n    @apply [overflow:clip] [transition:block-size_500ms_ease];\n  }\n  .s-bullet-delivery-inner {\n    @apply [overflow:clip] [transition:block-size_500ms_ease];\n    block-size: calc-size(auto, size);\n  }\n}\n\n@keyframes s-bullet-delivery-fade-in {\n  from {\n    opacity: 0;\n    transform: translateY(6px);\n  }\n  to {\n    opacity: 1;\n    transform: translateY(0);\n  }\n}";

const BRANCH_SEARCH_DEBOUNCE_MS = 1000;
const GEOLOCATION_TIMEOUT = 10000;
const SallaBulletDelivery = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        this.bulletDeliveryConfirmed = index.createEvent(this, "bulletDeliveryConfirmed");
        this.bulletDeliveryClosed = index.createEvent(this, "bulletDeliveryClosed");
        this.addressCreated = index.createEvent(this, "addressCreated");
        this.headerContextUpdate = index.createEvent(this, "headerContextUpdate");
        this.confirming = false;
        this.branchSearchDebounceTimer = null;
        this.componentReady = false;
        this.pendingOpen = false;
        this.tabChanging = false;
        this.intentStorageKey = "bullet_delivery_intent";
        this.sessionShownKey = "bullet_delivery_shown_this_session";
        this.cartItemAddedEvent = "cart::item.added";
        this.cartItemAddedHandler = null;
        this.useCartEventApi = false;
        this.bulletDeliveryOpenHandler = (eventData) => {
            this.preselectedAddressId = eventData?.preselected_address_id;
            this.preselectedBranchId = eventData?.preselected_branch_id;
            this.open();
        };
        this.closeHandler = null;
        this.guestLoginSyncHandler = null;
        /** True after saved addresses have been loaded (lazy: only when address tab is shown). */
        this.savedAddressesLoaded = false;
        // Core state
        this.loading = true;
        this.activeTab = "address";
        this.isLoggedIn = false;
        this.viewMode = "main";
        // Location data
        this.countries = [];
        this.regions = [];
        this.cities = [];
        this.districts = [];
        this.loadingRegions = false;
        this.selectedCountry = null;
        this.selectedRegion = null;
        this.selectedCity = null;
        this.selectedDistrict = null;
        this.districtName = "";
        // Saved addresses state
        this.savedAddresses = [];
        this.selectedSavedAddress = null;
        /** True when user selected the session address (guest address from before login, not yet saved) */
        this.selectedSessionAddress = false;
        // Pickup state
        this.branches = [];
        this.filteredBranches = [];
        this.selectedBranch = null;
        this.branchSearchQuery = "";
        // Loading states
        this.loadingCities = false;
        this.loadingDistricts = false;
        this.loadingBranches = false;
        this.loadingNearestBranch = false;
        this.locationError = "";
        this.savingAddress = false;
        this.loadingSavedAddresses = false;
        this.showCartWillBeClearedBanner = false;
        /** Shown when scopes/allocation returns 422 (address outside delivery coverage). Only on delivery tab. */
        this.allocationOutOfCoverageMessage = null;
        this.newAddressForm = {
            ...SallaBulletDelivery.INITIAL_ADDRESS_FORM,
        };
        /**
         * Submit add-address form: create address via API then switch back to list.
         */
        this.handleSubmitAddAddress = async (e) => {
            e.preventDefault();
            if (this.savingAddress)
                return;
            if (!this.selectedCountry || !this.selectedCity)
                return;
            if (!requireRegionAndDistrictForSA(this.selectedCountry?.code, this.selectedRegion?.id, this.selectedDistrict?.id ?? (this.districtName?.trim() || null)))
                return;
            this.savingAddress = true;
            try {
                const payload = this.buildAddressLocationPayload();
                const { success, address } = await bulletDeliveryAPI.saveAddressLocation(payload);
                if (success) {
                    await this.loadSavedAddresses();
                    const latest = address ?? this.savedAddresses[0] ?? null;
                    if (latest?.is_in_coverage) {
                        this.selectedSessionAddress = false;
                        this.selectedSavedAddress = latest;
                    }
                    this.addressCreated.emit({ address: latest });
                    this.viewMode = "main";
                    this.districtName = "";
                }
                else {
                    Salla.notify?.error(Salla.lang.get("common.errors.error_occurred"));
                }
            }
            catch (error) {
                console.error("SallaBulletDelivery: Error saving address", error);
                Salla.notify?.error(Salla.lang.get("common.errors.error_occurred"));
            }
            finally {
                this.savingAddress = false;
            }
        };
    }
    /** Whether to show delivery + pickup tabs. From store.settings.bullet_delivery.supports_pickup */
    get supportsPickup() {
        return !!Salla.config.get("store.settings.bullet_delivery.supports_pickup");
    }
    /** The modal opening strategy: 'first_visit' | 'on_cart_click' | 'after_add_to_cart' */
    get openingType() {
        return Salla.config.get("store.settings.bullet_delivery.settings.type");
    }
    /** Whether the modal is required (cannot be closed/skipped). From store.settings.bullet_delivery.settings.is_required */
    get isRequired() {
        return !!Salla.config.get("store.settings.bullet_delivery.settings.is_required");
    }
    /**
     * Opens the bullet delivery modal
     */
    async open() {
        // If component is not ready yet, mark as pending and return
        if (!this.componentReady || !this.modal) {
            this.pendingOpen = true;
            return;
        }
        // Reset state and set loading before opening
        this.resetState();
        this.loading = true;
        // Open modal immediately to show skeleton/loading state
        this.modal.open();
        try {
            // Check login status
            const userType = Salla.config.get("user.type");
            this.isLoggedIn = userType !== "guest";
            const savedIntent = this.getStoredIntent();
            // When opening from a stored intent, treat its IDs as preselected targets
            if (savedIntent?.type === "address" && savedIntent.address_id) {
                this.preselectedAddressId = savedIntent.address_id;
            }
            if (savedIntent?.type === "branch" && savedIntent.branch_id) {
                this.preselectedBranchId = savedIntent.branch_id;
            }
            const hasStoredAddressIds = !this.isLoggedIn &&
                savedIntent?.type === "address" &&
                getIntentCountryId(savedIntent) != null &&
                getIntentCityId(savedIntent) != null;
            // Set default tab first so we can lazy-load tab-specific data.
            if (this.supportsPickup &&
                savedIntent?.type &&
                (savedIntent.type === "address" || savedIntent.type === "branch")) {
                this.activeTab = savedIntent.type;
            }
            else {
                this.activeTab = "address";
            }
            // Load countries always; load saved addresses only when address tab is active (lazy tab data).
            await Promise.all([
                this.loadCountries(hasStoredAddressIds),
                this.isLoggedIn && this.activeTab === "address"
                    ? this.loadSavedAddresses().then(() => {
                        this.savedAddressesLoaded = true;
                    })
                    : Promise.resolve(),
            ]);
            // Auto-select preselected address if provided
            if (this.isLoggedIn && this.preselectedAddressId) {
                const address = this.savedAddresses.find((a) => a.id === this.preselectedAddressId);
                if (address) {
                    this.selectedSessionAddress = false;
                    this.selectedSavedAddress = address;
                }
            }
            // Load branches only when branch tab is active (lazy tab data)
            // Use lat/lng from session storage if available for more accurate results
            if (this.activeTab === "branch") {
                const intentLat = getIntentLatitude(savedIntent);
                const intentLng = getIntentLongitude(savedIntent);
                if (intentLat != null && intentLng != null) {
                    await this.loadBranchesWithLocation(intentLat, intentLng);
                }
                else {
                    await this.loadBranches();
                }
                // Auto-select preselected branch if provided
                if (this.preselectedBranchId) {
                    const branch = this.branches.find((b) => b.id === this.preselectedBranchId);
                    if (branch) {
                        this.selectedBranch = branch;
                    }
                }
            }
            // Prefill form data from session storage
            await this.prefillFromSessionStorage();
        }
        catch (e) {
            console.error("SallaBulletDelivery: Error loading data", e);
        }
        finally {
            // Only hide loading after all data is loaded and UI is ready
            this.loading = false;
            this.modal?.stopLoading();
        }
    }
    /**
     * Closes the bullet delivery modal
     */
    async close() {
        this.bulletDeliveryClosed.emit();
        return this.modal?.close();
    }
    resetState() {
        this.viewMode = "main";
        this.selectedCountry = null;
        this.selectedRegion = null;
        this.selectedCity = null;
        this.selectedDistrict = null;
        this.districtName = "";
        this.selectedSavedAddress = null;
        this.selectedSessionAddress = false;
        this.selectedBranch = null;
        this.branchSearchQuery = "";
        this.regions = [];
        this.cities = [];
        this.districts = [];
        this.branches = [];
        this.filteredBranches = [];
        this.newAddressForm = { ...SallaBulletDelivery.INITIAL_ADDRESS_FORM };
        // Reset new state variables
        this.showCartWillBeClearedBanner = false;
        this.allocationOutOfCoverageMessage = null;
        this.locationError = "";
        this.savedAddressesLoaded = false;
        this.loadingSavedAddresses = false;
    }
    getIntentStorage() {
        const rememberLastSession = Boolean(Salla.config.get("store.settings.bullet_delivery.settings.remember_last_session"));
        return Salla.storage[rememberLastSession ? "store" : "session"];
    }
    getStoredIntent() {
        return this.getIntentStorage().get(this.intentStorageKey);
    }
    setStoredIntent(intent) {
        this.getIntentStorage().set(this.intentStorageKey, intent);
    }
    removeStoredIntent() {
        this.getIntentStorage().remove(this.intentStorageKey);
    }
    /**
     * Single prefill flow: load geography from stored intent and set selected location + form.
     * Assumes this.countries is already loaded. No-op if no intent or no city in intent.
     */
    async prefillAddressFromStoredIntent() {
        const intent = this.getStoredIntent();
        if (!intent || intent.type !== "address" || getIntentCityId(intent) == null)
            return;
        const countryId = getIntentCountryId(intent);
        if (!countryId)
            return;
        const country = this.countries.find((c) => String(c.id) === String(countryId));
        if (!country)
            return;
        this.selectedCountry = country;
        this.updateNewAddressForm({ country_id: Number(country.id) });
        const isSA = isSaudiArabia(country.code);
        const cityId = getIntentCityId(intent);
        const regionId = getIntentRegionId(intent);
        const districtId = getIntentDistrictId(intent);
        const districtName = intent.address_details?.district?.name;
        const [regions, cities, districts] = await Promise.all([
            isSA ? bulletDeliveryAPI.getRegions(country.id) : Promise.resolve([]),
            bulletDeliveryAPI.getCities(country.id, isSA && regionId != null ? regionId : undefined),
            cityId != null
                ? bulletDeliveryAPI.getDistricts(Number(cityId))
                : Promise.resolve([]),
        ]);
        if (isSA)
            this.regions = regions;
        this.cities = cities;
        this.districts = districts;
        this.loadingCities = false;
        this.loadingRegions = false;
        this.loadingDistricts = false;
        if (isSA && regionId != null) {
            const region = this.regions.find((r) => String(r.id) === String(regionId));
            if (region) {
                this.selectedRegion = region;
                this.updateNewAddressForm({ region_id: region.id });
            }
        }
        if (cityId != null && this.cities.length > 0) {
            const city = this.cities.find((c) => String(c.id) === String(cityId));
            if (city) {
                this.selectedCity = city;
                this.updateNewAddressForm({ city_id: city.id, city });
            }
        }
        if (this.districts.length > 0) {
            const district = this.findPrefillDistrict(districtId, districtName);
            if (district) {
                this.updateNewAddressForm({ district_id: district.id, district });
                this.selectedDistrict = district;
            }
        }
    }
    updateNewAddressForm(partial) {
        this.newAddressForm = { ...this.newAddressForm, ...partial };
    }
    /** Whether allocation completed successfully this session (modal won't auto-open again). */
    hasBeenShownThisSession() {
        return !!Salla.storage.session.get(this.sessionShownKey);
    }
    /** Mark allocation as completed this session (called only after allocateScope succeeds). */
    markShownThisSession() {
        Salla.storage.session.set(this.sessionShownKey, true);
    }
    async loadCountries(skipEagerSubFetch = false) {
        const forBranch = this.activeTab === "branch";
        this.countries = await bulletDeliveryAPI.getCountries(forBranch);
        // Preselect country from bullet_delivery.customer_location.country (SA | EG | ...)
        const configCountryCode = Salla.config.get("store.settings.bullet_delivery.customer_location.country");
        const countryToSelect = configCountryCode && this.countries.length > 0
            ? this.countries.find((c) => String(c.code).toUpperCase() ===
                String(configCountryCode).toUpperCase())
            : null;
        if (countryToSelect) {
            this.selectedCountry = countryToSelect;
            this.updateNewAddressForm({
                country_id: Number(this.selectedCountry.id),
            });
            // Skip eager sub-fetches when stored intent IDs exist — prefillFromSessionStorage
            // will fire regions/cities/districts in parallel using the stored IDs directly.
            if (!skipEagerSubFetch) {
                if (isSaudiArabia(this.selectedCountry.code)) {
                    await this.loadRegions(this.selectedCountry.id);
                }
                else {
                    await this.loadCities(this.selectedCountry.id);
                }
            }
        }
    }
    async loadCities(countryId, regionId) {
        this.loadingCities = true;
        this.cities = [];
        this.districts = [];
        this.selectedCity = null;
        this.selectedDistrict = null;
        this.districtName = "";
        try {
            this.cities = await bulletDeliveryAPI.getCities(countryId, regionId);
        }
        finally {
            this.loadingCities = false;
        }
    }
    async loadDistricts(cityId) {
        this.loadingDistricts = true;
        this.districts = [];
        this.selectedDistrict = null;
        this.districtName = "";
        try {
            this.districts = await bulletDeliveryAPI.getDistricts(cityId);
        }
        finally {
            this.loadingDistricts = false;
        }
    }
    findPrefillDistrict(districtId, districtName) {
        if (this.districts.length === 0) {
            return undefined;
        }
        if (districtId != null) {
            const byId = this.districts.find((d) => String(d.id) === String(districtId));
            if (byId) {
                return byId;
            }
        }
        const normalizedName = districtName?.trim().toLowerCase();
        if (!normalizedName) {
            return undefined;
        }
        return this.districts.find((d) => {
            const arName = d.name?.trim().toLowerCase();
            const enName = d.name_en?.trim().toLowerCase();
            return arName === normalizedName || enName === normalizedName;
        });
    }
    async loadBranches() {
        this.loadingBranches = true;
        try {
            this.branches = await bulletDeliveryAPI.getBranches({
                country_id: this.selectedCountry?.id,
            });
            this.filterBranches();
        }
        finally {
            this.loadingBranches = false;
        }
    }
    filterBranches() {
        this.filteredBranches = filterBranches(this.branches, this.branchSearchQuery);
    }
    /**
     * Prefill form fields from session storage
     */
    async prefillFromSessionStorage() {
        try {
            const intent = this.getStoredIntent();
            if (!intent) {
                return;
            }
            // Prefill address tab for guest users (from stored address_details IDs only)
            if (this.activeTab === "address" &&
                intent.type === "address" &&
                !this.isLoggedIn) {
                await this.prefillAddressFromStoredIntent();
            }
            // Prefill branch tab
            if (this.activeTab === "branch" && intent.type === "branch") {
                const countryId = getIntentCountryId(intent);
                if (countryId != null) {
                    const country = this.countries.find((c) => String(c.id) === String(countryId));
                    if (country) {
                        this.selectedCountry = country;
                        this.updateNewAddressForm({ country_id: Number(country.id) });
                        const branchId = getIntentBranchId(intent);
                        if (branchId != null && this.branches.length > 0) {
                            const branch = this.branches.find((b) => b.id === branchId);
                            if (branch) {
                                this.selectedBranch = branch;
                            }
                        }
                    }
                }
            }
        }
        catch (error) {
            console.error("[BulletDelivery] Error prefilling from session storage:", error);
        }
    }
    async loadSavedAddresses() {
        this.savedAddresses = await bulletDeliveryAPI.getSavedAddresses();
        // If session address exists (guest intent before login), select it by default
        const intent = this.getStoredIntent();
        const hasSessionAddress = hasSessionAddressIntent(intent);
        if (hasSessionAddress) {
            this.selectedSessionAddress = true;
            this.selectedSavedAddress = null;
        }
        else {
            // Select default address if available and valid
            const defaultAddress = this.savedAddresses.find((a) => a.is_default && a.is_in_coverage !== false);
            if (defaultAddress) {
                this.selectedSavedAddress = defaultAddress;
            }
            else {
                // Select first valid address
                const validAddress = this.savedAddresses.find((a) => a.is_in_coverage !== false);
                if (validAddress) {
                    this.selectedSavedAddress = validAddress;
                }
            }
        }
    }
    applyCountryChange(newCountry) {
        this.selectedCountry = newCountry;
        this.selectedRegion = null;
        this.regions = [];
        this.cities = [];
        this.districts = [];
        this.selectedCity = null;
        this.selectedDistrict = null;
        this.districtName = "";
        this.allocationOutOfCoverageMessage = null;
        if (this.selectedCountry) {
            this.updateNewAddressForm({
                country_id: Number(this.selectedCountry.id),
                region_id: undefined,
                city_id: undefined,
                district_id: undefined,
                city: undefined,
                district: undefined,
            });
            if (this.activeTab === "branch") {
                this.loadBranches();
            }
            else if (isSaudiArabia(this.selectedCountry.code)) {
                this.loadRegions(this.selectedCountry.id);
            }
            else {
                this.loadCities(this.selectedCountry.id);
            }
        }
    }
    async loadRegions(countryId) {
        this.loadingRegions = true;
        this.regions = [];
        this.selectedRegion = null;
        this.cities = [];
        this.districts = [];
        this.selectedCity = null;
        this.selectedDistrict = null;
        this.districtName = "";
        try {
            this.regions = await bulletDeliveryAPI.getRegions(countryId);
        }
        finally {
            this.loadingRegions = false;
        }
    }
    /**
     * AC7: Guest-to-Registered Sync
     * Syncs guest session location to user profile on login
     */
    async handleGuestToRegisteredSync() {
        try {
            // Check if user is now logged in
            const userId = Salla.config.get("user.id");
            if (!userId)
                return;
            // Get the guest session location from session storage
            const guestIntent = this.getStoredIntent();
            if (!guestIntent || guestIntent.type !== "address") {
                // No guest address intent to sync, or it's a branch intent
                return;
            }
            // Get user's saved addresses
            const userAddresses = await bulletDeliveryAPI.getSavedAddresses();
            const matchingAddress = findMatchingAddress(guestIntent, userAddresses);
            if (matchingAddress) {
                await this.setAddressAsActive(matchingAddress.id);
                const updatedIntent = {
                    ...guestIntent,
                    address_id: matchingAddress.id,
                    address_details: this.buildIntentAddressFromSaved(matchingAddress),
                };
                this.setStoredIntent(updatedIntent);
            }
            else if (getIntentCityId(guestIntent)) {
                // No match - create new address from guest session data
                const newAddress = await this.createAddressFromGuestSession(guestIntent);
                if (newAddress) {
                    const updatedIntent = {
                        ...guestIntent,
                        address_id: newAddress.id,
                        address_details: this.buildIntentAddressFromSaved(newAddress),
                    };
                    this.setStoredIntent(updatedIntent);
                    this.addressCreated.emit({ address: newAddress });
                }
            }
            // Emit event to notify that sync is complete
            Salla.event.emit("bullet-delivery::guest-synced", {
                guest_intent: guestIntent,
                matched_address_id: matchingAddress?.id,
            });
        }
        catch (error) {
            console.error("SallaBulletDelivery: Error syncing guest session to user profile", error);
        }
    }
    /**
     * Set an address as active/last used
     */
    async setAddressAsActive(addressId) {
        try {
            // Call API to mark address as default/active
            await Salla.address?.api?.setDefault?.(addressId);
        }
        catch (error) {
            // If setDefault API doesn't exist, just update local state
            console.warn("SallaBulletDelivery: Could not set address as default", error);
        }
    }
    /**
     * Create a new address from guest session: only country, city and district IDs (no street, etc.).
     * Uses address/location API; after fetch we return the newly created address from the list.
     */
    async createAddressFromGuestSession(guestIntent) {
        const countryId = getIntentCountryId(guestIntent);
        const cityId = getIntentCityId(guestIntent);
        if (!countryId || !cityId)
            return null;
        const isSA = isSaudiArabia(getIntentCountryCode(guestIntent));
        const districtName = guestIntent.address_details?.district?.name ?? "";
        const payload = {
            country_id: countryId,
            region_id: getIntentRegionId(guestIntent),
            city_id: cityId,
            district_id: getIntentDistrictId(guestIntent),
            description: districtName,
        };
        if (isSA) {
            payload.local = districtName;
        }
        try {
            const { success, address } = await bulletDeliveryAPI.saveAddressLocation(payload);
            if (!success)
                return null;
            if (address)
                return address;
            const addresses = await bulletDeliveryAPI.getSavedAddresses();
            const districtId = getIntentDistrictId(guestIntent);
            const newAddress = addresses.find((a) => (a.city_id ?? a.city?.id) === cityId &&
                (a.district_id ?? a.district?.id) === districtId) ??
                addresses[0] ??
                null;
            return newAddress;
        }
        catch (error) {
            console.error("SallaBulletDelivery: Could not create address from guest session", error);
            return null;
        }
    }
    emitHeaderContextUpdate(result, address, branch) {
        let locationText = "";
        if (result.type === "address") {
            if (address) {
                locationText =
                    address.district?.name && address.city?.name
                        ? `${address.district.name}، ${address.city.name}`
                        : (address.city?.name ?? "");
            }
            else if (this.selectedCity) {
                locationText = this.selectedDistrict
                    ? `${this.selectedDistrict.name}، ${this.selectedCity.name}`
                    : this.selectedCity.name;
            }
            else if (result.address_details) {
                const parts = [
                    result.address_details.district?.name,
                    result.address_details.city.name,
                    result.address_details.country.name,
                ].filter(Boolean);
                locationText =
                    parts.join("، ") || result.address_details.short_address || "";
            }
        }
        else if (branch) {
            locationText = branch.name;
        }
        else if (result.branch_details) {
            locationText = result.branch_details.city ?? result.branch_details.name;
        }
        // Build display text with prefix (e.g., "Delivering to: Riyadh" or "Pickup from: Branch Name")
        const prefix = result.type === "address"
            ? Salla.lang.get("mobile_app.strings.delivery")
            : Salla.lang.get("mobile_app.strings.pickup");
        const displayText = locationText ? `${prefix}: ${locationText}` : "";
        this.headerContextUpdate.emit({
            type: result.type,
            display_text: displayText,
            country_code: result.country_code,
            city: this.selectedCity?.name ?? result.address_details?.city?.name,
            district: this.selectedDistrict?.name ?? result.address_details?.district?.name,
            branch_name: branch?.name ?? result.branch_details?.name,
        });
    }
    handleBranchSearch(event) {
        const input = event.target;
        const query = input.value;
        this.branchSearchQuery = query;
        if (this.branchSearchDebounceTimer)
            clearTimeout(this.branchSearchDebounceTimer);
        if (!query?.trim()) {
            this.branchSearchQuery = "";
            void this.loadBranches();
            return;
        }
        this.branchSearchDebounceTimer = setTimeout(async () => {
            if (query.trim().length < 2)
                return;
            this.loadingBranches = true;
            try {
                this.branches = await bulletDeliveryAPI.getBranches({
                    country_id: this.selectedCountry?.id,
                    query,
                });
                this.filterBranches();
            }
            finally {
                this.loadingBranches = false;
            }
        }, BRANCH_SEARCH_DEBOUNCE_MS);
    }
    async loadBranchesWithLocation(lat, lng) {
        this.loadingBranches = true;
        try {
            this.branches = await bulletDeliveryAPI.getBranches({
                country_id: this.selectedCountry?.id,
                lat,
                lng,
            });
            this.filterBranches();
            // Auto-select country from first branch if no country is selected
            if (!this.selectedCountry && this.branches.length > 0) {
                const firstBranch = this.branches[0];
                if (firstBranch.country) {
                    const matchingCountry = this.countries.find((c) => c.id === firstBranch.country.id);
                    if (matchingCountry) {
                        this.selectedCountry = matchingCountry;
                        this.updateNewAddressForm({
                            country_id: Number(matchingCountry.id),
                        });
                    }
                }
            }
        }
        finally {
            this.loadingBranches = false;
        }
    }
    /** Locate nearest branch using device geolocation (branches tab only). */
    async handleFindNearestBranch() {
        if (!navigator.geolocation) {
            this.locationError = Salla.lang.get("pages.checkout.current_location_not_supported");
            return;
        }
        this.loadingNearestBranch = true;
        this.locationError = "";
        try {
            const position = await new Promise((resolve, reject) => {
                navigator.geolocation.getCurrentPosition(resolve, reject, {
                    enableHighAccuracy: true,
                    timeout: GEOLOCATION_TIMEOUT,
                    maximumAge: 0,
                });
            });
            const { latitude, longitude } = position.coords;
            await this.loadBranchesWithLocation(latitude, longitude);
            const nearestBranch = findNearestBranch(this.branches, latitude, longitude);
            if (nearestBranch) {
                this.selectedBranch = nearestBranch;
                this.branchSearchQuery = "";
                this.filterBranches();
            }
        }
        catch (error) {
            const geoError = error;
            const code = typeof geoError?.code === "number" ? geoError.code : -1;
            this.locationError = getGeolocationErrorMessage(code);
        }
        finally {
            this.loadingNearestBranch = false;
        }
    }
    handleBranchSelect(branch) {
        this.selectedBranch = branch;
    }
    handleSavedAddressSelect(address) {
        if (address.is_in_coverage === false) {
            // Show toast notification for out of coverage address
            Salla.notify?.error(Salla.lang.get("pages.checkout.shipping_not_available"));
            return;
        }
        this.selectedSavedAddress = address;
        this.selectedSessionAddress = false;
    }
    handleSessionAddressSelect() {
        this.selectedSessionAddress = true;
        this.selectedSavedAddress = null;
    }
    /** Show delete confirmation UI for this address (do not call API yet). */
    /**
     * Switch to add-address form (unified form). Preselects country from session when present.
     */
    async handleAddNewAddress() {
        this.viewMode = "add-address";
        const intent = this.getStoredIntent();
        const sessionCountryId = getIntentCountryId(intent);
        if (sessionCountryId && this.countries.length > 0) {
            const country = this.countries.find((c) => c.id === sessionCountryId);
            if (country) {
                this.selectedCountry = country;
                this.updateNewAddressForm({ country_id: Number(country.id) });
                if (isSaudiArabia(country.code)) {
                    await this.loadRegions(country.id);
                }
                else {
                    await this.loadCities(country.id);
                }
            }
        }
    }
    handleBackToAddressList() {
        this.viewMode = "main";
    }
    /**
     * Build address/location payload from guest-form state (same fields as add-address).
     */
    buildAddressLocationPayload() {
        const districtName = this.selectedDistrict?.name ?? this.districtName?.trim() ?? "";
        return buildAddressLocationPayloadFromSelection({
            countryId: this.selectedCountry?.id ?? 0,
            countryCode: this.selectedCountry?.code,
            regionId: this.selectedRegion?.id,
            cityId: this.selectedCity?.id,
            districtId: this.selectedDistrict?.id,
            description: districtName,
        });
    }
    async handleTabChange(tab) {
        if (this.tabChanging)
            return;
        this.tabChanging = true;
        try {
            this.activeTab = tab;
            this.showCartWillBeClearedBanner = false;
            this.allocationOutOfCoverageMessage = null;
            // Reload countries with for_branch=1 for pickup tab, for_branch=0 for delivery tab
            const forBranch = tab === "branch";
            this.countries = await bulletDeliveryAPI.getCountries(forBranch);
            const previousCountryId = this.selectedCountry?.id;
            if (previousCountryId !== undefined && previousCountryId !== null) {
                this.selectedCountry =
                    this.countries.find((c) => c.id === previousCountryId) || null;
            }
            // Load appropriate data when switching tabs if country is selected
            if (this.selectedCountry) {
                if (tab === "branch") {
                    // Load branches when switching to branch tab
                    // Check session storage for lat/lng to get nearby branches
                    const savedIntent = this.getStoredIntent();
                    const intentLat = getIntentLatitude(savedIntent);
                    const intentLng = getIntentLongitude(savedIntent);
                    if (intentLat != null && intentLng != null) {
                        await this.loadBranchesWithLocation(intentLat, intentLng);
                    }
                    else {
                        await this.loadBranches();
                    }
                    // Preselect branch from session storage if available
                    const branchId = getIntentBranchId(savedIntent);
                    if (branchId != null && this.branches.length > 0) {
                        const branch = this.branches.find((b) => b.id === branchId);
                        if (branch) {
                            this.selectedBranch = branch;
                        }
                    }
                }
                else if (tab === "address") {
                    // Lazy-load saved addresses when switching to address tab (if not loaded yet).
                    if (this.isLoggedIn && !this.savedAddressesLoaded) {
                        this.loadingSavedAddresses = true;
                        try {
                            await this.loadSavedAddresses();
                            this.savedAddressesLoaded = true;
                        }
                        finally {
                            this.loadingSavedAddresses = false;
                        }
                    }
                    if (!this.isLoggedIn) {
                        const savedIntent = this.getStoredIntent();
                        if (savedIntent && getIntentCityId(savedIntent)) {
                            await this.prefillAddressFromStoredIntent();
                        }
                    }
                }
            }
        }
        finally {
            this.tabChanging = false;
        }
    }
    isConfirmDisabled() {
        if (this.activeTab === "branch") {
            return !this.selectedBranch;
        }
        if (this.activeTab === "address") {
            if (this.isLoggedIn && this.selectedSessionAddress) {
                return false; // Session address selected - confirm will save then allocate
            }
            if (this.isLoggedIn && this.selectedSavedAddress) {
                return this.selectedSavedAddress.is_in_coverage === false;
            }
            // Guest: require country + city; for SA require region + district
            if (!this.selectedCountry || !this.selectedCity)
                return true;
            return !requireRegionAndDistrictForSA(this.selectedCountry?.code, this.selectedRegion?.id, this.selectedDistrict?.id ?? (this.districtName?.trim() || null));
        }
        return false;
    }
    handleLogin() {
        Salla.event.dispatch("bullet-delivery::close");
        Salla.event.dispatch("login::open");
    }
    buildIntentAddress() {
        if (this.activeTab !== "address" ||
            !this.selectedCountry ||
            !this.selectedCity) {
            return undefined;
        }
        const country = {
            id: this.selectedCountry.id,
            name: this.selectedCountry.name,
            code: this.selectedCountry.code,
        };
        const city = {
            id: this.selectedCity.id,
            name: this.selectedCity.name,
        };
        const region = this.selectedRegion
            ? {
                id: this.selectedRegion.id,
                name: this.selectedRegion.name,
                code: this.selectedRegion.code,
            }
            : undefined;
        const district = this.selectedDistrict
            ? { id: this.selectedDistrict.id, name: this.selectedDistrict.name }
            : this.districtName
                ? { id: 0, name: this.districtName }
                : undefined;
        const short_address = this.selectedSavedAddress?.short_address;
        return {
            country,
            region,
            city,
            district,
            short_address,
        };
    }
    buildIntentAddressFromSaved(addr) {
        return {
            country: {
                id: addr.country?.id ?? addr.country_id ?? 0,
                name: addr.country?.name ?? "",
                code: addr.country?.code,
            },
            region: addr.region
                ? { id: addr.region.id, name: addr.region.name, code: addr.region.code }
                : undefined,
            city: {
                id: addr.city?.id ?? addr.city_id ?? 0,
                name: addr.city?.name ?? "",
            },
            district: addr.district
                ? { id: addr.district.id, name: addr.district.name }
                : undefined,
            short_address: addr.short_address,
        };
    }
    buildIntentBranch() {
        if (this.activeTab !== "branch" || !this.selectedBranch)
            return undefined;
        const b = this.selectedBranch;
        const lat = b.location?.lat != null ? Number(b.location.lat) : undefined;
        const lng = b.location?.lng != null ? Number(b.location.lng) : undefined;
        return {
            id: b.id,
            name: b.name,
            city: b.city?.name,
            latitude: lat,
            longitude: lng,
        };
    }
    async handleConfirm() {
        if (this.isConfirmDisabled() || this.confirming)
            return;
        this.confirming = true;
        this.confirmBtn?.load();
        let scheduledRedirect = false;
        try {
            const result = {
                type: this.activeTab,
                country_id: Number(this.selectedCountry?.id) || 0,
                country_code: this.selectedCountry?.code || "",
            };
            let selectedAddress;
            let selectedBranchData;
            let usedSessionAddress = false;
            if (this.activeTab === "address") {
                if (this.isLoggedIn && this.selectedSessionAddress) {
                    usedSessionAddress = true;
                    const savedAddress = await this.saveSessionAddressToProfile();
                    if (!savedAddress) {
                        Salla.notify?.error(Salla.lang.get("common.errors.error_occurred"));
                        this.confirmBtn?.stop();
                        return;
                    }
                    result.address_id = savedAddress.id;
                    selectedAddress = savedAddress;
                    this.selectedSessionAddress = false;
                    this.selectedSavedAddress = savedAddress;
                    result.address_details = this.buildIntentAddressFromSaved(savedAddress);
                }
                else if (this.isLoggedIn && this.selectedSavedAddress) {
                    result.address_id = this.selectedSavedAddress.id;
                    selectedAddress = this.selectedSavedAddress;
                    result.address_details = this.buildIntentAddressFromSaved(this.selectedSavedAddress);
                }
                else {
                    result.address_details = this.buildIntentAddress();
                }
            }
            else {
                result.branch_details = this.buildIntentBranch();
                result.branch_id = this.selectedBranch?.id;
                selectedBranchData = this.selectedBranch || undefined;
            }
            // Build scope allocation payload based on scenario
            let allocationPayload;
            if (this.activeTab === "address") {
                if (this.isLoggedIn && this.selectedSavedAddress) {
                    // Case 2: Logged-in user with saved address
                    allocationPayload = {
                        type: "address",
                        address_id: String(this.selectedSavedAddress.id),
                    };
                }
                else {
                    // Case 1: Guest user or logged-in user without saved address
                    allocationPayload = {
                        type: "address",
                        country_id: String(this.selectedCountry?.id || 0),
                        city_id: String(this.selectedCity?.id || ""),
                    };
                    if (isSaudiArabia(this.selectedCountry?.code ?? '')) {
                        if (this.selectedRegion?.id) {
                            allocationPayload.region_id = String(this.selectedRegion.id);
                        }
                        allocationPayload.district_id = String(this.selectedDistrict?.id || "");
                    }
                }
            }
            else {
                // Case 3: Pickup
                allocationPayload = {
                    type: "branch",
                    branch_id: String(this.selectedBranch?.id || ""),
                };
            }
            // Make API call to allocate scope
            const allocationResult = await bulletDeliveryAPI.allocateScope(allocationPayload);
            if (!allocationResult.success) {
                console.error("BulletDelivery: Scope allocation failed", allocationResult.error);
                const errorMessage = allocationResult.error ||
                    (this.activeTab === "address"
                        ? Salla.lang.get("pages.checkout.address_out_of_coverage")
                        : Salla.lang.get("pages.checkout.failed_to_set_pickup"));
                Salla.notify?.error(errorMessage);
                this.allocationOutOfCoverageMessage = errorMessage;
                this.confirmBtn?.stop();
                this.confirming = false;
                return;
            }
            // Process successful allocation response
            let allocatedScopeId;
            if (allocationResult.data) {
                const allocationData = allocationResult.data;
                allocatedScopeId = allocationData.id;
                const allocationType = allocationData?.allocation?.type;
                const allocationBranchId = allocationData?.allocation?.branch_id;
                // Set headers via Salla.api.setHeader
                if (allocationType) {
                    Salla.api.setHeader("s-scope-allocation-type", allocationType);
                }
                if (allocationBranchId) {
                    Salla.api.setHeader("s-scope-allocation-id", String(allocationBranchId));
                }
                // Persist allocation data in scope storage (single source of truth for all components)
                const existingScope = Salla.config.get("store.scope", Salla.storage.get("scope")) || {};
                Salla.storage.set("scope", {
                    ...existingScope,
                    id: allocatedScopeId,
                    allocation_type: allocationType,
                    allocation_id: allocationBranchId
                        ? String(allocationBranchId)
                        : undefined,
                });
                // Store headers in result object for persistence after page reload
                result.allocation_headers = {
                    "s-scope-allocation-type": allocationType,
                    "s-scope-allocation-id": allocationBranchId
                        ? String(allocationBranchId)
                        : undefined,
                };
            }
            // After allocation success: mark session so we don't show modal again this session
            this.markShownThisSession();
            // After allocation success: update session storage for guest prefill on next visit
            // When session address was confirmed, remove intent so logout shows a reset form
            if (usedSessionAddress) {
                this.removeStoredIntent();
            }
            else {
                this.setStoredIntent(result);
            }
            if (allocatedScopeId) {
                await bulletDeliveryAPI.setDeliveryScope(allocatedScopeId);
            }
            this.bulletDeliveryConfirmed.emit({
                ...result,
                address: selectedAddress,
                branch: selectedBranchData,
            });
            // Emit header context update for global header sync
            this.emitHeaderContextUpdate(result, selectedAddress, selectedBranchData);
            scheduledRedirect = true;
            setTimeout(() => {
                this.confirmBtn?.stop();
                this.close();
                this.confirming = false;
                if (allocatedScopeId) {
                    const scope = Salla.storage.get("scope") || {};
                    let url = Salla.helpers.addParamToUrl("scope", allocatedScopeId);
                    if (scope.allocation_id) {
                        url = Salla.helpers.addParamToUrl("allocation_id", scope.allocation_id, url);
                    }
                    if (scope.allocation_type) {
                        url = Salla.helpers.addParamToUrl("allocation_type", scope.allocation_type, url);
                    }
                    window.location.replace(url);
                }
                else {
                    window.location.reload();
                }
            }, 500);
        }
        finally {
            if (!scheduledRedirect) {
                this.confirming = false;
            }
        }
    }
    getCartViewConfig(intent) {
        if (!intent || !Salla.url.is_page("cart"))
            return null;
        const isBranch = intent.type === "branch";
        const isAddress = intent.type === "address";
        if ((isBranch && !this.supportsPickup) || (!isBranch && !isAddress))
            return null;
        return {
            title: isBranch
                ? Salla.lang.get("pages.checkout.pickup_option")
                : Salla.lang.get("pages.checkout.delivery_option"),
            subtitle: getIntentSubtitle(intent, Salla.lang.get("common.elements.to"), ""),
            buttonText: Salla.lang.get("common.elements.edit"),
            isAddress,
            intent,
        };
    }
    /** Creates the cart view DOM for mounting into cart.items.start hook */
    createCartViewElement() {
        const intent = this.getStoredIntent();
        const config = this.getCartViewConfig(intent);
        if (!config)
            return null;
        const { title, subtitle, buttonText, intent: configIntent } = config;
        const openPayload = configIntent.type === "branch"
            ? { preselected_branch_id: getIntentBranchId(configIntent) }
            : { preselected_address_id: configIntent.address_id };
        const icon = document.createElement("span");
        icon.className =
            "s-bullet-delivery-cart-view-icon s-bullet-delivery-tab-icon";
        icon.setAttribute("aria-hidden", "true");
        icon.innerHTML = location.Location;
        const titleEl = document.createElement("span");
        titleEl.className = "s-bullet-delivery-cart-view-title";
        titleEl.textContent = title;
        const subtitleEl = document.createElement("span");
        subtitleEl.className = "s-bullet-delivery-cart-view-subtitle";
        subtitleEl.textContent = subtitle;
        const text = document.createElement("div");
        text.className = "s-bullet-delivery-cart-view-text";
        text.append(titleEl, subtitleEl);
        const info = document.createElement("div");
        info.className = "s-bullet-delivery-cart-view-info";
        info.append(icon, text);
        const btn = document.createElement("salla-button");
        btn.className = "s-bullet-delivery-cart-view-button";
        btn.setAttribute("fill", "outline");
        btn.setAttribute("color", "primary");
        btn.textContent = buttonText;
        btn.addEventListener("click", () => Salla.event.emit("bullet-delivery::open", openPayload));
        const wrapper = document.createElement("div");
        wrapper.className = "s-bullet-delivery-cart-view";
        wrapper.append(info, btn);
        return wrapper;
    }
    renderSkeleton() {
        return (index.h("div", { slot: "loading", class: "s-bullet-delivery-skeleton" }, index.h("div", { class: "s-bullet-delivery-skeleton-header" }, index.h("salla-skeleton", { class: "s-bullet-delivery-skeleton-title", height: "28px", width: "68%" }), index.h("salla-skeleton", { class: "s-bullet-delivery-skeleton-subtitle", height: "18px", width: "84%" })), index.h("div", { class: {
                "s-bullet-delivery-skeleton-tabs": true,
                "s-bullet-delivery-skeleton-tabs--single": !this.supportsPickup,
            } }, index.h("div", { class: "s-bullet-delivery-skeleton-tab" }, index.h("salla-skeleton", { class: "s-bullet-delivery-skeleton-tab-icon", height: "32px", width: "32px" }), index.h("salla-skeleton", { class: "s-bullet-delivery-skeleton-tab-label", height: "18px", width: "92px" })), this.supportsPickup && (index.h("div", { class: "s-bullet-delivery-skeleton-tab" }, index.h("salla-skeleton", { class: "s-bullet-delivery-skeleton-tab-icon", height: "32px", width: "32px" }), index.h("salla-skeleton", { class: "s-bullet-delivery-skeleton-tab-label", height: "18px", width: "84px" })))), index.h("div", { class: "s-bullet-delivery-skeleton-content" }, index.h("div", { class: "s-bullet-delivery-skeleton-field" }, index.h("salla-skeleton", { class: "s-bullet-delivery-skeleton-field-label", height: "16px", width: "96px" }), index.h("salla-skeleton", { class: "s-bullet-delivery-skeleton-field-control", height: "56px", width: "100%" })), index.h("div", { class: "s-bullet-delivery-skeleton-field" }, index.h("salla-skeleton", { class: "s-bullet-delivery-skeleton-field-label", height: "16px", width: "104px" }), index.h("salla-skeleton", { class: "s-bullet-delivery-skeleton-field-control", height: "56px", width: "100%" })), index.h("div", { class: "s-bullet-delivery-skeleton-field" }, index.h("salla-skeleton", { class: "s-bullet-delivery-skeleton-field-label", height: "16px", width: "88px" }), index.h("salla-skeleton", { class: "s-bullet-delivery-skeleton-field-control", height: "56px", width: "100%" }))), index.h("div", { class: "s-bullet-delivery-skeleton-footer" }, index.h("salla-skeleton", { height: "52px", width: "100%" }))));
    }
    renderTabs() {
        // Show tabs only when pickup is supported; otherwise delivery-only view with no tabs
        if (!this.supportsPickup)
            return null;
        return (index.h("div", { class: "s-bullet-delivery-tabs" }, index.h("button", { type: "button", class: {
                "s-bullet-delivery-tab": true,
                "s-bullet-delivery-tab--active": this.activeTab === "address",
            }, onClick: () => this.handleTabChange("address") }, index.h("span", { class: "s-bullet-delivery-tab-icon", innerHTML: location.Location, "aria-hidden": "true" }), index.h("span", null, Salla.lang.get("mobile_app.strings.delivery"))), index.h("button", { type: "button", class: {
                "s-bullet-delivery-tab": true,
                "s-bullet-delivery-tab--active": this.activeTab === "branch",
            }, onClick: () => this.handleTabChange("branch") }, index.h("span", { class: "s-bullet-delivery-tab-icon", innerHTML: Store, "aria-hidden": "true" }), index.h("span", null, Salla.lang.get("mobile_app.strings.pickup")))));
    }
    renderAlert(message) {
        return (index.h("div", { class: "s-bullet-delivery-allocation-out-of-coverage", role: "alert" }, index.h("span", { class: "s-bullet-delivery-allocation-out-of-coverage-icon", "aria-hidden": "true" }, index.h("svg", { width: "20", height: "20", viewBox: "0 0 20 20", fill: "none", xmlns: "http://www.w3.org/2000/svg", role: "img", "aria-hidden": "true" }, index.h("title", null, "Information"), index.h("circle", { cx: "10", cy: "10", r: "9", stroke: "currentColor", "stroke-width": "1.5", fill: "currentColor" }), index.h("text", { x: "10", y: "14", "text-anchor": "middle", fill: "white", "font-size": "12", "font-weight": "bold" }, "i"))), index.h("span", { class: "s-bullet-delivery-allocation-out-of-coverage-text" }, message)));
    }
    renderAllocationOutOfCoverageAlert() {
        const message = this.allocationOutOfCoverageMessage ||
            Salla.lang.get("pages.checkout.shipping_not_available");
        return this.renderAlert(message);
    }
    renderCountrySelect() {
        return (index.h("div", { class: "s-bullet-delivery-field" }, index.h("label", { class: "s-bullet-delivery-label", htmlFor: "bullet-delivery-country" }, Salla.lang.get("blocks.buy_as_gift.receiver_country"), index.h("span", { class: "text-red-500" }, " *")), index.h("select", { id: "bullet-delivery-country", class: "form-input", onChange: (e) => {
                const select = e.target;
                const countryId = Number.parseInt(select.value);
                const country = this.countries.find((c) => c.id === countryId);
                // A "switch" means the user already had a country and picked a different one
                const isCountrySwitch = this.selectedCountry &&
                    country &&
                    this.selectedCountry.code !== country.code;
                // Show the info banner whenever a country is changed to a different one
                this.showCartWillBeClearedBanner = !!isCountrySwitch;
                this.applyCountryChange(country || null);
            } }, index.h("option", { value: "", disabled: true, selected: !this.selectedCountry }, Salla.lang.get("pages.checkout.select_country")), this.countries.map((country) => (index.h("option", { key: country.id, value: country.id, selected: this.selectedCountry?.id === country.id }, country.name))))));
    }
    renderSelectLoader() {
        return (index.h("span", { class: "s-bullet-delivery-select-loader", "aria-hidden": "true" }, index.h("salla-loading", { size: "16", width: "3" })));
    }
    renderLoadingLabel(label, htmlFor, isLoading = false, isRequired = true) {
        return (index.h("label", { class: "s-bullet-delivery-label", htmlFor: htmlFor }, index.h("span", { class: "s-bullet-delivery-label-content" }, index.h("span", null, label, isRequired && index.h("span", { class: "text-red-500" }, " *")), isLoading && this.renderSelectLoader())));
    }
    /** Region select (SA only): shown when SA is selected; city and district disabled until region selected. */
    renderRegionSelect() {
        const isSA = isSaudiArabia(this.selectedCountry?.code ?? '');
        if (!isSA || !this.selectedCountry)
            return null;
        return (index.h("div", { class: "s-bullet-delivery-field" }, this.renderLoadingLabel(Salla.lang.get("pages.checkout.region_field"), "bullet-delivery-region", this.loadingRegions), index.h("select", { id: "bullet-delivery-region", class: "form-input", disabled: this.loadingRegions || this.regions.length === 0, onChange: (e) => {
                const select = e.target;
                const regionId = Number.parseInt(select.value);
                const region = this.regions.find((r) => r.id === regionId);
                this.selectedRegion = region || null;
                this.updateNewAddressForm({
                    region_id: region?.id,
                    city_id: undefined,
                    district_id: undefined,
                    city: undefined,
                    district: undefined,
                });
                if (region && this.selectedCountry) {
                    this.loadCities(this.selectedCountry.id, region.id);
                }
            } }, index.h("option", { value: "", disabled: true, selected: !this.selectedRegion }, Salla.lang.get("pages.checkout.select_region")), this.regions.map((region) => (index.h("option", { key: region.id, value: region.id, selected: this.selectedRegion?.id === region.id }, region.name))))));
    }
    renderCityDistrictSelects() {
        // Guest form: region + district only for SA; other countries get country + city only
        const isSA = isSaudiArabia(this.selectedCountry?.code ?? '');
        const showDistrict = isSA;
        const showDistrictInput = showDistrict &&
            this.selectedCity &&
            !this.loadingDistricts &&
            this.districts.length === 0;
        // For SA: city disabled until region selected; for non-SA: city disabled until country selected
        const cityDisabled = !this.selectedCountry ||
            (isSA && !this.selectedRegion) ||
            this.loadingCities ||
            this.cities.length === 0;
        const districtDisabled = !this.selectedCountry ||
            (isSA && !this.selectedRegion) ||
            !this.selectedCity ||
            this.loadingDistricts;
        return (index.h("div", { class: "s-bullet-delivery-field-row" }, index.h("div", { class: "s-bullet-delivery-field" }, this.renderLoadingLabel(Salla.lang.get("blocks.buy_as_gift.receiver_city"), "bullet-delivery-city", this.loadingCities), index.h("select", { id: "bullet-delivery-city", class: "form-input", disabled: cityDisabled, onChange: (e) => {
                const select = e.target;
                const cityId = Number.parseInt(select.value);
                const city = this.cities.find((c) => c.id === cityId);
                this.selectedCity = city || null;
                this.selectedDistrict = null;
                this.districtName = "";
                if (city) {
                    this.updateNewAddressForm({ city: city, district: undefined });
                    if (showDistrict) {
                        this.loadDistricts(city.id);
                    }
                }
            } }, index.h("option", { value: "", disabled: true, selected: !this.selectedCity }, Salla.lang.get("pages.checkout.select_city")), this.cities.map((city) => (index.h("option", { key: city.id, value: city.id, selected: this.selectedCity?.id === city.id }, city.name))))), showDistrict && (index.h("div", { class: "s-bullet-delivery-field" }, this.renderLoadingLabel(Salla.lang.get("pages.checkout.district_field"), "bullet-delivery-district", this.loadingDistricts && !showDistrictInput), showDistrictInput ? (index.h("input", { id: "bullet-delivery-district", type: "text", class: "form-input", placeholder: Salla.lang.get("pages.checkout.select_district"), value: this.districtName, disabled: districtDisabled, onInput: (e) => {
                const input = e.target;
                this.districtName = input.value;
                this.updateNewAddressForm({ district: undefined });
            } })) : (index.h("select", { id: "bullet-delivery-district", class: "form-input", disabled: districtDisabled || this.districts.length === 0, onChange: (e) => {
                const select = e.target;
                const districtId = Number.parseInt(select.value);
                const district = this.districts.find((d) => d.id === districtId);
                this.selectedDistrict = district || null;
                this.districtName = "";
                if (district) {
                    this.updateNewAddressForm({ district });
                }
            } }, index.h("option", { value: "", disabled: true, selected: !this.selectedDistrict }, Salla.lang.get("pages.checkout.select_district")), this.districts.map((district) => {
            const langCode = Salla.config.get("user.language_code");
            const label = langCode === "ar"
                ? district.name
                : (district.name_en ?? district.name);
            return (index.h("option", { key: district.id, value: district.id, selected: this.selectedDistrict?.id === district.id }, label));
        })))))));
    }
    renderSavedAddressesEmptyState() {
        return (index.h("div", { class: "s-bullet-delivery-saved-addresses-empty" }, index.h("div", { class: "s-bullet-delivery-saved-addresses-empty-icon", "aria-hidden": "true" }, index.h("svg", { width: "28", height: "28", viewBox: "0 0 28 29", fill: "none", xmlns: "http://www.w3.org/2000/svg" }, index.h("title", null, "No Saved Adresses"), index.h("path", { d: "M20.16 0C21.4613 0 22.6773 0.341333 23.808 1.024C24.9387 1.70667 25.8347 2.61333 26.496 3.744C27.1787 4.91733 27.52 6.18667 27.52 7.552C27.52 9.45067 26.8267 11.232 25.44 12.896C24.4587 14.0907 23.0827 15.2427 21.312 16.352H21.28C20.9387 16.5653 20.5653 16.672 20.16 16.672C19.7547 16.672 19.3813 16.5653 19.04 16.352H19.008C17.2373 15.2213 15.8613 14.0693 14.88 12.896C13.4933 11.232 12.8 9.45067 12.8 7.552C12.8 6.18667 13.1413 4.91733 13.824 3.744C14.4853 2.61333 15.3813 1.70667 16.512 1.024C17.6427 0.341333 18.8587 0 20.16 0ZM11.488 1.984C9.67467 2.368 8.04267 3.14667 6.592 4.32C5.14133 5.49333 4.01067 6.92267 3.2 8.608C2.34667 10.3573 1.92 12.224 1.92 14.208C1.92 14.72 1.952 15.2427 2.016 15.776L2.08 15.744C2.57067 15.488 2.94933 15.3067 3.216 15.2C3.48267 15.0933 3.73867 15.0933 3.984 15.2C4.22933 15.3067 4.4 15.488 4.496 15.744C4.592 16 4.58667 16.256 4.48 16.512C4.37333 16.768 4.19733 16.9493 3.952 17.056C3.70667 17.1627 3.36 17.3227 2.912 17.536L2.656 17.664C2.57067 17.7067 2.48533 17.7387 2.4 17.76C2.89067 19.4667 3.70133 20.9973 4.832 22.352C5.96267 23.7067 7.296 24.7573 8.832 25.504C10.432 26.272 12.1173 26.656 13.888 26.656C15.4667 26.656 16.9813 26.3467 18.432 25.728C19.8187 25.1307 21.0667 24.288 22.176 23.2C21.92 23.0507 21.7653 22.8373 21.712 22.56C21.6587 22.2827 21.7013 22.0267 21.84 21.792C21.9787 21.5573 22.176 21.408 22.432 21.344C22.8373 21.2587 23.2 21.12 23.52 20.928C23.6907 20.8427 23.8613 20.8 24.032 20.8C24.8 19.5413 25.3333 18.176 25.632 16.704C25.6747 16.4267 25.808 16.2133 26.032 16.064C26.256 15.9147 26.496 15.8667 26.752 15.92C27.008 15.9733 27.2107 16.1173 27.36 16.352C27.5093 16.5867 27.552 16.832 27.488 17.088C27.0613 19.2853 26.1973 21.264 24.896 23.024C23.5947 24.784 22.0053 26.1547 20.128 27.136C18.1653 28.16 16.0853 28.672 13.888 28.672C11.3707 28.672 9.03467 28.0107 6.88 26.688C4.78933 25.408 3.12533 23.68 1.888 21.504C0.629333 19.264 0 16.832 0 14.208C0 11.904 0.490667 9.73867 1.472 7.712C2.41067 5.74933 3.728 4.09067 5.424 2.736C7.12 1.38133 9.01333 0.48 11.104 0.032C11.36 -0.032 11.6 0.0106667 11.824 0.16C12.048 0.309333 12.1867 0.522667 12.24 0.8C12.2933 1.07733 12.2453 1.328 12.096 1.552C11.9467 1.776 11.744 1.92 11.488 1.984ZM14.72 7.552C14.72 8.96 15.2533 10.304 16.32 11.584C17.1733 12.5867 18.3893 13.6 19.968 14.624C20.032 14.6453 20.096 14.656 20.16 14.656C20.224 14.656 20.288 14.6453 20.352 14.624C21.9307 13.6213 23.1467 12.608 24 11.584C25.0667 10.304 25.6 8.96 25.6 7.552C25.6 6.57067 25.3547 5.65333 24.864 4.8C24.3733 3.94667 23.7067 3.26933 22.864 2.768C22.0213 2.26667 21.12 2.016 20.16 2.016C19.2 2.016 18.2987 2.26667 17.456 2.768C16.6133 3.26933 15.9467 3.94667 15.456 4.8C14.9653 5.65333 14.72 6.57067 14.72 7.552ZM17.28 7.68C17.28 6.848 17.5627 6.13867 18.128 5.552C18.6933 4.96533 19.3707 4.672 20.16 4.672C20.9493 4.672 21.6267 4.96533 22.192 5.552C22.7573 6.13867 23.04 6.848 23.04 7.68C23.04 8.512 22.7573 9.216 22.192 9.792C21.6267 10.368 20.9493 10.656 20.16 10.656C19.3707 10.656 18.6933 10.368 18.128 9.792C17.5627 9.216 17.28 8.512 17.28 7.68ZM6.752 14.688C7.62667 14.7733 8.53333 15.0293 9.472 15.456C9.70667 15.5627 9.87733 15.744 9.984 16C10.0907 16.256 10.0907 16.512 9.984 16.768C9.87733 17.024 9.70133 17.2 9.456 17.296C9.21067 17.392 8.97067 17.3867 8.736 17.28C7.94667 16.9387 7.22133 16.736 6.56 16.672C6.28267 16.6293 6.064 16.5013 5.904 16.288C5.744 16.0747 5.68 15.8293 5.712 15.552C5.744 15.2747 5.86133 15.0507 6.064 14.88C6.26667 14.7093 6.496 14.6453 6.752 14.688ZM17.6 20.544C18.304 20.864 18.976 21.0987 19.616 21.248C19.872 21.312 20.0693 21.4613 20.208 21.696C20.3467 21.9307 20.3893 22.1813 20.336 22.448C20.2827 22.7147 20.144 22.9227 19.92 23.072C19.696 23.2213 19.4453 23.264 19.168 23.2C18.4213 23.0293 17.632 22.752 16.8 22.368C16.5653 22.2613 16.4 22.08 16.304 21.824C16.208 21.568 16.2133 21.312 16.32 21.056C16.4267 20.8 16.6027 20.624 16.848 20.528C17.0933 20.432 17.344 20.4373 17.6 20.544ZM12.832 17.44C13.0027 17.568 13.248 17.7493 13.568 17.984L14.496 18.688C14.7093 18.8587 14.8373 19.0773 14.88 19.344C14.9227 19.6107 14.8693 19.856 14.72 20.08C14.5707 20.304 14.368 20.4427 14.112 20.496C13.856 20.5493 13.6107 20.5013 13.376 20.352L11.744 19.072C11.5307 18.9227 11.4027 18.7093 11.36 18.432C11.3173 18.1547 11.3707 17.904 11.52 17.68C11.6693 17.456 11.872 17.3227 12.128 17.28C12.384 17.2373 12.6187 17.2907 12.832 17.44ZM20.16 8.672C20.416 8.672 20.64 8.576 20.832 8.384C21.024 8.192 21.12 7.95733 21.12 7.68C21.12 7.40267 21.024 7.16267 20.832 6.96C20.64 6.75733 20.416 6.656 20.16 6.656C19.904 6.656 19.68 6.75733 19.488 6.96C19.296 7.16267 19.2 7.40267 19.2 7.68C19.2 7.95733 19.296 8.192 19.488 8.384C19.68 8.576 19.904 8.672 20.16 8.672Z", fill: "#666666" }))), index.h("p", { class: "s-bullet-delivery-saved-addresses-empty-title" }, Salla.lang.get("pages.checkout.no_saved_addresses")), index.h("p", { class: "s-bullet-delivery-saved-addresses-empty-desc" }, Salla.lang.get("pages.checkout.add_address_simplify_shipping")), index.h("salla-button", { onClick: () => this.handleAddNewAddress(), onKeyUp: () => this.handleAddNewAddress() }, Salla.lang.get("pages.checkout.add_new_address"))));
    }
    /**
     * Save session address (guest intent) to user profile. Returns the newly saved address or null.
     */
    async saveSessionAddressToProfile() {
        const intent = this.getStoredIntent();
        const countryId = getIntentCountryId(intent);
        const cityId = getIntentCityId(intent);
        const districtId = getIntentDistrictId(intent);
        if (!intent || intent.type !== "address" || !countryId || !cityId)
            return null;
        const countryCode = getIntentCountryCode(intent);
        if (isSaudiArabia(countryCode) && !districtId)
            return null;
        const isSA = isSaudiArabia(countryCode);
        const districtName = intent.address_details?.district?.name ?? "";
        const payload = {
            country_id: countryId,
            region_id: getIntentRegionId(intent),
            city_id: cityId,
            district_id: districtId,
            description: districtName,
        };
        if (isSA) {
            payload.local = districtName;
        }
        try {
            const { success, address } = await bulletDeliveryAPI.saveAddressLocation(payload);
            if (!success)
                return null;
            const savedAddress = address ?? null;
            if (savedAddress) {
                this.addressCreated.emit({ address: savedAddress });
            }
            return savedAddress;
        }
        catch (error) {
            console.error("SallaBulletDelivery: Error saving session address", error);
            return null;
        }
    }
    renderSavedAddresses() {
        if (!this.isLoggedIn)
            return null;
        if (this.loadingSavedAddresses) {
            return (index.h("div", null, index.h("salla-loading", { size: "md" })));
        }
        const intent = this.getStoredIntent();
        const hasSessionAddress = hasSessionAddressIntent(intent);
        const sessionAddressTitle = hasSessionAddress && intent?.address_details
            ? [
                intent.address_details.district?.name,
                intent.address_details.city.name,
                intent.address_details.country.name,
            ]
                .filter(Boolean)
                .join("، ") ||
                [
                    intent.address_details.country.id,
                    intent.address_details.city.id,
                ].join(", ")
            : "";
        if (this.savedAddresses.length === 0 && !hasSessionAddress) {
            return this.renderSavedAddressesEmptyState();
        }
        return (index.h("div", { class: "s-bullet-delivery-addresses-list s-scrollbar" }, hasSessionAddress &&
            this.renderAddressRow({
                key: "session-address",
                line1: sessionAddressTitle,
                line2: "",
                selected: this.selectedSessionAddress,
                onClick: () => this.handleSessionAddressSelect(),
            }), this.savedAddresses.map((address) => {
            const builtLine1 = [address.building_number, address.street]
                .filter(Boolean)
                .join("، ");
            const line1 = builtLine1 || address.formatted?.address_one || address.short_address || "";
            const builtLine2 = [
                address.district?.name,
                address.city?.name,
                address.country?.name,
            ]
                .filter(Boolean)
                .join("، ");
            const line2 = (builtLine2 || address.formatted?.address_two) ?? "";
            const inCoverage = address.is_in_coverage !== false;
            return this.renderAddressRow({
                key: String(address.id),
                line1,
                line2,
                selected: this.selectedSavedAddress?.id === address.id,
                disabled: !inCoverage,
                onClick: () => inCoverage && this.handleSavedAddressSelect(address),
            });
        })));
    }
    renderAddressRow(options) {
        const { key, line1, line2, selected, disabled = false, onClick } = options;
        return (index.h("button", { key: key, class: {
                "s-bullet-delivery-address-item": true,
                "s-bullet-delivery-address-item--selected": selected,
                "s-bullet-delivery-address-item--disabled": disabled,
            }, type: "button", onClick: disabled ? undefined : onClick, onKeyDown: (e) => {
                if (e.key === "Enter" || e.key === " ") {
                    e.preventDefault();
                    if (!disabled)
                        onClick();
                }
            } }, index.h("input", { type: "radio", name: "saved_address", class: "s-bullet-delivery-radio", checked: selected, disabled: disabled, onClick: (e) => e.stopPropagation() }), index.h("span", { class: "s-bullet-delivery-address-map-icon", innerHTML: MiniMap, "aria-hidden": "true" }), index.h("div", { class: "s-bullet-delivery-address-content" }, index.h("div", { class: "s-bullet-delivery-address-lines" }, index.h("span", { class: "s-bullet-delivery-address-line1" }, line1), index.h("span", { class: "s-bullet-delivery-address-line2" }, line2)))));
    }
    renderGuestDeliveryForm() {
        const isSA = isSaudiArabia(this.selectedCountry?.code ?? '');
        return [
            this.renderCountrySelect(),
            isSA ? this.renderRegionSelect() : null,
            this.renderCityDistrictSelects(),
        ];
    }
    renderDeliveryTab() {
        return (index.h("div", { class: {
                "s-bullet-delivery-content": true,
                "s-bullet-delivery-content--active": this.activeTab === "address",
                "s-hidden": this.activeTab !== "address",
            } }, this.isLoggedIn
            ? this.renderSavedAddresses()
            : this.renderGuestDeliveryForm()));
    }
    renderBranchSearch() {
        return (index.h("div", { class: "s-bullet-delivery-branch-search" }, index.h("div", { class: "s-bullet-delivery-search-row" }, index.h("div", { class: "s-bullet-delivery-search s-bullet-delivery-branch-search-wrap" }, index.h("span", { class: "s-bullet-delivery-search-icon s-bullet-delivery-branch-search-icon", innerHTML: search.Search, "aria-hidden": "true" }), index.h("input", { type: "text", class: "s-bullet-delivery-branch-search-input form-input", placeholder: Salla.lang.get("pages.checkout.search_for_city_or_branch"), value: this.branchSearchQuery, onInput: (e) => this.handleBranchSearch(e), autocomplete: "off" })), index.h("button", { type: "button", class: {
                "s-bullet-delivery-location-btn": true,
                "s-bullet-delivery-location-btn--full": true,
                "s-bullet-delivery-location-btn--loading": this.loadingNearestBranch,
            }, onClick: () => this.handleFindNearestBranch(), disabled: this.loadingNearestBranch }, this.loadingNearestBranch ? (index.h("salla-loading", { size: "sm" })) : (index.h("span", { class: "s-bullet-delivery-location-icon", innerHTML: GPS, "aria-hidden": "true" })), index.h("span", null, Salla.lang.get("pages.checkout.nearest_to_my_location")))), this.locationError && this.activeTab === "branch" && (index.h("span", { class: "s-bullet-delivery-error" }, this.locationError))));
    }
    renderBranchList() {
        if (this.loadingBranches) {
            return (index.h("div", { class: "s-bullet-delivery-branches-wrap" }, index.h("div", { class: "s-bullet-delivery-branches-loading" }, [1, 2, 3, 4].map((v) => (index.h("div", { key: v, class: "s-bullet-delivery-branch-skeleton" }, index.h("salla-skeleton", { class: "s-bullet-delivery-branch-skeleton-radio", type: "circle", height: "16px", width: "16px" }), index.h("div", { class: "s-bullet-delivery-branch-skeleton-icon" }, index.h("salla-skeleton", { height: "34px", width: "34px" })), index.h("div", { class: "s-bullet-delivery-branch-skeleton-content" }, index.h("salla-skeleton", { class: "s-bullet-delivery-branch-skeleton-title", height: "12px", width: "124px" })), index.h("salla-skeleton", { class: "s-bullet-delivery-branch-skeleton-time", height: "12px", width: "88px" }), index.h("div", { class: "s-bullet-delivery-branch-skeleton-map" }, index.h("salla-skeleton", { height: "32px", width: "32px" }))))))));
        }
        if (this.filteredBranches.length === 0) {
            return (index.h("salla-placeholder", { alignment: "center", class: "s-bullet-delivery-placeholder" }, index.h("span", { slot: "title" }, Salla.lang.get("pages.checkout.no_pickup_branches_found"))));
        }
        return (index.h("div", { class: "s-bullet-delivery-branches-wrap" }, index.h("ul", { class: "s-bullet-delivery-branches-list s-scrollbar" }, this.filteredBranches.map((branch) => (index.h("li", { key: branch.id, class: {
                "s-bullet-delivery-branch-item": true,
                "s-bullet-delivery-branch-item--selected": this.selectedBranch?.id === branch.id,
                "s-bullet-delivery-branch-item--disabled": branch.is_open === false,
            }, tabIndex: branch.is_open === false ? -1 : 0, role: "option", "aria-disabled": branch.is_open === false, onClick: () => branch.is_open !== false && this.handleBranchSelect(branch), onKeyDown: (e) => {
                if (e.key === "Enter" || e.key === " ") {
                    e.preventDefault();
                    if (branch.is_open !== false)
                        this.handleBranchSelect(branch);
                }
            } }, index.h("input", { type: "radio", name: "branch", class: "s-bullet-delivery-radio", checked: this.selectedBranch?.id === branch.id, disabled: branch.is_open === false }), index.h("span", { class: "s-bullet-delivery-branch-icon", "aria-hidden": "true" }, index.h("span", { innerHTML: Store })), index.h("div", { class: "s-bullet-delivery-branch-info" }, index.h("span", { class: "s-bullet-delivery-branch-name" }, branch.name), branch.preparation_time ? (index.h("span", { class: "s-bullet-delivery-branch-prep" }, `${Salla.lang.get("pages.checkout.processing_time")}: ${branch.preparation_time}`)) : null), (() => {
            const slot = getBranchFirstSlot(branch.working_hours);
            return slot ? (index.h("span", { class: "s-bullet-delivery-branch-time" }, formatWorkingHoursDisplay(slot.from, slot.to, Salla.config.get("user.language_code") ||
                "ar", Salla.lang.get("pages.checkout.until")))) : null;
        })(), branch.location?.lat != null && branch.location?.lng != null && (index.h("a", { href: `https://www.google.com/maps?q=${branch.location.lat},${branch.location.lng}`, target: "_blank", rel: "noopener noreferrer", class: "s-bullet-delivery-branch-map-link", title: Salla.lang.get("common.elements.open_in_google_maps"), onClick: (e) => e.stopPropagation() }, index.h("span", { class: "s-bullet-delivery-branch-map-link-icon", innerHTML: GetDirections.replace(/mask0_25437_34440/g, `mask0_branch_${branch.id}`), "aria-hidden": "true" }), index.h("span", { class: "sr-only" }, Salla.lang.get("common.elements.open_in_google_maps"))))))))));
    }
    renderPickupTab() {
        return (index.h("div", { class: {
                "s-bullet-delivery-content": true,
                "s-bullet-delivery-content--active": this.activeTab === "branch",
                "s-hidden": this.activeTab !== "branch",
            } }, this.renderCountrySelect(), this.renderBranchSearch(), this.renderBranchList()));
    }
    /** Add-address form: same fields as guest form (country, region, city, district) only. */
    renderAddAddressForm() {
        const title = Salla.lang.get("pages.checkout.add_new_address");
        const isSA = isSaudiArabia(this.selectedCountry?.code ?? '');
        const canSubmit = this.selectedCountry &&
            this.selectedCity &&
            (!isSA || this.selectedRegion) &&
            (!isSA ||
                !!this.selectedDistrict ||
                !!(this.districtName && this.districtName.trim().length > 0));
        const submitLabel = this.savingAddress
            ? Salla.lang.get("pages.checkout.loading")
            : Salla.lang.get("pages.checkout.confirm_address");
        return (index.h("div", { class: "s-bullet-delivery-add-address" }, index.h("div", { class: "s-bullet-delivery-add-address-header" }, index.h("button", { type: "button", class: "s-bullet-delivery-add-address-title", onClick: () => this.handleBackToAddressList(), onKeyDown: (e) => e.key === "Enter" && this.handleBackToAddressList() }, index.h("span", { innerHTML: this.isRTL ? keyboard_arrow_right.ArrowRightIcon : keyboard_arrow_right.ArrowLeftIcon, "aria-hidden": "true" }), index.h("span", { class: "s-bullet-delivery-add-address-title-text" }, title))), index.h("form", { class: "s-bullet-delivery-add-address-form", onSubmit: (e) => this.handleSubmitAddAddress(e) }, this.renderGuestDeliveryForm(), index.h("div", { class: "s-bullet-delivery-form-actions s-bullet-delivery-form-actions--single" }, index.h("salla-button", { type: "submit", loading: this.savingAddress, disabled: !canSubmit || this.savingAddress, class: "s-bullet-delivery-add-address-submit" }, submitLabel)))));
    }
    renderFooter() {
        const showAddAddressInFooter = this.activeTab === "address" && this.isLoggedIn;
        return (index.h("div", { class: "s-bullet-delivery-footer" }, index.h("salla-button", { ref: (btn) => {
                this.confirmBtn = btn;
            }, disabled: this.isConfirmDisabled(), onClick: () => this.handleConfirm(), "loader-position": "center", width: "wide" }, this.activeTab === "branch"
            ? Salla.lang.get("pages.checkout.confirm_receipt_from_the_branch")
            : Salla.lang.get("pages.checkout.confirm_address")), showAddAddressInFooter && (index.h("salla-button", { onClick: () => this.handleAddNewAddress(), width: "wide", shape: "link" }, Salla.lang.get("pages.checkout.add_new_address"))), !this.isLoggedIn && (index.h("salla-button", { shape: "link", onClick: this.handleLogin }, Salla.lang.get("pages.checkout.login_to_use_saved_addresses")))));
    }
    renderMainView() {
        const isAddressEmptyState = this.isLoggedIn &&
            this.activeTab === "address" &&
            this.savedAddresses.length === 0 &&
            !hasSessionAddressIntent(this.getStoredIntent());
        const activeTabView = this.activeTab === "branch" && this.supportsPickup
            ? this.renderPickupTab()
            : this.renderDeliveryTab();
        return (index.h("div", null, index.h("div", { class: "s-bullet-delivery-header" }, index.h("h2", { class: "s-bullet-delivery-title" }, Salla.lang.get("blocks.home.select_the_receiving_method")), index.h("p", { class: "s-bullet-delivery-subtitle" }, Salla.lang.get("pages.checkout.delivery_pickup_availability_hint"))), this.renderTabs(), this.showCartWillBeClearedBanner &&
            this.renderAlert(Salla.lang.get("blocks.scope.empty_cart_warning")), this.allocationOutOfCoverageMessage &&
            this.renderAllocationOutOfCoverageAlert(), activeTabView, !isAddressEmptyState && this.renderFooter()));
    }
    handleAfterAddToCart() {
        if (this.openingType !== "after_add_to_cart")
            return;
        if (this.hasBeenShownThisSession())
            return;
        if (this.modal?.hasAttribute?.("visible"))
            return;
        this.open();
    }
    overrideScopeSwitchUI() {
        const intent = this.getStoredIntent();
        const btn = document.querySelector('[onclick*="scopes"]');
        if (!btn)
            return;
        const trunc = (s, max) => s && s.length > max ? `${s.slice(0, max)}...` : (s ?? "");
        const label = !intent
            ? Salla.lang.get("common.titles.shipping")
            : intent.type === "branch"
                ? `${Salla.lang.get("mobile_app.strings.pickup")} ${Salla.lang.get("common.elements.from")} ${trunc(intent.branch_details?.name, 20)}`
                : `${Salla.lang.get("mobile_app.strings.delivery")} ${Salla.lang.get("common.elements.to")} ${trunc(intent.address_details?.city?.name, 10)}`;
        btn.style.display = intent ? "" : "none";
        btn.onclick = () => this.open();
        const span = btn.querySelector('span:not([class*="sicon"]):not([class*="icon"])');
        if (span) {
            span.textContent = label;
        }
        else {
            const textNode = Array.from(btn.childNodes).find((n) => n.nodeType === Node.TEXT_NODE && n.textContent?.trim());
            if (textNode)
                textNode.textContent = ` ${label}`;
        }
    }
    async componentWillLoad() {
        await Salla.onReady();
        try {
            this.overrideScopeSwitchUI();
        }
        catch (e) {
            console.warn("SallaBulletDelivery: overrideScopeSwitchUI failed", e);
        }
        Salla.event.on("bullet-delivery::open", this.bulletDeliveryOpenHandler);
        this.closeHandler = this.close.bind(this);
        Salla.event.on("bullet-delivery::close", this.closeHandler);
        // Listen for successful login to sync guest session to user profile (AC7)
        this.guestLoginSyncHandler = this.handleGuestToRegisteredSync.bind(this);
        Salla.event.on("login::success", this.guestLoginSyncHandler);
        Salla.event.on("auth::logged-in", this.guestLoginSyncHandler);
        this.isRTL = Salla.config.get("theme.is_rtl");
        // Listen for add-to-cart events for 'after_add_to_cart' opening strategy (after salla is ready)
        this.cartItemAddedHandler = () => this.handleAfterAddToCart();
        if (typeof Salla.cart?.event?.onItemAdded === "function") {
            this.useCartEventApi = true;
            Salla.cart.event.onItemAdded(this.cartItemAddedHandler);
        }
        else {
            Salla.event.on(this.cartItemAddedEvent, this.cartItemAddedHandler);
        }
    }
    render() {
        return (index.h(index.Host, { key: '5d79f6404692b4a68504770a54f46c9e5161e533', class: "s-bullet-delivery" }, index.h("salla-modal", { key: 'f123012f8593e20eeb14ebf59fc8dc5af9bbe12b', ref: (modal) => {
                this.modal = modal;
            }, isClosable: !this.isRequired, class: "s-bullet-delivery-modal", isLoading: this.loading, width: "sm" }, this.loading ? (this.renderSkeleton()) : (index.h("div", { class: "s-bullet-delivery-inner" }, this.viewMode === "main" && this.renderMainView(), this.viewMode === "add-address" && this.renderAddAddressForm())))));
    }
    componentDidLoad() {
        // Mark component as ready
        this.componentReady = true;
        // Handle pending open request (if event was emitted before component was ready)
        if (this.pendingOpen) {
            this.pendingOpen = false;
            this.open();
            return;
        }
        // Strategy-based auto-open logic
        switch (this.openingType) {
            case "first_visit":
                // Open once per session when component mounts (per-session behaviour)
                if (!this.hasBeenShownThisSession()) {
                    this.open();
                }
                break;
            case "on_cart_click":
                // Open automatically when user is on the cart page
                if (Salla.url.is_page("cart")) {
                    Salla.storage.session.get(this.sessionShownKey) !== true &&
                        this.open();
                }
                break;
        }
        // Mount cart view into cart.items.start hook
        const cartViewEl = this.createCartViewElement();
        if (cartViewEl && typeof salla?.hooks?.mount === "function") {
            Salla.hooks.mount("cart:items.start", cartViewEl).catch(() => {
                // Hook may not exist on this page; ignore
            });
        }
    }
    disconnectedCallback() {
        this.componentReady = false;
        Salla.event.off("bullet-delivery::open", this.bulletDeliveryOpenHandler);
        if (this.closeHandler) {
            Salla.event.off("bullet-delivery::close", this.closeHandler);
            this.closeHandler = null;
        }
        if (this.guestLoginSyncHandler) {
            Salla.event.off("login::success", this.guestLoginSyncHandler);
            Salla.event.off("auth::logged-in", this.guestLoginSyncHandler);
            this.guestLoginSyncHandler = null;
        }
        if (this.cartItemAddedHandler) {
            if (this.useCartEventApi && typeof Salla.cart?.event?.offItemAdded === "function") {
                Salla.cart.event.offItemAdded(this.cartItemAddedHandler);
            }
            else {
                Salla.event.off(this.cartItemAddedEvent, this.cartItemAddedHandler);
            }
            this.cartItemAddedHandler = null;
        }
    }
};
/**
 * SA add-address view: 'confirmed' = selected address box + description + confirm, 'full' = all form fields (after Edit).
 */
// Unified form state (guest and add address)
SallaBulletDelivery.INITIAL_ADDRESS_FORM = {
    country_id: 0,
    is_default: false,
    city_id: undefined,
    district_id: undefined,
    street: undefined,
    building_number: undefined,
    additional_number: undefined,
    postal_code: undefined,
    description: undefined,
};
SallaBulletDelivery.style = sallaBulletDeliveryCss;

exports.salla_bullet_delivery = SallaBulletDelivery;
