// @salla.sa/twilight-components/dist/cjs/salla-payments.cjs.entry.js
/*!
 * Crafted with ❤ by Salla
 */
'use strict';

var index = require('./index-Bxxuur88.js');

const sallaPaymentsCss = ":host{display:block}";

const SallaPayments = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        /**
         * Array of the payments/items that are not supported, ex: ["sbc", "made-in-ksa", "cod", "mada", ""].
         */
        this.exclude = [];
    }
    componentWillLoad() {
        return salla.onReady()
            .then(() => {
            if (typeof this.exclude === 'string') {
                try {
                    this.exclude = JSON.parse(this.exclude);
                }
                catch (error) {
                    salla.logger.error('salla-payments:: failed to parse the exclude props!', this.exclude);
                    this.exclude = [];
                }
            }
            this.paymentSlot = this.host.querySelector('[slot="payment"]')?.innerHTML || `<li class="s-payments-list-item"><img width="100%" height="100%" decoding="async" loading="lazy" src="{image}" alt="{payment}" /></li>`;
            this.sbcSlot = this.host.querySelector('[slot="sbc"]')?.innerHTML || `<li class="s-payments-list-item"><a target="_blank" href="{link}"><img width="100%" height="100%" decoding="async" loading="lazy" src="{image}" class="s-payments-sbc-image" alt="SBC" /></a></li>`;
            this.codSlot = this.host.querySelector('[slot="cod"]')?.innerHTML || `<li class="s-payments-list-item"><img width="100%" height="100%" decoding="async" loading="lazy" src="{image}" alt="COD" /></li>`;
            this.payments = salla.config.get('store.settings.payments');
        });
    }
    render() {
        const madeInKSA = !this.exclude.includes('made-in-ksa') && salla.config.get('store.settings.made_in_ksa');
        const sbcId = this.exclude.includes('sbc') ? null : salla.config.get('store.settings.certificate.id');
        return (index.h(index.Host, { key: 'ac201f76fd82b48a5c3121890724519addac63d1', class: "s-payments-list-wrap" }, index.h("ul", { key: 'b82ff166164a2d1186312b3c246b4940a33acb52', class: "s-payments-list" }, madeInKSA && index.h("li", { key: '47fdd99738046820b52617f0101df65fcd1029f7', class: "s-payments-list-item" }, index.h("img", { key: 'cd7aa2fc1265eea115d2f89d1cdf9a1de34b59c1', width: "100%", height: "100%", decoding: "async", loading: "lazy", src: salla.url.cdn(`images/made-in-ksa.svg`, 58, 58), alt: "made in KSA certified" })), this.payments.map((payment) => (payment == "cod" ?
            index.h("div", { id: "cod-slot", innerHTML: this.codSlot
                    .replace(/\{image\}/g, salla.url.cdn(`images/payment/cod_mini.png`, 58, 58)) }) :
            index.h("div", { id: "payment-slot", innerHTML: this.paymentSlot
                    .replace(/\{image\}/g, salla.url.cdn(`images/payment/${payment}_mini.png`, 58, 58))
                    .replace(/\{payment\}/g, payment) }))), sbcId && index.h("div", { key: '628d3efa7636011713975fb3310bfe2a65dc774a', id: "sbc-slot", innerHTML: this.sbcSlot
                .replace(/\{image\}/g, salla.url.cdn(`images/sbc.png`, 58, 58))
                .replace(/\{link\}/g, `https://eauthenticate.saudibusiness.gov.sa/certificate-details/${sbcId}`) }))));
    }
    componentDidRender() {
        //todo:: double check about the following workaround to reduce the dom length, it looks unusal
        this.host.querySelectorAll('#payment-slot').forEach(el => el.replaceWith(el.firstChild));
        this.host.querySelectorAll('#sbc-slot').forEach(el => el.replaceWith(el.firstChild));
        this.host.querySelectorAll('#cod-slot').forEach(el => el.replaceWith(el.firstChild));
        this.host.querySelector('[slot="payment"]')?.remove();
        this.host.querySelector('[slot="sbc"]')?.remove();
        this.host.querySelector('[slot="cod"]')?.remove();
    }
    get host() { return index.getElement(this); }
};
SallaPayments.style = sallaPaymentsCss;

exports.salla_payments = SallaPayments;
