// @salla.sa/twilight-components/dist/cjs/salla-filters-widget.cjs.entry.js
/*!
 * Crafted with ❤ by Salla
 */
'use strict';

var index = require('./index-Bxxuur88.js');
var Helper = require('./Helper-CU4Xuiki.js');
var interfaces = require('./interfaces-DZDQwN6-.js');
require('./anime.es-BqW8JHZi.js');

const sallaFiltersWidgetCss = ":host{display:block}";

const SallaFiltersWidget = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        this.changed = index.createEvent(this, "changed");
        this.initialItemCount = 10; // Show 10 items initially
        this.expandedItemCount = 15; // Show 15 items when expanded
        this.calculatedInitialHeight = 0; // Calculated based on actual content
        this.calculatedExpandedHeight = 0; // Calculated based on actual content
        this.isOpen = true;
        this.isShowMore = false;
        this.showMoreLabel = "عرض المزيد";
        this.showLessLabel = "عرض أقل";
        this.page = salla.config.get('page');
        this.searchQuery = '';
        this.filteredCount = 0; // Track filtered results count
    }
    connectedCallback() {
        // Show "Show More" button if there are more than 10 items
        this.withLoadMore = this.option.key != 'price' && Array.isArray(this.option.values) && this.option.values.length > this.initialItemCount;
        salla.onReady(() => {
            this.page = salla.config.get('page');
        });
        salla.lang.onLoaded(() => {
            this.showMoreLabel = salla.lang.getWithDefault('common.titles.more', this.showMoreLabel);
            this.showLessLabel = salla.lang.getWithDefault('common.elements.show_less', this.showLessLabel);
        });
    }
    componentDidLoad() {
        // Use fixed height per item for predictable calculations
        // Each item is approximately 20px, so we calculate based on item count
        const itemHeight = 20; // Fixed height per item
        const paddingPerItem = 15; // Additional padding per item for margins/spacing
        const allItems = this.widgetValues.querySelectorAll('.s-filters-label');
        if (allItems.length > 0) {
            const itemsToMeasure = Math.min(this.initialItemCount, allItems.length);
            // Calculate initial height: (items × height) + (items × padding)
            this.calculatedInitialHeight = this.initialItemCount * (itemHeight + paddingPerItem);
            // Calculate expanded height based on first 15 items
            if (allItems.length > this.initialItemCount) {
                this.calculatedExpandedHeight = this.expandedItemCount * (itemHeight + paddingPerItem);
            }
            else {
                this.calculatedExpandedHeight = itemsToMeasure * (itemHeight + paddingPerItem);
            }
        }
        this.widgetValues.scrollHeight < this.calculatedInitialHeight && (this.withLoadMore = false);
        (this.withLoadMore && this.widgetValues) && (this.widgetValues.style.maxHeight = `${this.calculatedInitialHeight}px`);
        this.widgetContent.style.height = `${this.widgetContent.scrollHeight}px`;
    }
    /**
   * Asynchronously sets the height of a widget element to its current height, allowing for smooth transitions.
   * This function is often used in scenarios where the widget's content changes dynamically, and animating
   * the height adjustment is desired for a smoother user experience.
   *
   * @param {number} [delay=250] - Optional. The delay (in milliseconds) before updating the widget height.
   *                               Defaults to 250 milliseconds.
   *
   * @returns {Promise<void>} - A Promise that resolves once the widget height is set after the specified delay.
   *
   * @example
   * // Set widget height with the default delay (250 milliseconds)
   * await setWidgetHeight();
   *
   * // Set widget height with a custom delay (e.g., 500 milliseconds)
   * await setWidgetHeight(500);
   */
    async setWidgetHeight(delay = 250) {
        this.widgetContent.removeAttribute('style');
        setTimeout(() => {
            let currentWidgetHeight = this.widgetContent.scrollHeight;
            this.widgetContent.style.height = currentWidgetHeight + 'px';
        }, delay);
    }
    /**
     * Reset selected filter options.
     */
    async reset() {
        if (this.option.type === interfaces.FilterOptionTypes.RANGE) {
            this.priceRange.reset();
        }
        Array.from(this.host.querySelectorAll('input')).forEach(input => input.checked = false);
    }
    /**
     * Action to show more or less filter options.
     */
    async showMore() {
        this.isShowMore = !this.isShowMore;
        this.widgetContent.style.height = 'auto';
        // Reset scroll position to top when collapsing to prevent content cutting
        if (!this.isShowMore) {
            this.widgetValues.scrollTop = 0;
        }
        // Use calculated heights based on actual content
        const maxHeight = this.isShowMore ? this.calculatedExpandedHeight : this.calculatedInitialHeight;
        // Calculate actual item count
        const actualCount = this.searchQuery ? this.filteredCount : (Array.isArray(this.option.values) ? this.option.values.length : 0);
        // When showing more: expand to 15 items height with scrollbar for items beyond 15
        // When showing less: collapse to 10 items height
        this.widgetValues.style.maxHeight = `${maxHeight}px`;
        this.widgetValues.style.overflowY = this.isShowMore && actualCount > this.expandedItemCount ? 'auto' : 'hidden';
        setTimeout(() => {
            this.widgetContent.style.height = `${this.widgetContent.scrollHeight}px`;
        }, 400); // get height after time of collapse animation (duration-300)
    }
    /**
     * Action to toggle widget open or closed (expand/ collapse).
     */
    async toggleWidget() {
        this.isOpen = !this.isOpen;
        Helper.Helper.toggleElementClassIf(this.widgetContent, 's-filters-widget-opened', 's-filters-widget-closed', () => this.isOpen);
    }
    renderFilterOption(option) {
        if (![interfaces.FilterOptionTypes.VALUES, interfaces.FilterOptionTypes.MINIMUM, interfaces.FilterOptionTypes.VARIANTS].includes(option.type)) {
            return '';
        }
        //@ts-ignore
        const filteredValues = option.values.filter((filterOption) => {
            if (!this.searchQuery)
                return true;
            const label = typeof filterOption === 'number'
                ? String(filterOption)
                : (filterOption?.value || filterOption?.key || '');
            return String(label).toLowerCase().includes(this.searchQuery.toLowerCase());
        });
        // Update filtered count for show more button logic
        this.filteredCount = filteredValues.length;
        // Show empty state if no results
        if (filteredValues.length === 0) {
            return (index.h("div", { class: "s-filters-widget-empty-state" }, salla.lang.get('common.elements.no_options') || 'لا توجد نتائج'));
        }
        //@ts-ignore
        return filteredValues.map((filterOption, index$1) => {
            let value = typeof filterOption == 'number' ? filterOption : (filterOption.key || filterOption.value);
            return index.h("label", { class: "s-filters-label", htmlFor: `${option.key}-option-${value}`, key: `${option.key}-option-${value}` }, index.h("input", { id: `${option.key}-option-${value}`, name: option.key, type: "radio", checked: this.isSelectedOption(option, value), class: `s-filters-radio`, onChange: event => this.changed.emit({ event, option, value }) }), this.getOptionLabel(option, filterOption));
        });
    }
    isSelectedOption(option, value) {
        if (option.type === interfaces.FilterOptionTypes.MINIMUM) {
            return this.filtersData[option.key] == value;
        }
        if (option.type === interfaces.FilterOptionTypes.VARIANTS) {
            return this.filtersData[option.type] && this.filtersData[option.type][Object.keys(this.filtersData[option.type])[0]] == value;
        }
        if (option.type === interfaces.FilterOptionTypes.RANGE) {
            return this.filtersData[option.key] && this.filtersData[option.key].min == value.min && this.filtersData[option.key].max == value.max;
        }
        if (option.type === interfaces.FilterOptionTypes.VALUES) {
            return this.filtersData[option.key] && Number(this.filtersData[option.key]) == Number(value);
        }
        return false;
    }
    getOptionLabel(option, filterOption) {
        if (option.key == 'rating') {
            //in amazon has stars & up, should we add it, to avoid those people who will come to say I selected 4 why I see 5 sars products
            return index.h("salla-rating-stars", { size: "small", value: filterOption });
        }
        let label = filterOption.value || 'null';
        //label+=filterOption.count ? ` (${salla.helpers.number(filterOption.count)})` : '';
        return index.h("span", { class: "s-filters-option-name" }, label);
    }
    render() {
        return (index.h(index.Host, { key: 'aaa423868d5ecd62a1d3dbb05a5143f7983e15ab', class: "s-filters-widget-container" }, index.h("h3", { key: '444a72e027f870fcd3f2fb01d8a52d642b1d6fab', class: "s-filters-widget-title", onClick: () => this.toggleWidget() }, index.h("span", { key: '1813e58218aefebecdd0b2d8af9ce89ce5c17b9f' }, this.option.label), index.h("span", { key: '11926f6d1343095b29c93f010b99cfc07152918a', class: `s-filters-widget-plusminus ${this.isOpen ? 's-filters-widget-plusminus-active' : ''}` })), index.h("div", { key: '24f151541c5da697d157f6f110fb640b316ad533', class: "s-filters-widget-content", ref: (el) => this.widgetContent = el }, Array.isArray(this.option.values) && this.option.values.length > 5 && (index.h("div", { key: '12b4adb00888f657f9d67f84b377634190e0cedd', class: "s-filters-widget-search-wrapper", style: { padding: '8px 0', marginBottom: '8px' } }, index.h("input", { key: 'b14ff87ad64e594c3120be3b5e5d102c8d74cd98', type: "text", class: "s-filters-widget-search-input", placeholder: salla.lang.get('blocks.header.search_placeholder') || 'بحث...', value: this.searchQuery, onInput: (e) => this.searchQuery = e.target.value }))), index.h("div", { key: '9a9f76a9ca1a9ae7e474dcf771e0a58936dee7d7', class: "s-filters-widget-values", ref: (el) => this.widgetValues = el }, index.h("slot", { key: '020b614afb9c6ca5aadfccafe748e73e8f2fa77a' }), this.option.type !== interfaces.FilterOptionTypes.RANGE
            ? this.renderFilterOption(this.option)
            : index.h("salla-price-range", { onChanged: (event) => {
                    this.changed.emit(event.detail);
                }, ref: price => this.priceRange = price, filtersData: this.filtersData, option: this.option })), this.withLoadMore && (!this.searchQuery || this.filteredCount > this.initialItemCount) &&
            index.h("a", { key: '4d03eccbc3b357444e27d61c647b09d435ceb1be', class: "s-filters-widget-more", onClick: () => this.showMore() }, !this.isShowMore ? this.showMoreLabel : this.showLessLabel))));
    }
    get host() { return index.getElement(this); }
};
SallaFiltersWidget.style = sallaFiltersWidgetCss;

exports.salla_filters_widget = SallaFiltersWidget;
