// @salla.sa/twilight-components/dist/cjs/salla-booking-field_7.cjs.entry.js
/*!
 * Crafted with ❤ by Salla
 */
'use strict';

var index = require('./index-Bxxuur88.js');
var keyboard_arrow_down = require('./keyboard_arrow_down-DHJ3FFZq.js');
var minus = require('./minus-CCryh1qf.js');
var Helper = require('./Helper-CU4Xuiki.js');
require('./anime.es-BqW8JHZi.js');

function _interopNamespace(e) {
  if (e && e.__esModule) return e;
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () { return e[k]; }
        });
      }
    });
  }
  n.default = e;
  return Object.freeze(n);
}

var BookingTime = `<!-- Generated by IcoMoon.io -->
<svg version="1.1" xmlns="http://www.w3.org/2000/svg" width="32" height="32" viewBox="0 0 32 32">
<title>calendar-time</title>
<path d="M22.667 17.333c-0.737 0-1.333 0.596-1.333 1.333v2.667h-2.667c-0.737 0-1.333 0.596-1.333 1.333s0.596 1.333 1.333 1.333h4c0.737 0 1.333-0.596 1.333-1.333v-4c0-0.737-0.596-1.333-1.333-1.333zM28 2.667h-2.667v-1.333c0-0.736-0.596-1.333-1.333-1.333s-1.333 0.597-1.333 1.333v1.333h-13.333v-1.333c0-0.736-0.596-1.333-1.333-1.333s-1.333 0.597-1.333 1.333v1.333h-2.667c-2.205 0-4 1.795-4 4v21.333c0 2.205 1.795 4 4 4h5.363c0.737 0 1.333-0.596 1.333-1.333s-0.596-1.333-1.333-1.333h-5.363c-0.736 0-1.333-0.597-1.333-1.333v-21.333c0-0.736 0.597-1.333 1.333-1.333h2.667v2.667c0 0.736 0.596 1.333 1.333 1.333s1.333-0.597 1.333-1.333v-2.667h13.333v2.667c0 0.736 0.596 1.333 1.333 1.333s1.333-0.597 1.333-1.333v-2.667h2.667c0.736 0 1.333 0.599 1.333 1.333v2.696c0 0.736 0.596 1.333 1.333 1.333s1.333-0.597 1.333-1.333v-2.696c0-2.205-1.795-4-4-4zM22 12c-5.515 0-10 4.485-10 10s4.485 10 10 10 10-4.485 10-10-4.485-10-10-10zM22 29.333c-4.043 0-7.333-3.291-7.333-7.333s3.291-7.333 7.333-7.333 7.333 3.291 7.333 7.333-3.291 7.333-7.333 7.333z"></path>
</svg>
`;

var Calendar = `<!-- Generated by IcoMoon.io -->
<svg version="1.1" xmlns="http://www.w3.org/2000/svg" width="32" height="32" viewBox="0 0 32 32">
<title>calendar</title>
<path d="M28 2.667h-2.667v-1.333c0-0.736-0.597-1.333-1.333-1.333s-1.333 0.597-1.333 1.333v1.333h-13.333v-1.333c0-0.736-0.597-1.333-1.333-1.333s-1.333 0.597-1.333 1.333v1.333h-2.667c-2.205 0-4 1.795-4 4v21.333c0 2.205 1.795 4 4 4h24c2.205 0 4-1.795 4-4v-21.333c0-2.205-1.795-4-4-4zM29.333 28c0 0.736-0.599 1.333-1.333 1.333h-24c-0.736 0-1.333-0.599-1.333-1.333v-13.333h26.667zM29.333 12h-26.667v-5.333c0-0.736 0.599-1.333 1.333-1.333h2.667v2.667c0 0.736 0.597 1.333 1.333 1.333s1.333-0.597 1.333-1.333v-2.667h13.333v2.667c0 0.736 0.597 1.333 1.333 1.333s1.333-0.597 1.333-1.333v-2.667h2.667c0.736 0 1.333 0.599 1.333 1.333z"></path>
</svg>
`;

var TimeIcon = `<!-- Generated by IcoMoon.io -->
<svg version="1.1" xmlns="http://www.w3.org/2000/svg" width="32" height="32" viewBox="0 0 32 32">
<title>time</title>
<path d="M16 0c-8.823 0-16 7.177-16 16s7.177 16 16 16 16-7.177 16-16-7.177-16-16-16zM16 29.333c-7.352 0-13.333-5.981-13.333-13.333s5.981-13.333 13.333-13.333 13.333 5.981 13.333 13.333-5.981 13.333-13.333 13.333zM16 8c-0.736 0-1.333 0.596-1.333 1.333v6.667h-4c-0.736 0-1.333 0.596-1.333 1.333s0.597 1.333 1.333 1.333h5.333c0.736 0 1.333-0.596 1.333-1.333v-8c0-0.737-0.597-1.333-1.333-1.333z"></path>
</svg>
`;

const sallaBookingFieldCss = ":host{display:block}";

const SallaBookingField = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        this.invalidInput = index.createEvent(this, "invalidInput");
        this.bookNowLabel = salla.lang.get('pages.cart.book_an_appointment', 'حجز موعد');
        this.editLabel = salla.lang.get('pages.cart.edit_an_appointment', 'تعديل الموعد');
        this.bookedLabel = salla.lang.get('pages.cart.booked_successfully', 'تمت اضافة الموعد بنجاح');
        this.selectDate = salla.lang.get('pages.cart.select_appointment_date', 'حدد تاريخ الموعد');
        this.bookingUrl = '';
        this.iframeReady = false;
        this.reservations = [];
        // Load translations
        salla.lang.onLoaded(() => this.setTranslations());
        // Register event listeners
        Salla.event.on('booking::open', (data) => this.handleBookingOpen(data));
    }
    async setTranslations() {
        const setNestedAsync = (lang, key, value) => {
            return new Promise((resolve) => {
                salla.helpers.setNested(salla.lang.messages[lang], key, value);
                resolve(true);
            });
        };
        await setNestedAsync('ar.trans', 'pages.cart.book_an_appointment', 'حجز موعد');
        await setNestedAsync('en.trans', 'pages.cart.book_an_appointment', 'Book an Appointment');
        await setNestedAsync('ar.trans', 'pages.cart.edit_an_appointment', 'تعديل الموعد');
        await setNestedAsync('en.trans', 'pages.cart.edit_an_appointment', 'Edit an Appointment');
        await setNestedAsync('ar.trans', 'pages.cart.booked_successfully', 'تمت اضافة الموعد بنجاح');
        await setNestedAsync('en.trans', 'pages.cart.booked_successfully', 'Booked Successfully');
        await setNestedAsync('ar.trans', 'pages.cart.select_appointment_date', 'حدد تاريخ الموعد');
        await setNestedAsync('en.trans', 'pages.cart.select_appointment_date', 'Select appointment date');
        this.bookNowLabel = salla.lang.get('pages.cart.book_an_appointment');
        this.editLabel = salla.lang.get('pages.cart.edit_an_appointment');
        this.bookedLabel = salla.lang.get('pages.cart.booked_successfully');
        this.selectDate = salla.lang.get('pages.cart.select_appointment_date');
    }
    openBookingModal(event, afterReload = false) {
        if (afterReload && (!event.detail || typeof event.detail !== 'number' || event.detail !== this.productId)) {
            return;
        }
        if (salla.config.isGuest()) {
            this.setAfterReloadEvent('booking::open-after-reload', this.productId);
            salla.event.dispatch('login::open');
            return;
        }
        salla.booking.add(this.productId, false)
            .then((resp) => {
            if (resp.data.redirect.to !== 'booking') {
                throw new Error('Unexpected redirect!');
            }
            salla.event.dispatch('booking::open', { url: resp.data.redirect.url, id: this.productId });
        })
            .catch((error) => {
            salla.error(salla.lang.get('common.errors.error_occurred'));
            salla.logger.error(error.response || error);
        });
    }
    handleBookingOpen(data) {
        if (data.id !== this.productId)
            return;
        this.bookingUrl = salla.url.addParamToUrl('product_id', data.id, data.url);
        this.iframeReady = true;
        setTimeout(() => {
            this.modal.setTitle(this.selectDate);
            this.modal.open();
        }, 100);
    }
    setAfterReloadEvent(event, payload) {
        salla.storage.set('afterReloadEvent', { event, payload });
    }
    emitAfterReloadEvent() {
        const eventDetails = salla.storage.get('afterReloadEvent');
        if (eventDetails && eventDetails.event) {
            const customEvent = new CustomEvent(eventDetails.event, {
                detail: eventDetails.payload
            });
            window.dispatchEvent(customEvent);
            salla.storage.remove('afterReloadEvent');
        }
    }
    componentWillLoad() {
        if (this.option && this.option.details.length) {
            this.reservations = this.option.details;
        }
    }
    componentDidLoad() {
        window.addEventListener('booking::open-after-reload', (event) => this.openBookingModal(event, true));
        this.emitAfterReloadEvent();
        window.addEventListener('message', this.handleMessageEvent.bind(this));
        this.reservationsInput.addEventListener('invalid', e => this.invalidInput.emit(e));
        this.reservationsInput.addEventListener('input', () => {
            this.reservationsInput.setCustomValidity('');
            this.reservationsInput.reportValidity();
        });
    }
    handleMessageEvent(event) {
        if (event.data.source !== 'booking')
            return;
        const action = event.data.type;
        const value = event.data.message;
        if (localStorage.getItem('debug'))
            console.log(`Received an action:${action}`, event.data);
        if (action === 'error') {
            if (value.fields?.reservation) {
                salla.notify.error(value.fields.reservation[0]);
                return;
            }
            const errorList = Object.values(value.fields || [value.message]).flat().map(error => `<li>${error}</li>`).join('');
            salla.notify.error(`<ul>${errorList}</ul>`);
        }
        if (action === 'success') {
            if (Number(value.productId) !== Number(this.productId))
                return;
            this.reservations = value.data.reservations.map(schedule => {
                if (schedule.time && schedule.time.length > 0) {
                    const timeSlot = schedule.time[0];
                    return {
                        date: schedule.date,
                        day: schedule.day,
                        from_timestamp: timeSlot.from,
                        to_timestamp: timeSlot.to,
                    };
                }
                return null;
            }).filter(item => item !== null);
            salla.notify.success(this.bookedLabel);
            this.modal?.close();
            setTimeout(() => window.location.reload());
        }
        if (action === 'height') {
            this.iframe.height = value?.height + 'px';
        }
    }
    bookingModal() {
        return (index.h("salla-modal", { class: "s-booking-field-modal", ref: modal => (this.modal = modal), width: "md", position: "middle", noPadding: true }, index.h("iframe", { ref: iframe => (this.iframe = iframe), src: this.bookingUrl, frameborder: "0" })));
    }
    renderReservationDate(reservation) {
        return (index.h("span", { class: reservation.from_timestamp ? 's-booking-field-reservations-has-time' : '' }, index.h("i", { class: "s-booking-field-reservations-icon", innerHTML: Calendar }), reservation.date));
    }
    renderReservationTime(reservation) {
        if (!reservation.from_timestamp)
            return '';
        return (index.h("span", { class: "s-booking-field-reservations-time" }, index.h("i", { class: "s-booking-field-reservations-icon", innerHTML: TimeIcon }), index.h("span", null, reservation.from_timestamp, " - ", reservation.to_timestamp)));
    }
    render() {
        return (index.h(index.Host, { key: '04c9e3dc3dc825bd6905221d858c260f7ca34748' }, index.h("div", { key: '07ec38cec1bd1798640dd7ea9592bb3b17d945c6', class: "s-booking-field-main" }, this.option.required || this.reservations.length > 0 ? index.h("div", { class: "s-booking-field-price" }, index.h("span", { innerHTML: salla.money(this.option.price) })) : '', index.h("salla-button", { key: '6b71889cfc0429d15c466a6f194df31549f09786', class: "s-booking-field-book-now", size: "small", loaderPosition: "center", fill: "outline", onClick: event => this.openBookingModal(event, false) }, index.h("span", { key: 'a16f8db6f52c7751ea605efcae04057ca0dc5b61', class: "s-booking-field-book-now-content" }, index.h("span", { key: '95f6086a599bcc620c87ece6c6ae9620dba15581', innerHTML: BookingTime }), this.reservations.length ? this.editLabel : this.bookNowLabel))), this.reservations.length > 0 && (index.h("div", { key: 'b3615100bbfff3d98d0ac99145aaf53449790c94', class: "s-booking-field-reservations" }, this.reservations.map((reservation, index$1) => (index.h("div", { key: index$1, class: "s-booking-field-reservations-item" }, this.renderReservationDate(reservation), this.renderReservationTime(reservation)))))), index.h("input", { key: '126f14f2e520b8696bc7450e209bad30b267878b', class: "s-hidden", name: this.option.name, required: this.option.required, value: JSON.stringify(this.reservations) === '[]' ? '' : JSON.stringify(this.reservations), ref: reservations => this.reservationsInput = reservations }), this.iframeReady && this.bookingModal()));
    }
    get host() { return index.getElement(this); }
};
SallaBookingField.style = sallaBookingFieldCss;

const sallaColorPickerCss = "";

const SallaColorPicker = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        this.colorChanged = index.createEvent(this, "colorChanged");
        this.invalidInput = index.createEvent(this, "invalidInput");
        this.submitted = index.createEvent(this, "submitted");
        this.popupOpened = index.createEvent(this, "popupOpened");
        this.popupClosed = index.createEvent(this, "popupClosed");
        /** Lazy-loaded to avoid bundling vanilla-picker (Babel polyfills) into main/hydrate bundle. */
        this.picker = null;
        this.pickerReady = null;
        /**
         * File input name for the native formData
         */
        this.name = 'color';
        /**
         * Set if the color picker input is required or not
         */
        this.required = false;
        /**
         * How to display the selected color in the text field
         * (the text field still supports input in any format).
         */
        this.format = 'hex';
        /**
         * Whether to have a "Cancel" button which closes the popup.
         */
        this.showCancelButton = false;
        /**
         * Whether to show a text field for color value editing.
         */
        this.showTextField = true;
        /**
         * Whether to enable adjusting the alpha channel.
         */
        this.enableAlpha = false;
        this.updateViewportCache = () => {
            this.cachedViewportWidth = this.getViewportWidthFromBreakpoints();
        };
        this.cachedViewportWidth = 0; // Will be populated in componentDidLoad via requestAnimationFrame
    }
    colorChangeHandler(color) {
        if (this.colorInput)
            this.colorInput.value = color.hex;
        this.colorChanged.emit(color);
    }
    async submittedHandler(color) {
        await this.setColorValue(color.rgbaString, true);
        if (this.canvas)
            this.canvas.style.backgroundColor = color.rgbString;
        if (this.colorInput) {
            this.colorInput.value = color.hex;
            this.colorInput.dispatchEvent(new window.Event('change', { bubbles: true }));
        }
        this.submitted.emit(color);
    }
    popupOpenedHandler(color) {
        // Double rAF: defer setPopInPosition (getBoundingClientRect) until after paint to avoid forced reflow
        requestAnimationFrame(() => {
            requestAnimationFrame(() => this.setPopInPosition());
        });
        this.popupOpened.emit(color);
    }
    popupClosedHandler(color) {
        this.popupClosed.emit(color);
    }
    /** Methods */
    async ensurePicker() {
        if (this.picker)
            return this.picker;
        if (this.pickerReady)
            return await this.pickerReady;
        throw new Error('Color picker not initialized');
    }
    /**
     * Set the picker options.
     *
     * (Usually a new .parent and .color).
     * @param {Object} options
     */
    async setPickerOption(options) {
        const picker = await this.ensurePicker();
        picker.setOptions(options);
    }
    /**
     * Move the popup to a different parent, optionally opening it at the same time.
     *
     * (Usually a new .parent and .color).
     * @param {Options} option
     *
     * Whether to open the popup immediately.
     * @param {boolean} openImmediately
     */
    async movePopUp(options, openImmediately) {
        const picker = await this.ensurePicker();
        picker.movePopup(options, openImmediately);
    }
    /**
     * Set/initialize the picker's color.
     *
     * Color name, RGBA/HSLA/HEX string, or RGBA array.
     * @param {string} color
     *
     * If true, won't trigger onChange.
     * @param {boolean} triggerEvent
     */
    async setColorValue(color, triggerEvent) {
        const picker = await this.ensurePicker();
        picker.setColor(color, triggerEvent);
    }
    /**
     * Show/open the picker.
     */
    async openPicker() {
        const picker = await this.ensurePicker();
        picker.show();
    }
    /**
     * Close/Hide the picker.
     */
    async closePicker() {
        const picker = await this.ensurePicker();
        picker.hide();
    }
    /**
     * Release all resources used by this picker instance.
     */
    async destroyPicker() {
        const picker = await this.ensurePicker();
        picker.destroy();
        this.picker = null;
        this.pickerReady = null;
    }
    disconnectedCallback() {
        window.removeEventListener('resize', this.updateViewportCache);
    }
    componentWillLoad() {
        salla.onReady(() => {
            this.color = this.color ? this.color : salla.config.get('theme.color.primary', '#5dd5c4');
        });
    }
    /**
     * Returns viewport width from matchMedia breakpoints (no layout read).
     * Must match updateViewportCache breakpoints for consistency.
     */
    getViewportWidthFromBreakpoints() {
        if (typeof window === 'undefined')
            return 1024;
        if (window.matchMedia('(min-width: 1200px)').matches)
            return 1200;
        if (window.matchMedia('(min-width: 992px)').matches)
            return 992;
        if (window.matchMedia('(min-width: 768px)').matches)
            return 768;
        return 375;
    }
    setPopInPosition() {
        // Use cached viewport width – reading innerWidth at popup open causes forced reflow.
        // Fallback: matchMedia-based value when cache is 0 (e.g. SSR before hydration). Same breakpoints as updateViewportCache.
        const viewportWidth = this.cachedViewportWidth || this.getViewportWidthFromBreakpoints();
        const popup = this.host.querySelector('.picker_wrapper');
        const widgetEl = this.host.querySelector('.s-color-picker-widget');
        if (!popup || !widgetEl)
            return;
        const widgetPosition = widgetEl.getBoundingClientRect();
        const widgetToWindowEq = viewportWidth / 2 - widgetPosition.width / 2;
        const widgetInLeft = widgetToWindowEq > widgetPosition.x;
        const widgetInRight = widgetToWindowEq < widgetPosition.x;
        const widgetInCenter = widgetToWindowEq === widgetPosition.x;
        const isMobile = !window.matchMedia('(min-width: 768px)').matches;
        if (isMobile && widgetInLeft) {
            popup.style.left = '0';
        }
        if (isMobile && widgetInRight) {
            popup.style.left = 'auto';
        }
        if (!isMobile || (isMobile && ((!widgetInRight && !widgetInLeft) || widgetInCenter))) {
            popup.style.left = `-95px`;
        }
    }
    initColorPicker() {
        this.pickerReady = Promise.resolve().then(function () { return require('./vanilla-picker-DDwsUdkB.js'); }).then(function (n) { return n.vanillaPicker; }).then(mod => {
            const Picker = mod.default;
            const picker = new Picker({
                parent: this.host,
                color: this.color,
                popup: 'bottom',
                alpha: this.enableAlpha,
                editor: this.showTextField,
                editorFormat: this.format,
                cancelButton: this.showCancelButton,
                onChange: (color) => this.colorChangeHandler(color),
                onDone: (color) => this.submittedHandler(color),
                onOpen: (color) => this.popupOpenedHandler(color),
                onClose: (color) => this.popupClosedHandler(color),
            });
            this.picker = picker;
            return this.picker;
        });
    }
    render() {
        return (index.h(index.Host, { key: '1da3853e7ba946b680f99ec07040035086cdf383', class: "s-color-picker-main" }, index.h("slot", { key: '16812a7f96d476deab7b0fe4b59f29595cc9f302', name: "widget" }, index.h("div", { key: 'a2820835ad0ed81022871d76fada51885f2cd144', class: "s-color-picker-widget" }, index.h("div", { key: 'e65430e20d893fc372449c933a619cceec98b7db', class: "s-color-picker-widget-canvas", ref: dv => (this.canvas = dv) }), index.h("span", { key: 'f5cfd889738260dcdb1de6da426c673b4685cffe', innerHTML: keyboard_arrow_down.ArrowDown }))), index.h("input", { key: '073754e72beaae864bce94a4de29eee7d0c189ba', class: "s-hidden", name: this.name, required: this.required, value: this.color, ref: color => (this.colorInput = color) })));
    }
    componentDidLoad() {
        if (this.canvas)
            this.canvas.style.backgroundColor = this.color;
        this.initColorPicker();
        // Populate viewport cache in next frame (avoids forced reflow during initial render)
        requestAnimationFrame(() => this.updateViewportCache());
        window.addEventListener('resize', this.updateViewportCache);
        if (this.colorInput) {
            this.colorInput.addEventListener('invalid', e => {
                this.invalidInput.emit(e);
            });
            this.colorInput.addEventListener('input', () => {
                this.colorInput.setCustomValidity('');
                this.colorInput.reportValidity();
            });
        }
    }
    get host() { return index.getElement(this); }
};
SallaColorPicker.style = sallaColorPickerCss;

const SallaConditionalFields = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
    }
    hideAllOptions(optionId) {
        this.host.querySelectorAll(`[data-show-when^="options[${optionId}"]`).forEach((field) => {
            field.classList.add('hidden');
            this.hideAllOptions(field.dataset.optionId);
            this.disableInputs(field);
        });
    }
    disableInputs(field) {
        field.querySelectorAll('[name]').forEach((input) => {
            input.setAttribute('disabled', '');
            input.removeAttribute('required');
            if (input?.tagName?.toLowerCase() === 'select') {
                input.value = '';
            }
            if (['checkbox'].includes(input.getAttribute('type')) && input.hasOwnProperty('checked')) {
                // @ts-ignore
                input.checked = false;
            }
        });
    }
    changeHandler(event) {
        salla.event.emit('salla-onditional-fields::change', event);
        salla.log('Received the change/input event: ', event);
        if (!event.target ||
            (!['SELECT', 'INPUT', 'TEXTAREA'].includes(event.target.tagName) &&
                !['checkbox', 'radio', 'text'].includes(event.target.getAttribute('type')))) {
            salla.log('Ignore the event because is not a supported input: ' + (event?.target?.tagName || 'N/A'));
            return;
        }
        // For text inputs, debounce the handling to improve performance on mobile
        const isTextInput = ['INPUT', 'TEXTAREA'].includes(event.target.tagName) &&
            (!event.target.getAttribute('type') || event.target.getAttribute('type') === 'text');
        if (isTextInput && event.type === 'input') {
            clearTimeout(this.debounceTimeout);
            this.debounceTimeout = setTimeout(() => {
                this.processConditionalFields(event);
            }, 300); // 300ms debounce for text inputs
            return;
        }
        // Process immediately for change events and non-text inputs
        this.processConditionalFields(event);
    }
    processConditionalFields(event) {
        let optionId = event.target.name.replace('[]', '');
        let isMultiple = event.target.getAttribute('type') === 'checkbox';
        let isRadio = event.target.getAttribute('type') === 'radio';
        let isTextInput = ['INPUT', 'TEXTAREA'].includes(event.target.tagName) &&
            (!event.target.getAttribute('type') || event.target.getAttribute('type') === 'text');
        salla.log('Trying to find all elements with condition:', `[data-show-when^="${optionId}"]`);
        this.host.querySelectorAll(`[data-show-when^="${optionId}"]`)
            .forEach((field) => {
            let isEqual = !field?.dataset.showWhen.includes('!=');
            let value = field?.dataset.showWhen.replace(/(.*)(=|!=)(.*)/gm, '$3').trim();
            let isSelected;
            if (isMultiple) {
                let selectedValues = Array.from(this.host.querySelectorAll(`input[name="${event.target.name}"]:checked`), e => e?.value);
                isSelected = selectedValues.includes(value.toString());
            }
            else if (isRadio) {
                // Handle radio inputs.
                isSelected = event.target.checked && event.target.value === value;
            }
            else if (isTextInput) {
                // Handle text inputs and textareas - check if value matches or is not empty for boolean conditions
                isSelected = value === event.target.value || (value.toLowerCase() === 'true' && event.target.value.trim() !== '');
            }
            else {
                isSelected = value === event.target.value;
            }
            salla.log('The input is ', isMultiple ? 'Multiple' : isRadio ? 'Radio' : isTextInput ? 'Text' : 'Single', ' value:', isSelected);
            let showTheInput = (isEqual && isSelected) || (!isEqual && !isSelected);
            if (showTheInput) {
                field.classList.remove('hidden');
                field.querySelectorAll('[name]').forEach((input) => {
                    input.removeAttribute('disabled');
                    const closestProductOption = input.closest('.s-product-options-option');
                    if (closestProductOption.dataset.optionRequired === 'true') {
                        input.setAttribute('required', '');
                    }
                    if (input.getAttribute('type') === 'checkbox') {
                        const checkboxes = Array.from(document.querySelectorAll(`input[type="checkbox"][name="${input.getAttribute('name')}"]`));
                        const isAnyChecked = checkboxes.some((checkbox) => checkbox.checked);
                        if (isAnyChecked) {
                            checkboxes.forEach((checkbox) => {
                                checkbox.removeAttribute('required');
                            });
                        }
                    }
                });
            }
            else {
                this.hideAllOptions(field.dataset.optionId);
                field.classList.add('hidden');
                this.disableInputs(field);
            }
        });
    }
    componentDidRender() {
        this.host.querySelectorAll(`[data-show-when]`).forEach((field) => {
            // @ts-ignore
            let optionName = field?.dataset?.showWhen.replace(/(.*)(=|!=)(.*)/gm, '$1').trim();
            if (!optionName) {
                return;
            }
            this.changeHandler({
                target: this.host.querySelector('[name^="' + optionName + '"]')
            });
        });
    }
    render() {
        return (index.h(index.Host, { key: 'f10107a1f25a518a71fe009caf40baabff9dbdc7' }, index.h("slot", { key: '7fef4fea8387956cc13583759dc3920c4881f389' })));
    }
    get host() { return index.getElement(this); }
};

const sallaDatetimePickerCss = ".flatpickr-calendar{background:transparent;opacity:0;display:none;text-align:center;visibility:hidden;padding:0;-webkit-animation:none;animation:none;direction:ltr;border:0;font-size:14px;line-height:24px;border-radius:5px;position:absolute;width:307.875px;-webkit-box-sizing:border-box;box-sizing:border-box;-ms-touch-action:manipulation;touch-action:manipulation;background:#fff;-webkit-box-shadow:1px 0 0 #e6e6e6,-1px 0 0 #e6e6e6,0 1px 0 #e6e6e6,0 -1px 0 #e6e6e6,0 3px 13px rgba(0,0,0,0.08);box-shadow:1px 0 0 #e6e6e6,-1px 0 0 #e6e6e6,0 1px 0 #e6e6e6,0 -1px 0 #e6e6e6,0 3px 13px rgba(0,0,0,0.08)}.flatpickr-calendar.open,.flatpickr-calendar.inline{opacity:1;max-height:640px;visibility:visible}.flatpickr-calendar.open{display:inline-block;z-index:99999}.flatpickr-calendar.animate.open{-webkit-animation:fpFadeInDown 300ms cubic-bezier(.23,1,.32,1);animation:fpFadeInDown 300ms cubic-bezier(.23,1,.32,1)}.flatpickr-calendar.inline{display:block;position:relative;top:2px}.flatpickr-calendar.static{position:absolute;top:calc(100% + 2px)}.flatpickr-calendar.static.open{z-index:999;display:block}.flatpickr-calendar.multiMonth .flatpickr-days .dayContainer:nth-child(n+1) .flatpickr-day.inRange:nth-child(7n+7){-webkit-box-shadow:none !important;box-shadow:none !important}.flatpickr-calendar.multiMonth .flatpickr-days .dayContainer:nth-child(n+2) .flatpickr-day.inRange:nth-child(7n+1){-webkit-box-shadow:-2px 0 0 #e6e6e6,5px 0 0 #e6e6e6;box-shadow:-2px 0 0 #e6e6e6,5px 0 0 #e6e6e6}.flatpickr-calendar .hasWeeks .dayContainer,.flatpickr-calendar .hasTime .dayContainer{border-bottom:0;border-bottom-right-radius:0;border-bottom-left-radius:0}.flatpickr-calendar .hasWeeks .dayContainer{border-left:0}.flatpickr-calendar.hasTime .flatpickr-time{height:40px;border-top:1px solid #e6e6e6}.flatpickr-calendar.noCalendar.hasTime .flatpickr-time{height:auto}.flatpickr-calendar:before,.flatpickr-calendar:after{position:absolute;display:block;pointer-events:none;border:solid transparent;content:'';height:0;width:0;left:22px}.flatpickr-calendar.rightMost:before,.flatpickr-calendar.arrowRight:before,.flatpickr-calendar.rightMost:after,.flatpickr-calendar.arrowRight:after{left:auto;right:22px}.flatpickr-calendar.arrowCenter:before,.flatpickr-calendar.arrowCenter:after{left:50%;right:50%}.flatpickr-calendar:before{border-width:5px;margin:0 -5px}.flatpickr-calendar:after{border-width:4px;margin:0 -4px}.flatpickr-calendar.arrowTop:before,.flatpickr-calendar.arrowTop:after{bottom:100%}.flatpickr-calendar.arrowTop:before{border-bottom-color:#e6e6e6}.flatpickr-calendar.arrowTop:after{border-bottom-color:#fff}.flatpickr-calendar.arrowBottom:before,.flatpickr-calendar.arrowBottom:after{top:100%}.flatpickr-calendar.arrowBottom:before{border-top-color:#e6e6e6}.flatpickr-calendar.arrowBottom:after{border-top-color:#fff}.flatpickr-calendar:focus{outline:0}.flatpickr-wrapper{position:relative;display:inline-block}.flatpickr-months{display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex}.flatpickr-months .flatpickr-month{background:transparent;color:rgba(0,0,0,0.9);fill:rgba(0,0,0,0.9);height:34px;line-height:1;text-align:center;position:relative;-webkit-user-select:none;-moz-user-select:none;-ms-user-select:none;user-select:none;overflow:hidden;-webkit-box-flex:1;-webkit-flex:1;-ms-flex:1;flex:1}.flatpickr-months .flatpickr-prev-month,.flatpickr-months .flatpickr-next-month{-webkit-user-select:none;-moz-user-select:none;-ms-user-select:none;user-select:none;text-decoration:none;cursor:pointer;position:absolute;top:0;height:34px;padding:10px;z-index:3;color:rgba(0,0,0,0.9);fill:rgba(0,0,0,0.9)}.flatpickr-months .flatpickr-prev-month.flatpickr-disabled,.flatpickr-months .flatpickr-next-month.flatpickr-disabled{display:none}.flatpickr-months .flatpickr-prev-month i,.flatpickr-months .flatpickr-next-month i{position:relative}.flatpickr-months .flatpickr-prev-month.flatpickr-prev-month,.flatpickr-months .flatpickr-next-month.flatpickr-prev-month{left:0}.flatpickr-months .flatpickr-prev-month.flatpickr-next-month,.flatpickr-months .flatpickr-next-month.flatpickr-next-month{right:0}.flatpickr-months .flatpickr-prev-month:hover,.flatpickr-months .flatpickr-next-month:hover{color:#959ea9}.flatpickr-months .flatpickr-prev-month:hover svg,.flatpickr-months .flatpickr-next-month:hover svg{fill:#f64747}.flatpickr-months .flatpickr-prev-month svg,.flatpickr-months .flatpickr-next-month svg{width:14px;height:14px}.flatpickr-months .flatpickr-prev-month svg path,.flatpickr-months .flatpickr-next-month svg path{-webkit-transition:fill .1s;transition:fill .1s;fill:inherit}.numInputWrapper{position:relative;height:auto}.numInputWrapper input,.numInputWrapper span{display:inline-block}.numInputWrapper input{width:100%}.numInputWrapper input::-ms-clear{display:none}.numInputWrapper input::-webkit-outer-spin-button,.numInputWrapper input::-webkit-inner-spin-button{margin:0;-webkit-appearance:none}.numInputWrapper span{position:absolute;right:0;width:14px;padding:0 4px 0 2px;height:50%;line-height:50%;opacity:0;cursor:pointer;border:1px solid rgba(57,57,57,0.15);-webkit-box-sizing:border-box;box-sizing:border-box}.numInputWrapper span:hover{background:rgba(0,0,0,0.1)}.numInputWrapper span:active{background:rgba(0,0,0,0.2)}.numInputWrapper span:after{display:block;content:\"\";position:absolute}.numInputWrapper span.arrowUp{top:0;border-bottom:0}.numInputWrapper span.arrowUp:after{border-left:4px solid transparent;border-right:4px solid transparent;border-bottom:4px solid rgba(57,57,57,0.6);top:26%}.numInputWrapper span.arrowDown{top:50%}.numInputWrapper span.arrowDown:after{border-left:4px solid transparent;border-right:4px solid transparent;border-top:4px solid rgba(57,57,57,0.6);top:40%}.numInputWrapper span svg{width:inherit;height:auto}.numInputWrapper span svg path{fill:rgba(0,0,0,0.5)}.numInputWrapper:hover{background:rgba(0,0,0,0.05)}.numInputWrapper:hover span{opacity:1}.flatpickr-current-month{font-size:135%;line-height:inherit;font-weight:300;color:inherit;position:absolute;width:75%;left:12.5%;padding:7.48px 0 0 0;line-height:1;height:34px;display:inline-block;text-align:center;-webkit-transform:translate3d(0,0,0);transform:translate3d(0,0,0)}.flatpickr-current-month span.cur-month{font-family:inherit;font-weight:700;color:inherit;display:inline-block;margin-left:.5ch;padding:0}.flatpickr-current-month span.cur-month:hover{background:rgba(0,0,0,0.05)}.flatpickr-current-month .numInputWrapper{width:6ch;width:7ch\\0;display:inline-block}.flatpickr-current-month .numInputWrapper span.arrowUp:after{border-bottom-color:rgba(0,0,0,0.9)}.flatpickr-current-month .numInputWrapper span.arrowDown:after{border-top-color:rgba(0,0,0,0.9)}.flatpickr-current-month input.cur-year{background:transparent;-webkit-box-sizing:border-box;box-sizing:border-box;color:inherit;cursor:text;padding:0 0 0 .5ch;margin:0;display:inline-block;font-size:inherit;font-family:inherit;font-weight:300;line-height:inherit;height:auto;border:0;border-radius:0;vertical-align:initial;-webkit-appearance:textfield;-moz-appearance:textfield;appearance:textfield}.flatpickr-current-month input.cur-year:focus{outline:0}.flatpickr-current-month input.cur-year[disabled],.flatpickr-current-month input.cur-year[disabled]:hover{font-size:100%;color:rgba(0,0,0,0.5);background:transparent;pointer-events:none}.flatpickr-current-month .flatpickr-monthDropdown-months{appearance:menulist;background:transparent;border:none;border-radius:0;box-sizing:border-box;color:inherit;cursor:pointer;font-size:inherit;font-family:inherit;font-weight:300;height:auto;line-height:inherit;margin:-1px 0 0 0;outline:none;padding:0 0 0 .5ch;position:relative;vertical-align:initial;-webkit-box-sizing:border-box;-webkit-appearance:menulist;-moz-appearance:menulist;width:auto}.flatpickr-current-month .flatpickr-monthDropdown-months:focus,.flatpickr-current-month .flatpickr-monthDropdown-months:active{outline:none}.flatpickr-current-month .flatpickr-monthDropdown-months:hover{background:rgba(0,0,0,0.05)}.flatpickr-current-month .flatpickr-monthDropdown-months .flatpickr-monthDropdown-month{background-color:transparent;outline:none;padding:0}.flatpickr-weekdays{background:transparent;text-align:center;overflow:hidden;width:100%;display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;-webkit-box-align:center;-webkit-align-items:center;-ms-flex-align:center;align-items:center;height:28px}.flatpickr-weekdays .flatpickr-weekdaycontainer{display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;-webkit-box-flex:1;-webkit-flex:1;-ms-flex:1;flex:1}span.flatpickr-weekday{cursor:default;font-size:90%;background:transparent;color:rgba(0,0,0,0.54);line-height:1;margin:0;text-align:center;display:block;-webkit-box-flex:1;-webkit-flex:1;-ms-flex:1;flex:1;font-weight:bolder}.dayContainer,.flatpickr-weeks{padding:1px 0 0 0}.flatpickr-days{position:relative;overflow:hidden;display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;-webkit-box-align:start;-webkit-align-items:flex-start;-ms-flex-align:start;align-items:flex-start;width:307.875px}.flatpickr-days:focus{outline:0}.dayContainer{padding:0;outline:0;text-align:left;width:307.875px;min-width:307.875px;max-width:307.875px;-webkit-box-sizing:border-box;box-sizing:border-box;display:inline-block;display:-ms-flexbox;display:-webkit-box;display:-webkit-flex;display:flex;-webkit-flex-wrap:wrap;flex-wrap:wrap;-ms-flex-wrap:wrap;-ms-flex-pack:justify;-webkit-justify-content:space-around;justify-content:space-around;-webkit-transform:translate3d(0,0,0);transform:translate3d(0,0,0);opacity:1}.dayContainer+.dayContainer{-webkit-box-shadow:-1px 0 0 #e6e6e6;box-shadow:-1px 0 0 #e6e6e6}.flatpickr-day{background:none;border:1px solid transparent;border-radius:150px;-webkit-box-sizing:border-box;box-sizing:border-box;color:#393939;cursor:pointer;font-weight:400;width:14.2857143%;-webkit-flex-basis:14.2857143%;-ms-flex-preferred-size:14.2857143%;flex-basis:14.2857143%;max-width:39px;height:39px;line-height:39px;margin:0;display:inline-block;position:relative;-webkit-box-pack:center;-webkit-justify-content:center;-ms-flex-pack:center;justify-content:center;text-align:center}.flatpickr-day.inRange,.flatpickr-day.prevMonthDay.inRange,.flatpickr-day.nextMonthDay.inRange,.flatpickr-day.today.inRange,.flatpickr-day.prevMonthDay.today.inRange,.flatpickr-day.nextMonthDay.today.inRange,.flatpickr-day:hover,.flatpickr-day.prevMonthDay:hover,.flatpickr-day.nextMonthDay:hover,.flatpickr-day:focus,.flatpickr-day.prevMonthDay:focus,.flatpickr-day.nextMonthDay:focus{cursor:pointer;outline:0;background:#e6e6e6;border-color:#e6e6e6}.flatpickr-day.today{border-color:#959ea9}.flatpickr-day.today:hover,.flatpickr-day.today:focus{border-color:#959ea9;background:#959ea9;color:#fff}.flatpickr-day.selected,.flatpickr-day.startRange,.flatpickr-day.endRange,.flatpickr-day.selected.inRange,.flatpickr-day.startRange.inRange,.flatpickr-day.endRange.inRange,.flatpickr-day.selected:focus,.flatpickr-day.startRange:focus,.flatpickr-day.endRange:focus,.flatpickr-day.selected:hover,.flatpickr-day.startRange:hover,.flatpickr-day.endRange:hover,.flatpickr-day.selected.prevMonthDay,.flatpickr-day.startRange.prevMonthDay,.flatpickr-day.endRange.prevMonthDay,.flatpickr-day.selected.nextMonthDay,.flatpickr-day.startRange.nextMonthDay,.flatpickr-day.endRange.nextMonthDay{background:#569ff7;-webkit-box-shadow:none;box-shadow:none;color:#fff;border-color:#569ff7}.flatpickr-day.selected.startRange,.flatpickr-day.startRange.startRange,.flatpickr-day.endRange.startRange{border-radius:50px 0 0 50px}.flatpickr-day.selected.endRange,.flatpickr-day.startRange.endRange,.flatpickr-day.endRange.endRange{border-radius:0 50px 50px 0}.flatpickr-day.selected.startRange+.endRange:not(:nth-child(7n+1)),.flatpickr-day.startRange.startRange+.endRange:not(:nth-child(7n+1)),.flatpickr-day.endRange.startRange+.endRange:not(:nth-child(7n+1)){-webkit-box-shadow:-10px 0 0 #569ff7;box-shadow:-10px 0 0 #569ff7}.flatpickr-day.selected.startRange.endRange,.flatpickr-day.startRange.startRange.endRange,.flatpickr-day.endRange.startRange.endRange{border-radius:50px}.flatpickr-day.inRange{border-radius:0;-webkit-box-shadow:-5px 0 0 #e6e6e6,5px 0 0 #e6e6e6;box-shadow:-5px 0 0 #e6e6e6,5px 0 0 #e6e6e6}.flatpickr-day.flatpickr-disabled,.flatpickr-day.flatpickr-disabled:hover,.flatpickr-day.prevMonthDay,.flatpickr-day.nextMonthDay,.flatpickr-day.notAllowed,.flatpickr-day.notAllowed.prevMonthDay,.flatpickr-day.notAllowed.nextMonthDay{color:rgba(57,57,57,0.3);background:transparent;border-color:transparent;cursor:default}.flatpickr-day.flatpickr-disabled,.flatpickr-day.flatpickr-disabled:hover{cursor:not-allowed;color:rgba(57,57,57,0.1)}.flatpickr-day.week.selected{border-radius:0;-webkit-box-shadow:-5px 0 0 #569ff7,5px 0 0 #569ff7;box-shadow:-5px 0 0 #569ff7,5px 0 0 #569ff7}.flatpickr-day.hidden{visibility:hidden}.rangeMode .flatpickr-day{margin-top:1px}.flatpickr-weekwrapper{float:left}.flatpickr-weekwrapper .flatpickr-weeks{padding:0 12px;-webkit-box-shadow:1px 0 0 #e6e6e6;box-shadow:1px 0 0 #e6e6e6}.flatpickr-weekwrapper .flatpickr-weekday{float:none;width:100%;line-height:28px}.flatpickr-weekwrapper span.flatpickr-day,.flatpickr-weekwrapper span.flatpickr-day:hover{display:block;width:100%;max-width:none;color:rgba(57,57,57,0.3);background:transparent;cursor:default;border:none}.flatpickr-innerContainer{display:block;display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;-webkit-box-sizing:border-box;box-sizing:border-box;overflow:hidden}.flatpickr-rContainer{display:inline-block;padding:0;-webkit-box-sizing:border-box;box-sizing:border-box}.flatpickr-time{text-align:center;outline:0;display:block;height:0;line-height:40px;max-height:40px;-webkit-box-sizing:border-box;box-sizing:border-box;overflow:hidden;display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex}.flatpickr-time:after{content:\"\";display:table;clear:both}.flatpickr-time .numInputWrapper{-webkit-box-flex:1;-webkit-flex:1;-ms-flex:1;flex:1;width:40%;height:40px;float:left}.flatpickr-time .numInputWrapper span.arrowUp:after{border-bottom-color:#393939}.flatpickr-time .numInputWrapper span.arrowDown:after{border-top-color:#393939}.flatpickr-time.hasSeconds .numInputWrapper{width:26%}.flatpickr-time.time24hr .numInputWrapper{width:49%}.flatpickr-time input{background:transparent;-webkit-box-shadow:none;box-shadow:none;border:0;border-radius:0;text-align:center;margin:0;padding:0;height:inherit;line-height:inherit;color:#393939;font-size:14px;position:relative;-webkit-box-sizing:border-box;box-sizing:border-box;-webkit-appearance:textfield;-moz-appearance:textfield;appearance:textfield}.flatpickr-time input.flatpickr-hour{font-weight:bold}.flatpickr-time input.flatpickr-minute,.flatpickr-time input.flatpickr-second{font-weight:400}.flatpickr-time input:focus{outline:0;border:0}.flatpickr-time .flatpickr-time-separator,.flatpickr-time .flatpickr-am-pm{height:inherit;float:left;line-height:inherit;color:#393939;font-weight:bold;width:2%;-webkit-user-select:none;-moz-user-select:none;-ms-user-select:none;user-select:none;-webkit-align-self:center;-ms-flex-item-align:center;align-self:center}.flatpickr-time .flatpickr-am-pm{outline:0;width:18%;cursor:pointer;text-align:center;font-weight:400}.flatpickr-time input:hover,.flatpickr-time .flatpickr-am-pm:hover,.flatpickr-time input:focus,.flatpickr-time .flatpickr-am-pm:focus{background:#eee}.flatpickr-input[readonly]{cursor:pointer}@-webkit-keyframes fpFadeInDown{from{opacity:0;-webkit-transform:translate3d(0,-20px,0);transform:translate3d(0,-20px,0)}to{opacity:1;-webkit-transform:translate3d(0,0,0);transform:translate3d(0,0,0)}}@keyframes fpFadeInDown{from{opacity:0;-webkit-transform:translate3d(0,-20px,0);transform:translate3d(0,-20px,0)}to{opacity:1;-webkit-transform:translate3d(0,0,0);transform:translate3d(0,0,0)}}";

const SallaDatetimePicker = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        this.picked = index.createEvent(this, "picked");
        this.invalidInput = index.createEvent(this, "invalidInput");
        /**
         * Two way data binding to retrieve the selected date[time] value
         */
        this.value = null;
        /**
         * Placeholder text to show on the input element
         */
        this.placeholder = salla.lang.get('blocks.buy_as_gift.select_send_date_and_time');
        /**
         * Allows the user to enter a date directly into the input field. By default, direct entry is disabled.
         */
        this.allowInput = true;
        /**
         * Allows the preloading of an invalid date. When disabled, the field will be cleared if the provided date is invalid
         */
        this.allowInvalidPreload = false;
        /**
         * Exactly the same as date format, but for the altInput field.
         */
        this.altFormat = "F j, Y";
        /**
         * Show the user a readable date (as per altFormat), but return something totally different to the server.
         */
        this.altInput = false;
        /**
         * Instead of body, appends the calendar to the specified node instead.
         */
        this.appendTo = undefined;
        /**
         * Defines how the date will be formatted in the aria-label for calendar days,
         * using the same tokens as dateFormat. If you change this, you should choose a
         * value that will make sense if a screen reader reads it out loud.
         */
        this.ariaDateFormat = "F j, Y";
        /**
         * Whether the default time should be auto-filled when the input is empty and gains or loses focus.
         */
        this.autoFillDefaultTime = true;
        /**
         * Whether clicking on the input should open the picker.
         * Set it to false if you only want to open the calendar programmatically with [open()]
         */
        this.clickOpens = true;
        /**
         * Whether calendar should close after date selection or not
         */
        this.closeOnSelect = true;
        /**
         * A string of characters which are used to define how the date will be displayed in the input box.
         * The supported characters are defined in the table below.
         */
        this.dateFormat = "Y-m-d";
        /**
         * Initial value of the hour element, when no date is selected
         */
        this.defaultHour = 12;
        /**
         * Initial value of the minute element, when no date is selected
         */
        this.defaultMinute = 0;
        /**
         * Initial value of the seconds element, when no date is selected
         */
        this.defaultSeconds = 0;
        /**
         * Disables certain dates, preventing them from being selected.
         * See https://chmln.github.io/flatpickr/examples/#disabling-specific-dates
         */
        this.disable = [];
        /**
         * Set this to true to always use the non-native picker on mobile devices.
         * By default, Flatpickr utilizes native datetime widgets unless certain options (e.g. disable) are used.
         */
        this.disableMobile = false;
        /**
         * Disables all dates except these specified. See https://chmln.github.io/flatpickr/examples/#disabling-all-dates-except-select-few
         */
        this.enable = [(_) => true];
        /**
         * Enables seconds selection in the time picker.
         */
        this.enableSeconds = false;
        /**
         * Enables the time picker
         */
        this.enableTime = false;
        /**
         * Adjusts the step for the hour input (incl. scrolling)
         */
        this.hourIncrement = 1;
        /**
         * Displays the calendar inline
         */
        this.inline = false;
        /**
         * The locale, either as a string (e.g. "ar", "en") or as an object.
         * See https://chmln.github.io/flatpickr/localization/
         */
        this.locale = "en";
        /**
         * The maximum date that a user can pick to (inclusive).
         */
        this.maxDate = null;
        /**
         * The minimum date that a user can start picking from (inclusive).
         */
        this.maxTime = null;
        /**
         * The minimum date that a user can start picking from (inclusive).
         */
        this.minDate = null;
        /**
         * The minimum time that a user can start picking from (inclusive).
         */
        this.minTime = null;
        /**
         * Adjusts the step for the minute input (incl. scrolling) Defaults to 5
         */
        this.minuteIncrement = 5;
        /**
         * Date selection mode, defaults to "single"
         */
        this.mode = "single";
        /**
         * How the month should be displayed in the header of the calendar.
         * If showMonths has a value greater than 1, the month is always shown as static.
         */
        this.monthSelectorType = "dropdown";
        /**
         * HTML for the arrow icon, used to switch months.
         */
        this.nextArrow = '<span class="sicon-keyboard_arrow_right"></span>';
        /**
         * Hides the day selection in calendar. Use it along with enableTime to create a time picker.
         */
        this.noCalendar = false;
        /**
         * How the calendar should be positioned with regards to the input. Defaults to "auto"
         */
        this.position = "auto";
        /**
         * HTML for the left arrow icon, used to switch months.
         */
        this.prevArrow = '<span class="sicon-keyboard_arrow_left"></span>';
        /**
         * Whether to display the current month name in shorthand mode, e.g. "Sep" instead "September"
         */
        this.shorthandCurrentMonth = false;
        /**
         * Position the calendar inside the wrapper and next to the input element*.
         */
        this.static = false;
        /**
         * The number of months to be shown at the same time when displaying the calendar.
         */
        this.showMonths = 1;
        /**
         * Displays time picker in 24 hour mode without AM/PM selection when enabled.
         */
        this.time_24hr = false;
        /**
         * Enables display of week numbers in calendar.
         */
        this.weekNumbers = false;
        /**
         * See https://chmln.github.io/flatpickr/examples/#flatpickr-external-elements
         */
        this.wrap = false;
        /**
         * Whether the input is disabled
         */
        this.disabled = false;
    }
    /**
     * Lazy load flatpickr library
     * This reduces initial bundle size by ~35-40KB
     */
    async loadFlatpickr() {
        if (this.flatpickr)
            return;
        try {
            const flatpickrModule = await Promise.resolve().then(function () { return require('./index-D-dRbHuJ.js'); });
            this.flatpickr = flatpickrModule.default;
            // Load locale if not English
            if (this.locale && this.locale !== 'en') {
                try {
                    await (function (t) { return Promise.resolve().then(function () { return /*#__PURE__*/_interopNamespace(require(t)); }); })(`flatpickr/dist/l10n/${this.locale}.js`);
                }
                catch (e) {
                    console.warn(`Flatpickr locale ${this.locale} not found, using default`);
                }
            }
        }
        catch (error) {
            console.error('Failed to load Flatpickr:', error);
            salla.notify?.error?.('Failed to load date picker. Please refresh the page.');
            throw error;
        }
    }
    render() {
        return index.h("div", { key: 'a10e37bc3c3d5e13006fefd3569f2fb7660543bd', class: "s-datetime-picker" }, index.h("input", { key: '1e838ac042a0ca43e7d6a871eee2d6f2840434a0', type: "datetime", name: this.name, value: this.value, required: this.required, disabled: this.disabled, placeholder: this.placeholder, class: "s-datetime-picker-input", ref: (el) => this.dateInput = el }));
    }
    async componentDidLoad() {
        // Load flatpickr before initializing
        await this.loadFlatpickr();
        let options = {
            allowInput: this.allowInput,
            allowInvalidPreload: this.allowInvalidPreload,
            altFormat: this.altFormat,
            altInput: this.altInput,
            altInputClass: this.altInputClass,
            appendTo: this.appendTo,
            ariaDateFormat: this.ariaDateFormat,
            autoFillDefaultTime: this.autoFillDefaultTime,
            clickOpens: this.clickOpens,
            closeOnSelect: this.closeOnSelect,
            conjunction: this.conjunction,
            dateFormat: this.dateFormat,
            defaultDate: !!this.value ? Date.parse(this.value) : this.defaultDate,
            defaultHour: this.defaultHour,
            defaultMinute: this.defaultMinute,
            defaultSeconds: this.defaultSeconds,
            disable: this.disable,
            disableMobile: this.disableMobile,
            enable: this.enable,
            enableSeconds: this.enableSeconds,
            enableTime: this.enableTime,
            formatDate: this.formatDate,
            hourIncrement: this.hourIncrement,
            inline: this.inline,
            locale: this.locale,
            maxDate: this.maxDate,
            maxTime: this.maxTime,
            minDate: this.minDate,
            minTime: this.minTime,
            minuteIncrement: this.minuteIncrement,
            mode: this.mode,
            monthSelectorType: this.monthSelectorType,
            nextArrow: this.nextArrow,
            noCalendar: this.noCalendar,
            parseDate: this.dateParser,
            position: this.position,
            positionElement: this.positionElement,
            prevArrow: this.prevArrow,
            shorthandCurrentMonth: this.shorthandCurrentMonth,
            static: this.static,
            showMonths: this.showMonths,
            time_24hr: this.time_24hr,
            weekNumbers: this.weekNumbers,
            wrap: this.wrap,
            // @ts-ignore
            onChange: (selectedDates, dateStr) => this.picked.emit(this.value = dateStr)
            // onOpen: this.handleOnOpen(selectedDates, dateStr, instance)
            // onClose: this.handleOnClose(selectedDates, dateStr, instance)
        };
        this.flatpickr(this.dateInput, options);
        this.dateInput.addEventListener('invalid', e => {
            this.invalidInput.emit(e);
        });
        this.dateInput.addEventListener('input', () => {
            this.dateInput.setCustomValidity('');
            this.dateInput.reportValidity();
        });
    }
};
SallaDatetimePicker.style = sallaDatetimePickerCss;

const sallaFileUploadCss = "/*!\n * FilePond 4.32.8\n * Licensed under MIT, https://opensource.org/licenses/MIT/\n * Please visit https://pqina.nl/filepond/ for details.\n */.filepond--assistant{position:absolute;overflow:hidden;height:1px;width:1px;padding:0;border:0;clip:rect(1px,1px,1px,1px);-webkit-clip-path:inset(50%);clip-path:inset(50%);white-space:nowrap}.filepond--browser.filepond--browser{position:absolute;margin:0;padding:0;left:1em;top:1.75em;width:calc(100% - 2em);opacity:0;font-size:0}.filepond--data{position:absolute;width:0;height:0;padding:0;margin:0;border:none;visibility:hidden;pointer-events:none;contain:strict}.filepond--drip{position:absolute;top:0;left:0;right:0;bottom:0;overflow:hidden;opacity:.1;pointer-events:none;border-radius:.5em;background:rgba(0,0,0,.01)}.filepond--drip-blob{-webkit-transform-origin:center center;transform-origin:center center;width:8em;height:8em;margin-left:-4em;margin-top:-4em;background:#292625;border-radius:50%}.filepond--drip-blob,.filepond--drop-label{position:absolute;top:0;left:0;will-change:transform,opacity}.filepond--drop-label{right:0;margin:0;color:#4f4f4f;display:flex;justify-content:center;align-items:center;height:0;-webkit-user-select:none;-moz-user-select:none;-ms-user-select:none;user-select:none}.filepond--drop-label.filepond--drop-label label{display:block;margin:0;padding:.5em}.filepond--drop-label label{cursor:default;font-size:.875em;font-weight:400;text-align:center;line-height:1.5}.filepond--label-action{text-decoration:underline;-webkit-text-decoration-skip:ink;text-decoration-skip-ink:auto;-webkit-text-decoration-color:#a7a4a4;text-decoration-color:#a7a4a4;cursor:pointer}.filepond--root[data-disabled] .filepond--drop-label label{opacity:.5}.filepond--file-action-button.filepond--file-action-button{font-size:1em;width:1.625em;height:1.625em;font-family:inherit;line-height:inherit;margin:0;padding:0;border:none;outline:none;will-change:transform,opacity}.filepond--file-action-button.filepond--file-action-button span{position:absolute;overflow:hidden;height:1px;width:1px;padding:0;border:0;clip:rect(1px,1px,1px,1px);-webkit-clip-path:inset(50%);clip-path:inset(50%);white-space:nowrap}.filepond--file-action-button.filepond--file-action-button svg{width:100%;height:100%}.filepond--file-action-button.filepond--file-action-button:after{position:absolute;left:-.75em;right:-.75em;top:-.75em;bottom:-.75em;content:\"\"}.filepond--file-action-button{cursor:auto;color:#fff;border-radius:50%;background-color:rgba(0,0,0,.5);background-image:none;box-shadow:0 0 0 0 hsla(0,0%,100%,0);transition:box-shadow .25s ease-in}.filepond--file-action-button:focus,.filepond--file-action-button:hover{box-shadow:0 0 0 .125em hsla(0,0%,100%,.9)}.filepond--file-action-button[disabled]{color:hsla(0,0%,100%,.5);background-color:rgba(0,0,0,.25)}.filepond--file-action-button[hidden]{display:none}.filepond--action-edit-item.filepond--action-edit-item{width:2em;height:2em;padding:.1875em}.filepond--action-edit-item.filepond--action-edit-item[data-align*=center]{margin-left:-.1875em}.filepond--action-edit-item.filepond--action-edit-item[data-align*=bottom]{margin-bottom:-.1875em}.filepond--action-edit-item-alt{border:none;line-height:inherit;background:transparent;font-family:inherit;color:inherit;outline:none;padding:0;margin:0 0 0 .25em;pointer-events:all;position:absolute}.filepond--action-edit-item-alt svg{width:1.3125em;height:1.3125em}.filepond--action-edit-item-alt span{font-size:0;opacity:0}.filepond--file-info{position:static;display:flex;flex-direction:column;align-items:flex-start;flex:1;margin:0 .5em 0 0;min-width:0;will-change:transform,opacity;pointer-events:none;-webkit-user-select:none;-moz-user-select:none;-ms-user-select:none;user-select:none}.filepond--file-info *{margin:0}.filepond--file-info .filepond--file-info-main{font-size:.75em;line-height:1.2;text-overflow:ellipsis;overflow:hidden;white-space:nowrap;width:100%}.filepond--file-info .filepond--file-info-sub{font-size:.625em;opacity:.5;transition:opacity .25s ease-in-out;white-space:nowrap}.filepond--file-info .filepond--file-info-sub:empty{display:none}.filepond--file-status{position:static;display:flex;flex-direction:column;align-items:flex-end;flex-grow:0;flex-shrink:0;margin:0;min-width:2.25em;text-align:right;will-change:transform,opacity;pointer-events:none;-webkit-user-select:none;-moz-user-select:none;-ms-user-select:none;user-select:none}.filepond--file-status *{margin:0;white-space:nowrap}.filepond--file-status .filepond--file-status-main{font-size:.75em;line-height:1.2}.filepond--file-status .filepond--file-status-sub{font-size:.625em;opacity:.5;transition:opacity .25s ease-in-out}.filepond--file-wrapper.filepond--file-wrapper{border:none;margin:0;padding:0;min-width:0;height:100%}.filepond--file-wrapper.filepond--file-wrapper>legend{position:absolute;overflow:hidden;height:1px;width:1px;padding:0;border:0;clip:rect(1px,1px,1px,1px);-webkit-clip-path:inset(50%);clip-path:inset(50%);white-space:nowrap}.filepond--file{position:static;display:flex;height:100%;align-items:flex-start;padding:.5625em;color:#fff;border-radius:.5em}.filepond--file .filepond--file-status{margin-left:auto;margin-right:2.25em}.filepond--file .filepond--processing-complete-indicator{pointer-events:none;-webkit-user-select:none;-moz-user-select:none;-ms-user-select:none;user-select:none;z-index:3}.filepond--file .filepond--file-action-button,.filepond--file .filepond--processing-complete-indicator,.filepond--file .filepond--progress-indicator{position:absolute}.filepond--file [data-align*=left]{left:.5625em}.filepond--file [data-align*=right]{right:.5625em}.filepond--file [data-align*=center]{left:calc(50% - .8125em)}.filepond--file [data-align*=bottom]{bottom:1.125em}.filepond--file [data-align=center]{top:calc(50% - .8125em)}.filepond--file .filepond--progress-indicator{margin-top:.1875em}.filepond--file .filepond--progress-indicator[data-align*=right]{margin-right:.1875em}.filepond--file .filepond--progress-indicator[data-align*=left]{margin-left:.1875em}[data-filepond-item-state*=error] .filepond--file-info,[data-filepond-item-state*=invalid] .filepond--file-info,[data-filepond-item-state=cancelled] .filepond--file-info{margin-right:2.25em}[data-filepond-item-state~=processing] .filepond--file-status-sub{opacity:0}[data-filepond-item-state~=processing] .filepond--action-abort-item-processing~.filepond--file-status .filepond--file-status-sub{opacity:.5}[data-filepond-item-state=processing-error] .filepond--file-status-sub{opacity:0}[data-filepond-item-state=processing-error] .filepond--action-retry-item-processing~.filepond--file-status .filepond--file-status-sub{opacity:.5}[data-filepond-item-state=processing-complete] .filepond--action-revert-item-processing svg{-webkit-animation:fall .5s linear .125s both;animation:fall .5s linear .125s both}[data-filepond-item-state=processing-complete] .filepond--file-status-sub{opacity:.5}[data-filepond-item-state=processing-complete] .filepond--file-info-sub,[data-filepond-item-state=processing-complete] .filepond--processing-complete-indicator:not([style*=hidden])~.filepond--file-status .filepond--file-status-sub{opacity:0}[data-filepond-item-state=processing-complete] .filepond--action-revert-item-processing~.filepond--file-info .filepond--file-info-sub{opacity:.5}[data-filepond-item-state*=error] .filepond--file-wrapper,[data-filepond-item-state*=error] .filepond--panel,[data-filepond-item-state*=invalid] .filepond--file-wrapper,[data-filepond-item-state*=invalid] .filepond--panel{-webkit-animation:shake .65s linear both;animation:shake .65s linear both}[data-filepond-item-state*=busy] .filepond--progress-indicator svg{-webkit-animation:spin 1s linear infinite;animation:spin 1s linear infinite}@-webkit-keyframes spin{0%{-webkit-transform:rotate(0deg);transform:rotate(0deg)}to{-webkit-transform:rotate(1turn);transform:rotate(1turn)}}@keyframes spin{0%{-webkit-transform:rotate(0deg);transform:rotate(0deg)}to{-webkit-transform:rotate(1turn);transform:rotate(1turn)}}@-webkit-keyframes shake{10%,90%{-webkit-transform:translateX(-.0625em);transform:translateX(-.0625em)}20%,80%{-webkit-transform:translateX(.125em);transform:translateX(.125em)}30%,50%,70%{-webkit-transform:translateX(-.25em);transform:translateX(-.25em)}40%,60%{-webkit-transform:translateX(.25em);transform:translateX(.25em)}}@keyframes shake{10%,90%{-webkit-transform:translateX(-.0625em);transform:translateX(-.0625em)}20%,80%{-webkit-transform:translateX(.125em);transform:translateX(.125em)}30%,50%,70%{-webkit-transform:translateX(-.25em);transform:translateX(-.25em)}40%,60%{-webkit-transform:translateX(.25em);transform:translateX(.25em)}}@-webkit-keyframes fall{0%{opacity:0;-webkit-transform:scale(.5);transform:scale(.5);-webkit-animation-timing-function:ease-out;animation-timing-function:ease-out}70%{opacity:1;-webkit-transform:scale(1.1);transform:scale(1.1);-webkit-animation-timing-function:ease-in-out;animation-timing-function:ease-in-out}to{-webkit-transform:scale(1);transform:scale(1);-webkit-animation-timing-function:ease-out;animation-timing-function:ease-out}}@keyframes fall{0%{opacity:0;-webkit-transform:scale(.5);transform:scale(.5);-webkit-animation-timing-function:ease-out;animation-timing-function:ease-out}70%{opacity:1;-webkit-transform:scale(1.1);transform:scale(1.1);-webkit-animation-timing-function:ease-in-out;animation-timing-function:ease-in-out}to{-webkit-transform:scale(1);transform:scale(1);-webkit-animation-timing-function:ease-out;animation-timing-function:ease-out}}.filepond--hopper[data-hopper-state=drag-over]>*{pointer-events:none}.filepond--hopper[data-hopper-state=drag-over]:after{content:\"\";position:absolute;left:0;top:0;right:0;bottom:0;z-index:100}.filepond--progress-indicator{z-index:103}.filepond--file-action-button{z-index:102}.filepond--file-status{z-index:101}.filepond--file-info{z-index:100}.filepond--item{position:absolute;top:0;left:0;right:0;z-index:1;padding:0;margin:.25em;will-change:transform,opacity;touch-action:auto}.filepond--item>.filepond--panel{z-index:-1}.filepond--item>.filepond--panel .filepond--panel-bottom{box-shadow:0 .0625em .125em -.0625em rgba(0,0,0,.25)}.filepond--item>.filepond--file-wrapper,.filepond--item>.filepond--panel{transition:opacity .15s ease-out}.filepond--item[data-drag-state]{cursor:-webkit-grab;cursor:grab}.filepond--item[data-drag-state]>.filepond--panel{transition:box-shadow .125s ease-in-out;box-shadow:0 0 0 transparent}.filepond--item[data-drag-state=drag]{cursor:-webkit-grabbing;cursor:grabbing}.filepond--item[data-drag-state=drag]>.filepond--panel{box-shadow:0 .125em .3125em rgba(0,0,0,.325)}.filepond--item[data-drag-state]:not([data-drag-state=idle]){z-index:2}.filepond--item-panel{background-color:#64605e}[data-filepond-item-state=processing-complete] .filepond--item-panel{background-color:#369763}[data-filepond-item-state*=error] .filepond--item-panel,[data-filepond-item-state*=invalid] .filepond--item-panel{background-color:#c44e47}.filepond--item-panel{border-radius:.5em;transition:background-color .25s}.filepond--list-scroller{position:absolute;top:0;left:0;right:0;margin:0;will-change:transform}.filepond--list-scroller[data-state=overflow] .filepond--list{bottom:0;right:0}.filepond--list-scroller[data-state=overflow]{overflow-y:scroll;overflow-x:hidden;-webkit-overflow-scrolling:touch;-webkit-mask:linear-gradient(180deg,#000 calc(100% - .5em),transparent);mask:linear-gradient(180deg,#000 calc(100% - .5em),transparent)}.filepond--list-scroller::-webkit-scrollbar{background:transparent}.filepond--list-scroller::-webkit-scrollbar:vertical{width:1em}.filepond--list-scroller::-webkit-scrollbar:horizontal{height:0}.filepond--list-scroller::-webkit-scrollbar-thumb{background-color:rgba(0,0,0,.3);border-radius:99999px;border:.3125em solid transparent;background-clip:content-box}.filepond--list.filepond--list{position:absolute;top:0;margin:0;padding:0;list-style-type:none;will-change:transform}.filepond--list{left:.75em;right:.75em}.filepond--root[data-style-panel-layout~=integrated]{width:100%;height:100%;max-width:none;margin:0}.filepond--root[data-style-panel-layout~=circle] .filepond--panel-root,.filepond--root[data-style-panel-layout~=integrated] .filepond--panel-root{border-radius:0}.filepond--root[data-style-panel-layout~=circle] .filepond--panel-root>*,.filepond--root[data-style-panel-layout~=integrated] .filepond--panel-root>*{display:none}.filepond--root[data-style-panel-layout~=circle] .filepond--drop-label,.filepond--root[data-style-panel-layout~=integrated] .filepond--drop-label{bottom:0;height:auto;display:flex;justify-content:center;align-items:center;z-index:7}.filepond--root[data-style-panel-layout~=circle] .filepond--item-panel,.filepond--root[data-style-panel-layout~=integrated] .filepond--item-panel{display:none}.filepond--root[data-style-panel-layout~=compact] .filepond--list-scroller,.filepond--root[data-style-panel-layout~=integrated] .filepond--list-scroller{overflow:hidden;height:100%;margin-top:0;margin-bottom:0}.filepond--root[data-style-panel-layout~=compact] .filepond--list,.filepond--root[data-style-panel-layout~=integrated] .filepond--list{left:0;right:0;height:100%}.filepond--root[data-style-panel-layout~=compact] .filepond--item,.filepond--root[data-style-panel-layout~=integrated] .filepond--item{margin:0}.filepond--root[data-style-panel-layout~=compact] .filepond--file-wrapper,.filepond--root[data-style-panel-layout~=integrated] .filepond--file-wrapper{height:100%}.filepond--root[data-style-panel-layout~=compact] .filepond--drop-label,.filepond--root[data-style-panel-layout~=integrated] .filepond--drop-label{z-index:7}.filepond--root[data-style-panel-layout~=circle]{border-radius:99999rem;overflow:hidden}.filepond--root[data-style-panel-layout~=circle]>.filepond--panel{border-radius:inherit}.filepond--root[data-style-panel-layout~=circle]>.filepond--panel>*{display:none}.filepond--root[data-style-panel-layout~=circle] .filepond--file-info,.filepond--root[data-style-panel-layout~=circle] .filepond--file-status{display:none}.filepond--root[data-style-panel-layout~=circle] .filepond--action-edit-item{opacity:1!important;visibility:visible!important}@media not all and (min-resolution:0.001dpcm){@supports (-webkit-appearance:none) and (stroke-color:transparent){.filepond--root[data-style-panel-layout~=circle]{will-change:transform}}}.filepond--panel-root{border-radius:.5em;background-color:#f1f0ef}.filepond--panel{position:absolute;left:0;top:0;right:0;margin:0;height:100%!important;pointer-events:none}.filepond-panel:not([data-scalable=false]){height:auto!important}.filepond--panel[data-scalable=false]>div{display:none}.filepond--panel[data-scalable=true]{-webkit-transform-style:preserve-3d;transform-style:preserve-3d;background-color:transparent!important;border:none!important}.filepond--panel-bottom,.filepond--panel-center,.filepond--panel-top{position:absolute;left:0;top:0;right:0;margin:0;padding:0}.filepond--panel-bottom,.filepond--panel-top{height:.5em}.filepond--panel-top{border-bottom-left-radius:0!important;border-bottom-right-radius:0!important;border-bottom:none!important}.filepond--panel-top:after{content:\"\";position:absolute;height:2px;left:0;right:0;bottom:-1px;background-color:inherit}.filepond--panel-bottom,.filepond--panel-center{will-change:transform;-webkit-backface-visibility:hidden;backface-visibility:hidden;-webkit-transform-origin:left top;transform-origin:left top;-webkit-transform:translate3d(0,.5em,0);transform:translate3d(0,.5em,0)}.filepond--panel-bottom{border-top-left-radius:0!important;border-top-right-radius:0!important;border-top:none!important}.filepond--panel-bottom:before{content:\"\";position:absolute;height:2px;left:0;right:0;top:-1px;background-color:inherit}.filepond--panel-center{height:100px!important;border-top:none!important;border-bottom:none!important;border-radius:0!important}.filepond--panel-center:not([style]){visibility:hidden}.filepond--progress-indicator{position:static;width:1.25em;height:1.25em;color:#fff;margin:0;pointer-events:none;will-change:transform,opacity}.filepond--progress-indicator svg{width:100%;height:100%;vertical-align:top;transform-box:fill-box}.filepond--progress-indicator path{fill:none;stroke:currentColor}.filepond--list-scroller{z-index:6}.filepond--drop-label{z-index:5}.filepond--drip{z-index:3}.filepond--root>.filepond--panel{z-index:2}.filepond--browser{z-index:1}.filepond--root{box-sizing:border-box;position:relative;margin-bottom:1em;font-size:1rem;line-height:normal;font-family:-apple-system,BlinkMacSystemFont,Segoe UI,Roboto,Helvetica,Arial,sans-serif,Apple Color Emoji,Segoe UI Emoji,Segoe UI Symbol;font-weight:450;text-align:left;text-rendering:optimizeLegibility;direction:ltr;contain:layout style size}.filepond--root *{box-sizing:inherit;line-height:inherit}.filepond--root :not(text){font-size:inherit}.filepond--root[data-disabled]{pointer-events:none}.filepond--root[data-disabled] .filepond--list-scroller{pointer-events:all}.filepond--root[data-disabled] .filepond--list{pointer-events:none}.filepond--root .filepond--drop-label{min-height:4.75em}.filepond--root .filepond--list-scroller{margin-top:1em;margin-bottom:1em}.filepond--root .filepond--credits{position:absolute;right:0;opacity:.4;line-height:.85;font-size:11px;color:inherit;text-decoration:none;z-index:3;bottom:-14px}.filepond--root .filepond--credits[style]{top:0;bottom:auto;margin-top:14px}/*!\n * FilePondPluginImagePreview 4.6.12\n * Licensed under MIT, https://opensource.org/licenses/MIT/\n * Please visit https://pqina.nl/filepond/ for details.\n */.filepond--image-preview-markup{position:absolute;left:0;top:0}.filepond--image-preview-wrapper{z-index:2}.filepond--image-preview-overlay{display:block;position:absolute;left:0;top:0;width:100%;min-height:5rem;max-height:7rem;margin:0;opacity:0;z-index:2;pointer-events:none;-webkit-user-select:none;-moz-user-select:none;-ms-user-select:none;user-select:none}.filepond--image-preview-overlay svg{width:100%;height:auto;color:inherit;max-height:inherit}.filepond--image-preview-overlay-idle{mix-blend-mode:multiply;color:rgba(40, 40, 40, 0.85)}.filepond--image-preview-overlay-success{mix-blend-mode:normal;color:rgba(54, 151, 99, 1)}.filepond--image-preview-overlay-failure{mix-blend-mode:normal;color:rgba(196, 78, 71, 1)}@supports (-webkit-marquee-repetition: infinite) and\n  ((-o-object-fit: fill) or (object-fit: fill)){.filepond--image-preview-overlay-idle{mix-blend-mode:normal}}.filepond--image-preview-wrapper{-webkit-user-select:none;-moz-user-select:none;-ms-user-select:none;user-select:none;position:absolute;left:0;top:0;right:0;height:100%;margin:0;border-radius:0.45em;overflow:hidden;background:rgba(0, 0, 0, 0.01)}.filepond--image-preview{position:absolute;left:0;top:0;z-index:1;display:flex;align-items:center;height:100%;width:100%;pointer-events:none;background:#222;will-change:transform, opacity}.filepond--image-clip{position:relative;overflow:hidden;margin:0 auto;}.filepond--image-clip[data-transparency-indicator='grid'] img,.filepond--image-clip[data-transparency-indicator='grid'] canvas{background-color:#fff;background-image:url(\"data:image/svg+xml,%3Csvg viewBox='0 0 100 100' xmlns='http://www.w3.org/2000/svg' fill='%23eee'%3E%3Cpath d='M0 0 H50 V50 H0'/%3E%3Cpath d='M50 50 H100 V100 H50'/%3E%3C/svg%3E\");background-size:1.25em 1.25em}.filepond--image-bitmap,.filepond--image-vector{position:absolute;left:0;top:0;will-change:transform}.filepond--root[data-style-panel-layout~='integrated'] .filepond--image-preview-wrapper{border-radius:0}.filepond--root[data-style-panel-layout~='integrated'] .filepond--image-preview{height:100%;display:flex;justify-content:center;align-items:center}.filepond--root[data-style-panel-layout~='circle'] .filepond--image-preview-wrapper{border-radius:99999rem}.filepond--root[data-style-panel-layout~='circle'] .filepond--image-preview-overlay{top:auto;bottom:0;-webkit-transform:scaleY(-1);transform:scaleY(-1)}.filepond--root[data-style-panel-layout~='circle'] .filepond--file .filepond--file-action-button[data-align*='bottom']:not([data-align*='center']){margin-bottom:0.325em}.filepond--root[data-style-panel-layout~='circle'] .filepond--file [data-align*='left']{left:calc(50% - 3em)}.filepond--root[data-style-panel-layout~='circle'] .filepond--file [data-align*='right']{right:calc(50% - 3em)}.filepond--root[data-style-panel-layout~='circle'] .filepond--progress-indicator[data-align*='bottom'][data-align*='left'],.filepond--root[data-style-panel-layout~='circle'] .filepond--progress-indicator[data-align*='bottom'][data-align*='right']{margin-bottom:calc(0.325em + 0.1875em)}.filepond--root[data-style-panel-layout~='circle'] .filepond--progress-indicator[data-align*='bottom'][data-align*='center']{margin-top:0;margin-bottom:0.1875em;margin-left:0.1875em}/*!\n * FilePondPluginImageEdit 1.6.3\n * Licensed under MIT, https://opensource.org/licenses/MIT/\n * Please visit https://pqina.nl/filepond/ for details.\n */.filepond--action-edit-item.filepond--action-edit-item{width:2em;height:2em;padding:0.1875em}.filepond--action-edit-item.filepond--action-edit-item[data-align*='center']{margin-left:-0.1875em}.filepond--action-edit-item.filepond--action-edit-item[data-align*='bottom']{margin-bottom:-0.1875em}.filepond--action-edit-item-alt{border:none;line-height:inherit;background:transparent;font-family:inherit;color:inherit;outline:none;padding:0;margin:0 0 0 0.25em;pointer-events:all;position:absolute}.filepond--action-edit-item-alt svg{width:1.3125em;height:1.3125em}.filepond--action-edit-item-alt span{font-size:0;opacity:0}.filepond--root[data-style-panel-layout~='circle'] .filepond--action-edit-item{opacity:1 !important;visibility:visible !important}.s-file-upload-wrapper .filepond--file-info-main{-webkit-line-clamp:1;display:-webkit-box;-webkit-box-orient:vertical}";

const SallaFileUpload = class {
    /**
     * Lazy load FilePond and all plugins
     * This reduces initial bundle size by ~200-250KB
     */
    async loadFilePond() {
        if (this.filePondLoaded)
            return;
        try {
            const [FilePondModule, FilePondPluginFileValidateSize, FilePondPluginImageExifOrientation, FilePondPluginImagePreview, FilePondPluginImageEdit, FilePondPluginFileValidateType, FilePondPluginFilePoster] = await Promise.all([
                Promise.resolve().then(function () { return require('./filepond-CsqfBzs3.js'); }).then(function (n) { return n.filepond; }),
                Promise.resolve().then(function () { return require('./filepond-plugin-file-validate-size-noIyy5Yp.js'); }).then(function (n) { return n.filepondPluginFileValidateSize; }),
                Promise.resolve().then(function () { return require('./filepond-plugin-image-exif-orientation-DLFISFCC.js'); }).then(function (n) { return n.filepondPluginImageExifOrientation; }),
                Promise.resolve().then(function () { return require('./filepond-plugin-image-preview-CtTDfIQp.js'); }).then(function (n) { return n.filepondPluginImagePreview; }),
                Promise.resolve().then(function () { return require('./filepond-plugin-image-edit-BkgUR4kG.js'); }).then(function (n) { return n.filepondPluginImageEdit; }),
                Promise.resolve().then(function () { return require('./filepond-plugin-file-validate-type-N76Wh7ZL.js'); }).then(function (n) { return n.filepondPluginFileValidateType; }),
                Promise.resolve().then(function () { return require('./filepond-plugin-file-poster-BWhvanmU.js'); }).then(function (n) { return n.filepondPluginFilePoster; })
            ]);
            this.FilePond = FilePondModule;
            // Register plugins
            this.FilePond.registerPlugin(FilePondPluginImagePreview.default, FilePondPluginImageExifOrientation.default, FilePondPluginFileValidateSize.default, FilePondPluginImageEdit.default, FilePondPluginFileValidateType.default, FilePondPluginFilePoster.default);
            this.filePondLoaded = true;
        }
        catch (error) {
            console.error('Failed to load FilePond:', error);
            salla.notify?.error?.('Failed to load file uploader. Please refresh the page.');
            throw error;
        }
    }
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        this.added = index.createEvent(this, "added");
        this.invalidInput = index.createEvent(this, "invalidInput");
        this.uploaded = index.createEvent(this, "uploaded");
        this.removed = index.createEvent(this, "removed");
        this.filePondLoaded = false;
        this.max_images_count = (count) => salla.lang.get('common.uploader.max_images_count', { count });
        this.max_count_hint = (count) => salla.lang.get('common.uploader.max_count_hint', { count });
        this.uploadedFiles = [];
        this.isListScrollerListenerAdded = false;
        this.showMaxCountHint = false;
        /**
         * Set the component to be profile image uploader with a preview and a circular shape
         */
        this.profileImage = false;
        /**
         * Allow to pass extra params to be sent with the upload request
         */
        this.payloadParams = {};
        /**
         * Accepted file types
         */
        this.accept = "image/png, image/jpeg, image/jpg, image/gif";
        /**
         * The submit request method.
         */
        this.method = 'POST';
        /**
         * json formData to be injected in the submit request
         */
        this.formData = "{}";
        /**
         * Enable or disable drag n' drop
         */
        this.allowDrop = true;
        /**
         * Enable or disable file browser
         */
        this.allowBrowse = true;
        /**
         * Allow drop to replace a file, only works when allowMultiple is false
         */
        this.allowReplace = true;
        /**
         * Enable or disable the revert processing button
         */
        this.allowRevert = true;
        /**
         * When set to false the remove button is hidden and disabled
         */
        this.allowRemove = true;
        /**
         * Enable or disable preview mode
         */
        this.allowImagePreview = true;
        /**
         * Fixed image preview height, overrides min and max preview height
         */
        this.imagePreviewHeight = null;
        /**
         * Fixed image poster height, overrides min and max preview height
         */
        this.filePosterHeight = null;
        /**
         * The maximum number of files that the pond can handle
         */
        this.maxFilesCount = 1;
        /**
         * The maxmimum number of files that can be uploaded in parallel
         */
        this.maxParallelUploads = 3;
        /**
         * Set to 'after' to add files to end of list (when dropped at the top of the list or added using browse or paste),
         * set to 'before' to add files at start of list.
         * Set to a compare function to automatically sort items when added
         */
        this.itemInsertLocation = 'after';
        /**
         * The interval to use before showing each item being added to the list
         */
        this.itemInsertInterval = 75;
        /**
         * Require drop on the FilePond element itself to catch the file.
         */
        this.dropOnElement = true;
        /**
         * Ignored file names when handling dropped directories. Dropping directories is not supported on all browsers.
         */
        this.ignoredFiles = ['.ds_store', 'thumbs.db', 'desktop.ini'];
        /**
         * The size of a chunk in bytes
         */
        this.chunkSize = 5000000;
        /**
         * Amount of times, and delayes, between retried uploading of a chunk
         */
        this.chunkRetryDelays = [500, 1000, 3000];
        /// Labels Props
        /**
         * The decimal separator used to render numbers. By default this is determined automatically.
         */
        this.labelDecimalSeparator = undefined;
        /**
         * The thousdands separator used to render numbers. By default this is determined automatically.
         */
        this.labelThousandsSeparator = undefined;
        /// SVG Icons
        /**
         * The icon used for remove actions
         */
        this.iconRemove = '<svg>...</svg>';
        /**
         * The icon used for process actions
         */
        this.iconProcess = '<svg>...</svg>';
        /**
         * The icon used for retry actions
         */
        this.iconRetry = '<svg>...</svg>';
        /**
         * The icon used for undo actions
         */
        this.iconUndo = '<svg>...</svg>';
        this.openFileBrowser = () => {
            this.filepond.browse();
        };
        //if cartItemIdIsPassed, we need to set the url
        if (!this.url && this.cartItemId) {
            this.url = salla.cart.getUploadImageEndpoint();
            this.instantUpload = true;
        }
        if (!this.url && this.profileImage) {
            this.url = salla.url.api('profile/update');
            this.instantUpload = true;
            this.host.hasAttribute('name') || (this.name = 'avatar');
        }
        if (!this.name) {
            this.name = 'file';
        }
        salla.lang.onLoaded(() => {
            this.host.querySelectorAll('.filepond--drop-label>label').forEach(label => label.innerHTML = this.getLabel());
            if (this.filepond) {
                this.filepond.labelFileTypeNotAllowed = salla.lang.get('common.uploader.invalid_type');
                this.filepond.labelMaxFileSizeExceeded = salla.lang.get('common.uploader.too_large');
                this.filepond.labelFileSizeNotAvailable = salla.lang.get('common.uploader.size_not_available');
                this.filepond.labelFileLoading = salla.lang.get('common.elements.loading');
                this.filepond.labelFileLoadError = salla.lang.get('common.uploader.failed_to_load');
                this.filepond.labelFileProcessingAborted = salla.lang.get('common.uploader.upload_cancelled');
                this.filepond.labelTapToCancel = salla.lang.get("common.elements.cancel");
                this.filepond.labelTapToRetry = salla.lang.get("common.elements.retry");
                this.filepond.labelButtonRemoveItem = salla.lang.get("common.elements.remove");
                this.filepond.labelButtonUndoItemProcessing = salla.lang.get("common.elements.undo");
                this.filepond.labelButtonProcessItem = salla.lang.get("common.uploader.upload");
                this.filepond.labelFileProcessingComplete = salla.lang.get('common.uploader.file_processing_complete'),
                    this.filepond.labelTapToUndo = salla.lang.get('common.uploader.tap_to_undo'),
                    this.filepond.labelFileProcessing = salla.lang.get('common.uploader.file_processing'),
                    this.filepond.labelFileProcessingError = salla.lang.get('common.uploader.file_processing_error');
            }
            const setNestedAsync = (lang, key, value) => {
                return new Promise((resolve) => {
                    salla.helpers.setNested(salla.lang.messages[lang], key, value);
                    resolve(true);
                });
            };
            const setTranslations = async () => {
                await setNestedAsync('ar.trans', 'common.uploader.max_images_count', '.أضف :count صور كحد أقصى');
                await setNestedAsync('en.trans', 'common.uploader.max_images_count', 'Add up to :count images maximum.');
                await setNestedAsync('ar.trans', 'common.uploader.max_count_hint', 'يمكنك رفع حتى :count صور');
                await setNestedAsync('en.trans', 'common.uploader.max_count_hint', 'You can upload up to :count images.');
            };
            this.max_count_hint = (count) => salla.lang.get('common.uploader.max_count_hint', { count });
            this.max_images_count = (count) => salla.lang.get('common.uploader.max_images_count', { count });
            setTranslations();
        });
    }
    addedHandler(error, file) {
        this.added.emit({ error: error, file: file });
        //if the file passed on initiate will not have type
        this.host.querySelector('.filepond--root').classList.remove('s-file-upload-has-error');
        if (error || !file.file.type) {
            this.host.querySelector('.filepond--root').classList.add('s-file-upload-has-error');
            return;
        }
        let container = new DataTransfer;
        let fileInput = this.getFormDataFileInput();
        // @ts-ignore
        container.items.add(file.file);
        fileInput.type = 'file';
        fileInput.files = container.files;
        fileInput.dispatchEvent(new window.Event('change', { bubbles: true }));
    }
    uploadedHandler() {
        let fileInput = this.getFormDataFileInput();
        fileInput.type = 'hidden';
        fileInput.value = this.value;
        this.hiddenInput.value = this.value;
        fileInput.dispatchEvent(new window.Event('change', { bubbles: true }));
        return this.uploaded.emit(this.value);
    }
    removedHandler(deletedFile) {
        let fileInput = this.getFormDataFileInput();
        fileInput.type = 'hidden';
        fileInput.value = '';
        this.host.closest('.s-product-options-option')?.removeAttribute('data-has-value');
        if (salla.url.is_page("cart") && deletedFile.getMetadata('id')) {
            salla.cart.api.deleteImage(deletedFile.getMetadata('id'));
        }
        if (this.height) {
            setTimeout(() => this.host.querySelector('.filepond--root').style.height = this.height, 1000);
        }
        this.hiddenInput.value = '';
        fileInput.dispatchEvent(new window.Event('change', { bubbles: true }));
        let toDelete = this.allowMultiple ? deletedFile.getMetadata('id') : deletedFile;
        return this.removed.emit(toDelete);
    }
    /**
     *  Method to set option for filepond
     * */
    async setOption(key, value) {
        this.filepond[key] = value;
    }
    getLabel() {
        if (this.labelIdle) {
            //some times we are passing label before translations is loaded, so here we will make sure that all translations are translated
            return this.labelIdle
                .replace('common.uploader.drag_and_drop', salla.lang.get('common.uploader.drag_and_drop'))
                .replace('common.uploader.browse', salla.lang.get('common.uploader.browse'));
        }
        return `${salla.lang.get('common.uploader.drag_and_drop')}<span class="filepond--label-action"> ${salla.lang.get('common.uploader.browse')} </span>`;
    }
    /**
     *
     * This method will fire head request to get the file size, it's head request,so it will be too fast.
     */
    getFileSize(url) {
        let http = new XMLHttpRequest();
        http.open('HEAD', url, false); // false = Synchronous
        http.send(null); // it will stop here until this http request is complete
        return http.status === 200 ? http.getResponseHeader('content-length') : '';
    }
    getFormDataFileInput() {
        return this.host.querySelector('.filepond--data input');
    }
    getFiles() {
        if (!this.value && !this.files) {
            return [];
        }
        try {
            let files = this.files
                ? typeof this.files === 'string' ? JSON.parse(this.files) : this.files
                : this.value.split(',').map(file => ({ url: file }));
            if (files.length) {
                this.host.closest('.s-product-options-option')?.setAttribute('data-has-value', 'true');
            }
            return files.map(file => ({
                source: file.id ? `${file.id}` : file.url,
                options: {
                    type: 'local',
                    file: {
                        name: file.url.substring(file.url.lastIndexOf('/') + 1),
                        size: this.getFileSize(file.url)
                    },
                    metadata: { poster: file.url, name: file.url, id: file.id },
                },
            }));
        }
        catch (e) {
            salla.log('failed To get files from: ' + (this.files || this.value));
        }
        return [];
    }
    async componentWillLoad() {
        // Load FilePond before component renders (reduces initial bundle by ~200-250KB)
        await this.loadFilePond();
        if (!this.labelIdle) {
            this.labelIdle = this.host.innerHTML;
            this.host.innerHTML = '';
        }
    }
    render() {
        return (index.h(index.Host, { key: 'b33365d20c42081e38ebf7c8b63bc0580807bc4e', class: {
                "s-file-upload": true,
                "s-file-upload-profile-image": this.profileImage,
                "s-file-upload-multiple": this.allowMultiple,
            } }, index.h("input", { key: '6e4458218bbeae75c21db781c85a376b29d5b11b', type: "file", name: this.name, value: this.value, ref: ele => this.fileUploader = ele, required: this.required, class: "s-file-upload-wrapper s-file-upload-input", accept: this.accept }), this.allowMultiple && this.showMaxCountHint ? index.h("div", { class: "s-file-upload-hint s-hidden" }, this.max_count_hint(this.maxFilesCount)) : '', index.h("input", { key: 'f06888dc84c34236d11c7b2ad7250bb41d8b9bfe', class: "s-hidden", name: 'hidden-' + this.name, required: this.required, value: this.value, ref: input => this.hiddenInput = input })));
    }
    componentDidLoad() {
        let files = this.getFiles();
        this.filepond = this.FilePond.create(this.fileUploader, {
            name: this.payloadName || this.name,
            //@ts-ignore
            files: files,
            required: this.required,
            disabled: this.disabled,
            allowDrop: this.allowDrop,
            allowBrowse: this.allowBrowse,
            allowPaste: this.allowPaste,
            allowMultiple: this.allowMultiple, //||files.length,
            allowReplace: this.allowReplace,
            allowRevert: this.allowRevert,
            allowProcess: this.allowProcess,
            allowReorder: this.allowReorder,
            storeAsFile: this.storeAsFile,
            forceRevert: this.forceRevert,
            maxFiles: this.maxFilesCount,
            imagePreviewHeight: this.imagePreviewHeight,
            filePosterHeight: this.filePosterHeight,
            allowImagePreview: this.allowImagePreview,
            maxParallelUploads: this.maxParallelUploads,
            checkValidity: this.checkValidity,
            itemInsertLocation: this.itemInsertLocation,
            itemInsertInterval: this.itemInsertInterval,
            credits: this.credits,
            dropOnPage: this.dropOnPage,
            dropOnElement: this.dropOnElement,
            dropValidation: this.dropValidation,
            ignoredFiles: this.ignoredFiles,
            onaddfile: (error, file) => this.addedHandler(error, file),
            onwarning: (error, _file, _status) => {
                if (error.body === "Max files") {
                    salla.notify.error(this.max_images_count(this.maxFilesCount));
                }
            },
            onprocessfile: (error, file) => {
                if (this.allowMultiple && !error) {
                    // Find the uploaded file data using the serverId returned from onload
                    const serverId = file.serverId;
                    const uploadedFile = this.uploadedFiles.find(f => f.id === serverId);
                    if (uploadedFile) {
                        // Set metadata on the file
                        file.setMetadata('id', uploadedFile.id, true);
                        file.setMetadata('url', uploadedFile.url, true);
                        this.value = {
                            id: uploadedFile.id,
                            url: uploadedFile.url
                        };
                        this.uploadedHandler();
                    }
                }
            },
            onupdatefiles: (fileItems) => {
                const dropLabel = this.host.querySelector('.filepond--drop-label');
                const listScroller = this.host.querySelector('.filepond--list-scroller');
                if (this.allowMultiple) {
                    if (fileItems.length > 0) {
                        dropLabel.classList.add('s-hidden');
                        if (!this.isListScrollerListenerAdded && fileItems.length < this.maxFilesCount) {
                            this.showMaxCountHint = true;
                            listScroller.addEventListener('click', this.openFileBrowser);
                            this.isListScrollerListenerAdded = true;
                        }
                    }
                    else {
                        dropLabel.classList.remove('s-hidden');
                        if (this.isListScrollerListenerAdded) {
                            listScroller.removeEventListener('click', this.openFileBrowser);
                            this.isListScrollerListenerAdded = false;
                        }
                    }
                    // Check if fileItems length equals maxFilesCount and remove event listener
                    if (fileItems.length === this.maxFilesCount && this.isListScrollerListenerAdded) {
                        listScroller.removeEventListener('click', this.openFileBrowser);
                        this.isListScrollerListenerAdded = false;
                        this.showMaxCountHint = false;
                    }
                }
            },
            server: {
                url: this.url,
                method: this.method,
                revert: null,
                // @ts-ignore
                process: {
                    onload: (response) => {
                        let responseData = JSON.parse(response).data;
                        if ((this.instantUpload || this.allowMultiple) && Array.isArray(responseData) && responseData.length > 0) {
                            const fileId = `${this.host.id}-${Math.floor(100 + Math.random() * 900)}`;
                            const fileUrl = responseData[0];
                            // Store the upload result to be picked up by onprocessfile
                            this.uploadedFiles.push({
                                id: fileId,
                                url: fileUrl
                            });
                            // Return the serverId that FilePond will use to identify this file
                            // We'll use the fileId as the serverId
                            return fileId;
                        }
                        else {
                            this.value = responseData.filePath || responseData.url;
                            this.uploadedHandler();
                            return this.value;
                        }
                    },
                    headers: salla.api.getHeaders(),
                    onerror: response => {
                        try {
                            let field = JSON.parse(response).error.fields;
                            field = field[this.payloadName || this.name];
                            return (field && field[0]) || salla.lang.get('common.errors.error_occurred');
                        }
                        catch (e) {
                            salla.log('FileUpload:: Error', e);
                            return salla.lang.get('common.errors.error_occurred');
                        }
                    },
                    ondata: formData => {
                        if (this.allowMultiple) {
                            const newFormData = new FormData();
                            // ? This line suppose to send all files in one request but its not working properly
                            // const filesArray = this.filepond.getFiles().map(file => file.file);
                            formData.forEach((value) => {
                                newFormData.append(`files[]`, value);
                            });
                            newFormData.append("type", this.type);
                            return newFormData;
                        }
                        if (this.payloadName && this.payloadName != this.name && !this.allowMultiple) {
                            formData.append(this.payloadName, this.filepond.getFile(0).file);
                            formData.delete(this.name);
                        }
                        if (this.payloadParams) {
                            for (const [k, v] of Object.entries(this.payloadParams)) {
                                formData.append(k, v);
                            }
                        }
                        if (this.cartItemId) {
                            formData.append('cart_item_id', this.cartItemId);
                        }
                        try {
                            Object.entries(JSON.parse(this.formData)).forEach((value) => value[1] && formData.append(value[0], value[1]));
                        }
                        catch (e) {
                            salla.log('Error to inject formData', e);
                        }
                        return formData;
                    }
                },
            },
            instantUpload: this.instantUpload,
            chunkUploads: this.chunkUploads,
            chunkForce: this.chunkForce,
            chunkSize: this.chunkSize,
            chunkRetryDelays: this.chunkRetryDelays,
            labelDecimalSeparator: this.labelDecimalSeparator,
            labelThousandsSeparator: this.labelThousandsSeparator,
            labelIdle: this.getLabel(),
            acceptedFileTypes: this.accept.replace(/\s+/g, '').split(','),
            labelFileTypeNotAllowed: salla.lang.get('common.uploader.invalid_type'),
            labelFileProcessingComplete: salla.lang.get('common.uploader.file_processing_complete'),
            labelTapToUndo: salla.lang.get('common.uploader.tap_to_undo'),
            labelFileProcessing: salla.lang.get('common.uploader.file_processing'),
            labelFileProcessingError: salla.lang.get('common.uploader.file_processing_error'),
            fileValidateTypeLabelExpectedTypes: '{allButLastType}, {lastType}',
            labelMaxFileSizeExceeded: salla.lang.get('common.uploader.too_large'),
            labelMaxFileSize: '{filesize}',
            maxFileSize: `${salla.config.get('store.settings.upload_size')}KB` || this.maxFileSize,
            beforeRemoveFile: (deletedFile) => new Promise((resolve) => {
                this.removedHandler(deletedFile);
                resolve(true);
            }),
        });
        this.hiddenInput.addEventListener('invalid', e => {
            this.invalidInput.emit(e);
        });
        this.hiddenInput.addEventListener('change', () => {
            this.hiddenInput.setCustomValidity('');
            this.hiddenInput.reportValidity();
        });
    }
    get host() { return index.getElement(this); }
};
SallaFileUpload.style = sallaFileUploadCss;

const sallaQuantityInputCss = "";

const SallaQuantityInput = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        /**
         * Based on store setting: store.settings.product.manual_quantity
         */
        this.disableInput = false;
        this.hostAttributes = {};
        this.hasIncrementSlot = false;
        this.hasDecrementSlot = false;
        this.didLoaded = false;
        this.quantity = 1;
        this.fireChangeEvent = true;
    }
    /**
     * Workaround to fire change event for the input.
     */
    watchPropHandler() {
        if (!this.didLoaded) {
            return;
        }
        if (!this.fireChangeEvent) {
            this.fireChangeEvent = true;
            return;
        }
        Helper.Helper.debounce(() => {
            salla.document.event.fireEvent(this.textInput, 'change', {
                bubbles: true,
                detail: { productId: this.cartItemId, quantity: this.quantity }
            });
        });
    }
    async componentWillLoad() {
        await salla.onReady();
        this.disableInput = !salla.config.get('store.settings.product.manual_quantity');
        this.quantity = parseInt(this.host.getAttribute('value')) || 1;
        this.hasIncrementSlot = !!this.host.querySelector('[slot="increment-button"]');
        this.hasDecrementSlot = !!this.host.querySelector('[slot="decrement-button"]');
    }
    componentDidLoad() {
        this.didLoaded = true;
        this.textInput.addEventListener('input', (event) => salla.helpers.inputDigitsOnly(event.target));
    }
    getInputAttributes() {
        for (let i = 0; i < this.host.attributes.length; i++) {
            if (!['id', 'value', 'min', 'class'].includes(this.host.attributes[i].name)) {
                this.hostAttributes[this.host.attributes[i].name] = this.host.attributes[i].value;
            }
        }
        return this.hostAttributes;
    }
    /**
     * decrease quantity by one.
     * @return HTMLSallaQuantityInputElement
     */
    async decrease() {
        return this.setValue(this.quantity - 1);
    }
    /**
     * increase quantity by one.
     * @return HTMLSallaQuantityInputElement
     */
    async increase() {
        return this.setValue(Number(this.quantity) + 1);
    }
    /**
     * set quantity by one.
     * @return HTMLSallaQuantityInputElement
     */
    async setValue(value, fireChangeEvent = true) {
        this.fireChangeEvent = fireChangeEvent;
        let maxQuantity = parseInt(this.host.getAttribute('max'));
        if (maxQuantity && value > maxQuantity) {
            value = maxQuantity;
        }
        if (value <= 1) {
            value = 1;
        }
        this.quantity = value;
        return this.host;
    }
    render() {
        const inputAttributes = this.getInputAttributes();
        if (!inputAttributes['aria-label'] && !inputAttributes['aria-labelledby']) {
            inputAttributes['aria-label'] = salla.lang.getWithDefault('common.elements.quantity', 'Quantity');
        }
        return (index.h(index.Host, { key: 'e7d9d747cb9498c384b7723067c81147dba9731c', class: "s-quantity-input" }, index.h("div", { key: '7b9b7849e00a68091647a8de34243481ddb5c2ed', class: "s-quantity-input-container" }, index.h("button", { key: '72032114d6b4294faf654c19909c8ddc6b46b6b6', onClick: () => this.increase(), class: "s-quantity-input-increase-button s-quantity-input-button", type: "button", "aria-label": salla.lang.getWithDefault('common.elements.increase_quantity', 'Increase quantity') }, !this.hasIncrementSlot ? index.h("span", { innerHTML: minus.plusIcon }) : '', index.h("slot", { key: 'c229ef5206240d5ee27371498bf8fabf6a6c1050', name: "increment-button" })), index.h("input", { key: 'bb9c2ac443c6920085261e19c6c32ac705360020', class: "s-quantity-input-input", ...inputAttributes, ref: (el) => this.textInput = el, onInput: (event) => this.setValue(event.target.value), min: "1", readOnly: this.disableInput, value: this.quantity }), index.h("button", { key: '1f3dfd851cc88f97c8f80a3bcd741e950061312e', class: "s-quantity-input-decrease-button s-quantity-input-button", onClick: () => this.decrease(), type: "button", "aria-label": salla.lang.getWithDefault('common.elements.decrease_quantity', 'Decrease quantity') }, !this.hasDecrementSlot ? index.h("span", { innerHTML: minus.minusIcon }) : '', index.h("slot", { key: '1058e6b14c8918f960a74ee18e7ac5d77bc5e3bc', name: "decrement-button" })))));
    }
    get host() { return index.getElement(this); }
    static get watchers() { return {
        "quantity": ["watchPropHandler"]
    }; }
};
SallaQuantityInput.style = sallaQuantityInputCss;

const sallaTelInputCss = ".iti{width:100%;position:relative;display:inline-block}.iti *{box-sizing:border-box;-moz-box-sizing:border-box}.iti__hide{display:none}.iti__v-hide{visibility:hidden}.iti input,.iti input[type=text],.iti input[type=tel]{position:relative;z-index:0;direction:ltr !important;margin-top:0 !important;margin-bottom:0 !important;padding-left:95px !important;padding-right:15px !important;margin-left:0 !important}.iti__flag-container{width:100%;position:absolute;top:0;bottom:0;left:0;padding:1px}.iti__selected-flag{width:80px;z-index:1;position:relative;display:flex;align-items:center;height:100%;float:left;padding:0 15px;box-shadow:none !important;outline:none !important}.ltr .iti__selected-flag{justify-content:flex-end}.iti__selected-flag:before{content:\"\";display:block;width:1px;height:70%;background:#eee;position:absolute;top:15%;right:0}.iti__selected-flag .iti__flag{display:none}.iti__arrow{width:15px;height:15px;line-height:15px;position:absolute;left:10px;top:50%;transform:translateY(-50%)}.iti__arrow:after{content:url(\"data:image/svg+xml,<svg version='1.1' xmlns='http://www.w3.org/2000/svg' viewBox='0 0 32 32'><title>keyboard_arrow_down</title><path d='M9.875 11.104l6.125 6.125 6.125-6.125 1.875 1.875-8 8-8-8z'></path></svg>\");font-size:16px;color:#999;opacity:0.65}.iti__arrow--up:after{transform:rotate(180deg)}.iti__country-list{width:100%;max-height:160px;position:absolute;top:100%;left:0;z-index:2;list-style:none;margin:0;padding:0;margin:5px 0 0 0;border-radius:5px;box-shadow:0px 2px 4px 2px rgba(0, 0, 0, 0.05);border:1px solid #f5f7f9;background-color:white;white-space:nowrap;overflow:hidden;overflow-y:auto;-webkit-overflow-scrolling:touch}.iti__country-list::-webkit-scrollbar{width:4px}.dark .iti__country-list{background-color:rgb(82, 82, 91);border-color:rgb(82, 82, 91)}.iti__country-list li{display:flex;align-items:center;justify-content:flex-start;flex-direction:row;padding:8px 10px;transition:background-color 0.3s}.iti__country-list li span{font-size:12px}.iti__country-list li:hover,.iti__country-list li.highlighted{background-color:rgba(0, 0, 0, 0.05)}.iti__country-list--dropup{bottom:100%;top:unset;margin:0 0 5px 0}@media (max-width: 500px){.iti__country-list{white-space:normal}}.iti__flag-box{display:inline-block;width:20px;margin:0 0 0 5px}.iti__divider{padding-bottom:5px;margin-bottom:5px;border-bottom:1px solid #eee}.iti__country{padding:5px 10px;outline:none}.iti__dial-code{font-family:Arial;color:#999;unicode-bidi:plaintext;letter-spacing:1px}.iti__country.iti__highlight{background-color:rgba(0, 0, 0, 0.05)}.iti__flag-box,.iti__country-name,.iti__dial-code{vertical-align:middle}.iti__country-name{flex:auto;margin:0;padding:0 10px;white-space:normal;line-height:1.3}.iti--allow-dropdown .iti__flag-container,.iti--separate-dial-code .iti__flag-container{left:0}.iti--allow-dropdown .iti__flag-container:hover{cursor:pointer}.iti--allow-dropdown input[disabled]+.iti__flag-container:hover,.iti--allow-dropdown input[readonly]+.iti__flag-container:hover{cursor:default}.iti--allow-dropdown input[disabled]+.iti__flag-container:hover .iti__selected-flag,.iti--allow-dropdown input[readonly]+.iti__flag-container:hover .iti__selected-flag{background-color:transparent}.iti--separate-dial-code .iti__selected-dial-code{unicode-bidi:plaintext;direction:ltr;letter-spacing:1px;font-family:\"Arial\", serif;font-size:13px}.iti--container{top:-18px !important;left:0 !important;padding:1px}.iti--container:hover{cursor:pointer}.iti-mobile .iti--container{inset:0;margin:20px;width:calc(100% - 40px);position:fixed;z-index:9999}.iti-mobile .iti__country-list{max-height:100%;width:100%;inset:0}.iti-mobile .iti__country{padding:10px 10px;line-height:1.5em}.iti__flag{width:20px}.iti__flag.iti__be{width:18px}.iti__flag.iti__ch{width:15px}.iti__flag.iti__mc{width:19px}.iti__flag.iti__ne{width:18px}.iti__flag.iti__np{width:13px}.iti__flag.iti__va{width:15px}@media (-webkit-min-device-pixel-ratio: 2), (min-resolution: 192dpi){.iti__flag{background-size:5652px 15px}}.iti__flag.iti__ac{height:10px;background-position:0px 0px}.iti__flag.iti__ad{height:14px;background-position:-22px 0px}.iti__flag.iti__ae{height:10px;background-position:-44px 0px}.iti__flag.iti__af{height:14px;background-position:-66px 0px}.iti__flag.iti__ag{height:14px;background-position:-88px 0px}.iti__flag.iti__ai{height:10px;background-position:-110px 0px}.iti__flag.iti__al{height:15px;background-position:-132px 0px}.iti__flag.iti__am{height:10px;background-position:-154px 0px}.iti__flag.iti__ao{height:14px;background-position:-176px 0px}.iti__flag.iti__aq{height:14px;background-position:-198px 0px}.iti__flag.iti__ar{height:13px;background-position:-220px 0px}.iti__flag.iti__as{height:10px;background-position:-242px 0px}.iti__flag.iti__at{height:14px;background-position:-264px 0px}.iti__flag.iti__au{height:10px;background-position:-286px 0px}.iti__flag.iti__aw{height:14px;background-position:-308px 0px}.iti__flag.iti__ax{height:13px;background-position:-330px 0px}.iti__flag.iti__az{height:10px;background-position:-352px 0px}.iti__flag.iti__ba{height:10px;background-position:-374px 0px}.iti__flag.iti__bb{height:14px;background-position:-396px 0px}.iti__flag.iti__bd{height:12px;background-position:-418px 0px}.iti__flag.iti__be{height:15px;background-position:-440px 0px}.iti__flag.iti__bf{height:14px;background-position:-460px 0px}.iti__flag.iti__bg{height:12px;background-position:-482px 0px}.iti__flag.iti__bh{height:12px;background-position:-504px 0px}.iti__flag.iti__bi{height:12px;background-position:-526px 0px}.iti__flag.iti__bj{height:14px;background-position:-548px 0px}.iti__flag.iti__bl{height:14px;background-position:-570px 0px}.iti__flag.iti__bm{height:10px;background-position:-592px 0px}.iti__flag.iti__bn{height:10px;background-position:-614px 0px}.iti__flag.iti__bo{height:14px;background-position:-636px 0px}.iti__flag.iti__bq{height:14px;background-position:-658px 0px}.iti__flag.iti__br{height:14px;background-position:-680px 0px}.iti__flag.iti__bs{height:10px;background-position:-702px 0px}.iti__flag.iti__bt{height:14px;background-position:-724px 0px}.iti__flag.iti__bv{height:15px;background-position:-746px 0px}.iti__flag.iti__bw{height:14px;background-position:-768px 0px}.iti__flag.iti__by{height:10px;background-position:-790px 0px}.iti__flag.iti__bz{height:14px;background-position:-812px 0px}.iti__flag.iti__ca{height:10px;background-position:-834px 0px}.iti__flag.iti__cc{height:10px;background-position:-856px 0px}.iti__flag.iti__cd{height:15px;background-position:-878px 0px}.iti__flag.iti__cf{height:14px;background-position:-900px 0px}.iti__flag.iti__cg{height:14px;background-position:-922px 0px}.iti__flag.iti__ch{height:15px;background-position:-944px 0px}.iti__flag.iti__ci{height:14px;background-position:-961px 0px}.iti__flag.iti__ck{height:10px;background-position:-983px 0px}.iti__flag.iti__cl{height:14px;background-position:-1005px 0px}.iti__flag.iti__cm{height:14px;background-position:-1027px 0px}.iti__flag.iti__cn{height:14px;background-position:-1049px 0px}.iti__flag.iti__co{height:14px;background-position:-1071px 0px}.iti__flag.iti__cp{height:14px;background-position:-1093px 0px}.iti__flag.iti__cr{height:12px;background-position:-1115px 0px}.iti__flag.iti__cu{height:10px;background-position:-1137px 0px}.iti__flag.iti__cv{height:12px;background-position:-1159px 0px}.iti__flag.iti__cw{height:14px;background-position:-1181px 0px}.iti__flag.iti__cx{height:10px;background-position:-1203px 0px}.iti__flag.iti__cy{height:14px;background-position:-1225px 0px}.iti__flag.iti__cz{height:14px;background-position:-1247px 0px}.iti__flag.iti__de{height:12px;background-position:-1269px 0px}.iti__flag.iti__dg{height:10px;background-position:-1291px 0px}.iti__flag.iti__dj{height:14px;background-position:-1313px 0px}.iti__flag.iti__dk{height:15px;background-position:-1335px 0px}.iti__flag.iti__dm{height:10px;background-position:-1357px 0px}.iti__flag.iti__do{height:14px;background-position:-1379px 0px}.iti__flag.iti__dz{height:14px;background-position:-1401px 0px}.iti__flag.iti__ea{height:14px;background-position:-1423px 0px}.iti__flag.iti__ec{height:14px;background-position:-1445px 0px}.iti__flag.iti__ee{height:13px;background-position:-1467px 0px}.iti__flag.iti__eg{height:14px;background-position:-1489px 0px}.iti__flag.iti__eh{height:10px;background-position:-1511px 0px}.iti__flag.iti__er{height:10px;background-position:-1533px 0px}.iti__flag.iti__es{height:14px;background-position:-1555px 0px}.iti__flag.iti__et{height:10px;background-position:-1577px 0px}.iti__flag.iti__eu{height:14px;background-position:-1599px 0px}.iti__flag.iti__fi{height:12px;background-position:-1621px 0px}.iti__flag.iti__fj{height:10px;background-position:-1643px 0px}.iti__flag.iti__fk{height:10px;background-position:-1665px 0px}.iti__flag.iti__fm{height:11px;background-position:-1687px 0px}.iti__flag.iti__fo{height:15px;background-position:-1709px 0px}.iti__flag.iti__fr{height:14px;background-position:-1731px 0px}.iti__flag.iti__ga{height:15px;background-position:-1753px 0px}.iti__flag.iti__gb{height:10px;background-position:-1775px 0px}.iti__flag.iti__gd{height:12px;background-position:-1797px 0px}.iti__flag.iti__ge{height:14px;background-position:-1819px 0px}.iti__flag.iti__gf{height:14px;background-position:-1841px 0px}.iti__flag.iti__gg{height:14px;background-position:-1863px 0px}.iti__flag.iti__gh{height:14px;background-position:-1885px 0px}.iti__flag.iti__gi{height:10px;background-position:-1907px 0px}.iti__flag.iti__gl{height:14px;background-position:-1929px 0px}.iti__flag.iti__gm{height:14px;background-position:-1951px 0px}.iti__flag.iti__gn{height:14px;background-position:-1973px 0px}.iti__flag.iti__gp{height:14px;background-position:-1995px 0px}.iti__flag.iti__gq{height:14px;background-position:-2017px 0px}.iti__flag.iti__gr{height:14px;background-position:-2039px 0px}.iti__flag.iti__gs{height:10px;background-position:-2061px 0px}.iti__flag.iti__gt{height:13px;background-position:-2083px 0px}.iti__flag.iti__gu{height:11px;background-position:-2105px 0px}.iti__flag.iti__gw{height:10px;background-position:-2127px 0px}.iti__flag.iti__gy{height:12px;background-position:-2149px 0px}.iti__flag.iti__hk{height:14px;background-position:-2171px 0px}.iti__flag.iti__hm{height:10px;background-position:-2193px 0px}.iti__flag.iti__hn{height:10px;background-position:-2215px 0px}.iti__flag.iti__hr{height:10px;background-position:-2237px 0px}.iti__flag.iti__ht{height:12px;background-position:-2259px 0px}.iti__flag.iti__hu{height:10px;background-position:-2281px 0px}.iti__flag.iti__ic{height:14px;background-position:-2303px 0px}.iti__flag.iti__id{height:14px;background-position:-2325px 0px}.iti__flag.iti__ie{height:10px;background-position:-2347px 0px}.iti__flag.iti__il{height:15px;background-position:-2369px 0px}.iti__flag.iti__im{height:10px;background-position:-2391px 0px}.iti__flag.iti__in{height:14px;background-position:-2413px 0px}.iti__flag.iti__io{height:10px;background-position:-2435px 0px}.iti__flag.iti__iq{height:14px;background-position:-2457px 0px}.iti__flag.iti__ir{height:12px;background-position:-2479px 0px}.iti__flag.iti__is{height:15px;background-position:-2501px 0px}.iti__flag.iti__it{height:14px;background-position:-2523px 0px}.iti__flag.iti__je{height:12px;background-position:-2545px 0px}.iti__flag.iti__jm{height:10px;background-position:-2567px 0px}.iti__flag.iti__jo{height:10px;background-position:-2589px 0px}.iti__flag.iti__jp{height:14px;background-position:-2611px 0px}.iti__flag.iti__ke{height:14px;background-position:-2633px 0px}.iti__flag.iti__kg{height:12px;background-position:-2655px 0px}.iti__flag.iti__kh{height:13px;background-position:-2677px 0px}.iti__flag.iti__ki{height:10px;background-position:-2699px 0px}.iti__flag.iti__km{height:12px;background-position:-2721px 0px}.iti__flag.iti__kn{height:14px;background-position:-2743px 0px}.iti__flag.iti__kp{height:10px;background-position:-2765px 0px}.iti__flag.iti__kr{height:14px;background-position:-2787px 0px}.iti__flag.iti__kw{height:10px;background-position:-2809px 0px}.iti__flag.iti__ky{height:10px;background-position:-2831px 0px}.iti__flag.iti__kz{height:10px;background-position:-2853px 0px}.iti__flag.iti__la{height:14px;background-position:-2875px 0px}.iti__flag.iti__lb{height:14px;background-position:-2897px 0px}.iti__flag.iti__lc{height:10px;background-position:-2919px 0px}.iti__flag.iti__li{height:12px;background-position:-2941px 0px}.iti__flag.iti__lk{height:10px;background-position:-2963px 0px}.iti__flag.iti__lr{height:11px;background-position:-2985px 0px}.iti__flag.iti__ls{height:14px;background-position:-3007px 0px}.iti__flag.iti__lt{height:12px;background-position:-3029px 0px}.iti__flag.iti__lu{height:12px;background-position:-3051px 0px}.iti__flag.iti__lv{height:10px;background-position:-3073px 0px}.iti__flag.iti__ly{height:10px;background-position:-3095px 0px}.iti__flag.iti__ma{height:14px;background-position:-3117px 0px}.iti__flag.iti__mc{height:15px;background-position:-3139px 0px}.iti__flag.iti__md{height:10px;background-position:-3160px 0px}.iti__flag.iti__me{height:10px;background-position:-3182px 0px}.iti__flag.iti__mf{height:14px;background-position:-3204px 0px}.iti__flag.iti__mg{height:14px;background-position:-3226px 0px}.iti__flag.iti__mh{height:11px;background-position:-3248px 0px}.iti__flag.iti__mk{height:10px;background-position:-3270px 0px}.iti__flag.iti__ml{height:14px;background-position:-3292px 0px}.iti__flag.iti__mm{height:14px;background-position:-3314px 0px}.iti__flag.iti__mn{height:10px;background-position:-3336px 0px}.iti__flag.iti__mo{height:14px;background-position:-3358px 0px}.iti__flag.iti__mp{height:10px;background-position:-3380px 0px}.iti__flag.iti__mq{height:14px;background-position:-3402px 0px}.iti__flag.iti__mr{height:14px;background-position:-3424px 0px}.iti__flag.iti__ms{height:10px;background-position:-3446px 0px}.iti__flag.iti__mt{height:14px;background-position:-3468px 0px}.iti__flag.iti__mu{height:14px;background-position:-3490px 0px}.iti__flag.iti__mv{height:14px;background-position:-3512px 0px}.iti__flag.iti__mw{height:14px;background-position:-3534px 0px}.iti__flag.iti__mx{height:12px;background-position:-3556px 0px}.iti__flag.iti__my{height:10px;background-position:-3578px 0px}.iti__flag.iti__mz{height:14px;background-position:-3600px 0px}.iti__flag.iti__na{height:14px;background-position:-3622px 0px}.iti__flag.iti__nc{height:10px;background-position:-3644px 0px}.iti__flag.iti__ne{height:15px;background-position:-3666px 0px}.iti__flag.iti__nf{height:10px;background-position:-3686px 0px}.iti__flag.iti__ng{height:10px;background-position:-3708px 0px}.iti__flag.iti__ni{height:12px;background-position:-3730px 0px}.iti__flag.iti__nl{height:14px;background-position:-3752px 0px}.iti__flag.iti__no{height:15px;background-position:-3774px 0px}.iti__flag.iti__np{height:15px;background-position:-3796px 0px}.iti__flag.iti__nr{height:10px;background-position:-3811px 0px}.iti__flag.iti__nu{height:10px;background-position:-3833px 0px}.iti__flag.iti__nz{height:10px;background-position:-3855px 0px}.iti__flag.iti__om{height:10px;background-position:-3877px 0px}.iti__flag.iti__pa{height:14px;background-position:-3899px 0px}.iti__flag.iti__pe{height:14px;background-position:-3921px 0px}.iti__flag.iti__pf{height:14px;background-position:-3943px 0px}.iti__flag.iti__pg{height:15px;background-position:-3965px 0px}.iti__flag.iti__ph{height:10px;background-position:-3987px 0px}.iti__flag.iti__pk{height:14px;background-position:-4009px 0px}.iti__flag.iti__pl{height:13px;background-position:-4031px 0px}.iti__flag.iti__pm{height:14px;background-position:-4053px 0px}.iti__flag.iti__pn{height:10px;background-position:-4075px 0px}.iti__flag.iti__pr{height:14px;background-position:-4097px 0px}.iti__flag.iti__ps{height:10px;background-position:-4119px 0px}.iti__flag.iti__pt{height:14px;background-position:-4141px 0px}.iti__flag.iti__pw{height:13px;background-position:-4163px 0px}.iti__flag.iti__py{height:11px;background-position:-4185px 0px}.iti__flag.iti__qa{height:8px;background-position:-4207px 0px}.iti__flag.iti__re{height:14px;background-position:-4229px 0px}.iti__flag.iti__ro{height:14px;background-position:-4251px 0px}.iti__flag.iti__rs{height:14px;background-position:-4273px 0px}.iti__flag.iti__ru{height:14px;background-position:-4295px 0px}.iti__flag.iti__rw{height:14px;background-position:-4317px 0px}.iti__flag.iti__sa{height:14px;background-position:-4339px 0px}.iti__flag.iti__sb{height:10px;background-position:-4361px 0px}.iti__flag.iti__sc{height:10px;background-position:-4383px 0px}.iti__flag.iti__sd{height:10px;background-position:-4405px 0px}.iti__flag.iti__se{height:13px;background-position:-4427px 0px}.iti__flag.iti__sg{height:14px;background-position:-4449px 0px}.iti__flag.iti__sh{height:10px;background-position:-4471px 0px}.iti__flag.iti__si{height:10px;background-position:-4493px 0px}.iti__flag.iti__sj{height:15px;background-position:-4515px 0px}.iti__flag.iti__sk{height:14px;background-position:-4537px 0px}.iti__flag.iti__sl{height:14px;background-position:-4559px 0px}.iti__flag.iti__sm{height:15px;background-position:-4581px 0px}.iti__flag.iti__sn{height:14px;background-position:-4603px 0px}.iti__flag.iti__so{height:14px;background-position:-4625px 0px}.iti__flag.iti__sr{height:14px;background-position:-4647px 0px}.iti__flag.iti__ss{height:10px;background-position:-4669px 0px}.iti__flag.iti__st{height:10px;background-position:-4691px 0px}.iti__flag.iti__sv{height:12px;background-position:-4713px 0px}.iti__flag.iti__sx{height:14px;background-position:-4735px 0px}.iti__flag.iti__sy{height:14px;background-position:-4757px 0px}.iti__flag.iti__sz{height:14px;background-position:-4779px 0px}.iti__flag.iti__ta{height:10px;background-position:-4801px 0px}.iti__flag.iti__tc{height:10px;background-position:-4823px 0px}.iti__flag.iti__td{height:14px;background-position:-4845px 0px}.iti__flag.iti__tf{height:14px;background-position:-4867px 0px}.iti__flag.iti__tg{height:13px;background-position:-4889px 0px}.iti__flag.iti__th{height:14px;background-position:-4911px 0px}.iti__flag.iti__tj{height:10px;background-position:-4933px 0px}.iti__flag.iti__tk{height:10px;background-position:-4955px 0px}.iti__flag.iti__tl{height:10px;background-position:-4977px 0px}.iti__flag.iti__tm{height:14px;background-position:-4999px 0px}.iti__flag.iti__tn{height:14px;background-position:-5021px 0px}.iti__flag.iti__to{height:10px;background-position:-5043px 0px}.iti__flag.iti__tr{height:14px;background-position:-5065px 0px}.iti__flag.iti__tt{height:12px;background-position:-5087px 0px}.iti__flag.iti__tv{height:10px;background-position:-5109px 0px}.iti__flag.iti__tw{height:14px;background-position:-5131px 0px}.iti__flag.iti__tz{height:14px;background-position:-5153px 0px}.iti__flag.iti__ua{height:14px;background-position:-5175px 0px}.iti__flag.iti__ug{height:14px;background-position:-5197px 0px}.iti__flag.iti__um{height:11px;background-position:-5219px 0px}.iti__flag.iti__un{height:14px;background-position:-5241px 0px}.iti__flag.iti__us{height:11px;background-position:-5263px 0px}.iti__flag.iti__uy{height:14px;background-position:-5285px 0px}.iti__flag.iti__uz{height:10px;background-position:-5307px 0px}.iti__flag.iti__va{height:15px;background-position:-5329px 0px}.iti__flag.iti__vc{height:14px;background-position:-5346px 0px}.iti__flag.iti__ve{height:14px;background-position:-5368px 0px}.iti__flag.iti__vg{height:10px;background-position:-5390px 0px}.iti__flag.iti__vi{height:14px;background-position:-5412px 0px}.iti__flag.iti__vn{height:14px;background-position:-5434px 0px}.iti__flag.iti__vu{height:12px;background-position:-5456px 0px}.iti__flag.iti__wf{height:14px;background-position:-5478px 0px}.iti__flag.iti__ws{height:10px;background-position:-5500px 0px}.iti__flag.iti__xk{height:15px;background-position:-5522px 0px}.iti__flag.iti__ye{height:14px;background-position:-5544px 0px}.iti__flag.iti__yt{height:14px;background-position:-5566px 0px}.iti__flag.iti__za{height:14px;background-position:-5588px 0px}.iti__flag.iti__zm{height:14px;background-position:-5610px 0px}.iti__flag.iti__zw{height:10px;background-position:-5632px 0px}.iti__flag{height:15px;box-shadow:0px 0px 1px 0px #888;background-image:url(\"https://cdn.salla.network/images/flags.png\");background-repeat:no-repeat;background-color:#f5f7f9;background-position:20px 0}@media (-webkit-min-device-pixel-ratio: 2), (min-resolution: 192dpi){.iti__flag{background-image:url(\"https://cdn.salla.network/images/flags@2x.png\")}}.iti__flag.iti__np{background-color:transparent}";

const SallaTelInput = class {
    /**
     * Lazy load intl-tel-input library
     * This reduces initial bundle size by ~80-90KB
     */
    async loadTelInput() {
        if (this.TelInput)
            return;
        try {
            const telInputModule = await Promise.resolve().then(function () { return require('./index-B-K6ToCR.js'); }).then(function (n) { return n.index; });
            this.TelInput = telInputModule.default;
        }
        catch (error) {
            console.error('Failed to load Tel Input:', error);
            salla.notify?.error?.('Failed to load phone input. Please refresh the page.');
            throw error;
        }
    }
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        this.phoneEntered = index.createEvent(this, "phoneEntered");
        /**
         * input name
         */
        this.name = 'phone';
        /**
         * input name
         */
        this.disabled = false;
        /**
         * Current country_code
         */
        this.countryCode = salla.config.get('user.country_code', 'SA') || 'SA';
        this.countryCodeLabel = salla.lang.get('common.country_code');
        this.mobileLabel = salla.lang.get('common.elements.mobile');
        this.tooShort = salla.lang.get('common.errors.too_short', { attribute: this.mobileLabel });
        this.tooLong = salla.lang.get('common.errors.too_long', { attribute: this.mobileLabel });
        this.invalidCountryCode = salla.lang.get('common.errors.invalid_value', { attribute: this.countryCodeLabel });
        this.invalidNumber = salla.lang.get('common.errors.invalid_value', { attribute: this.mobileLabel });
        this.errorMap = [this.invalidNumber, this.invalidCountryCode, this.tooShort, this.tooLong, this.invalidNumber];
        salla.lang.onLoaded(() => {
            this.mobileLabel = salla.lang.get('common.elements.mobile');
            this.countryCodeLabel = salla.lang.get('common.elements.country_code');
            this.invalidNumber = salla.lang.get('common.errors.invalid_value', { attribute: this.mobileLabel });
            this.invalidCountryCode = salla.lang.get('common.errors.invalid_value', { attribute: this.countryCodeLabel });
            this.tooShort = salla.lang.get('common.errors.too_short', { attribute: this.mobileLabel });
            this.tooLong = salla.lang.get('common.errors.too_long', { attribute: this.mobileLabel });
            this.mobileRequired = salla.lang.get('common.errors.field_required', { attribute: this.mobileLabel });
            this.errorMap = [this.invalidNumber, this.invalidCountryCode, this.tooShort, this.tooLong, this.invalidNumber];
        });
    }
    /**
     * Get current values
     * @return {{mobile:number,countryCode:'SA'|string}}
     */
    async getValues() {
        return {
            [this.name]: this.phone = this.phoneInput.value,
            countryCode: this.countryCode = this.countryCodeInput.value,
            countryKey: this.host.querySelector('.iti__selected-dial-code').innerText
        };
    }
    /**
     * Is current data valid or not
     * @return {boolean}
     */
    async isValid() {
        this.reset();
        if (this.iti.isValidNumber())
            return true;
        if (!this.phoneInput.value.trim()) {
            this.phoneInput.classList.add("s-has-error");
            this.errorMsg.innerText = this.mobileRequired || 'The mobile is required';
            return;
        }
        this.phoneInput.classList.add("s-has-error");
        let errorCode = this.iti.getValidationError();
        this.errorMsg.innerText = this.errorMap[errorCode] || '';
        salla.logger.info('Phone number (' + this.countryCode + ' - ' + this.phone + ') is not valid, error code ' + errorCode);
        return false;
    }
    async initTelInput() {
        // Load intl-tel-input before initializing
        await this.loadTelInput();
        salla.helpers.inputDigitsOnly(this.phoneInput);
        this.iti = this.TelInput(this.phoneInput, {
            initialCountry: this.countryCode?.toLowerCase() || 'sa',
            preferredCountries: ['sa', 'ae', 'kw', 'bh', 'qa', 'iq', 'om', 'ye', 'eg', 'jo', 'ps', 'sd', 'lb', 'dz', 'tn', 'ma', 'ly'],
            excludeCountries: ['sy', 'ir', 'cu', 'kp'],
            formatOnDisplay: false,
            separateDialCode: true,
            autoPlaceholder: 'aggressive',
            utilsScript: 'https://cdnjs.cloudflare.com/ajax/libs/intl-tel-input/21.2.7/js/utils.min.js',
            dropdownContainer: this.host
        });
        this.autofocus && this.phoneInput.focus();
        this.phoneInput.addEventListener("countrychange", () => {
            let data = this.iti.getSelectedCountryData();
            let value = data.iso2.toUpperCase();
            this.countryCodeInput.value = value;
            this.countryCode = value;
            this.phoneEntered.emit({ number: this.phone, country_code: value });
        });
        // on blur: validate
        // this.phoneInput.addEventListener('blur', () => this.isValid());
        // on keyup / change flag: reset
        this.phoneInput.addEventListener('input', (e) => {
            salla.helpers.inputDigitsOnly(e.target);
            this.reset();
            this.phoneEntered.emit({ number: e.target.value, country_code: this.countryCode });
        });
    }
    reset() {
        this.phoneInput.classList.remove("s-has-error");
        this.errorMsg.innerHTML = "";
    }
    ;
    handleCountryInput(event) {
        if (!!this.phone) {
            this.phoneEntered.emit({ number: event.target.value, country_code: this.countryCode });
        }
    }
    render() {
        return (index.h(index.Host, { key: 'd48f549012efc0bc465bd7da8bd5693506514d25', class: "s-tel-input" }, index.h("input", { key: '04fa4306973163688b74effbbbd8838b39deb6ad', type: "tel", disabled: this.disabled, name: this.name, value: this.phone, onChange: (event) => this.handleCountryInput(event), ref: el => this.phoneInput = el, enterkeyhint: "next", autocomplete: "tel", class: "s-tel-input-control tel-input s-ltr" }), index.h("span", { key: '9ae2ddda777dfc2672756f8a6ada215169c1b486', class: "s-tel-input-error-msg", ref: el => this.errorMsg = el }), index.h("input", { key: '83961dde155948b5bfe9030575a084470d707442', type: "hidden", name: "country_code", value: this.countryCode, ref: el => this.countryCodeInput = el, class: "country_code" })));
    }
    componentDidLoad() {
        this.initTelInput();
    }
    get host() { return index.getElement(this); }
};
SallaTelInput.style = sallaTelInputCss;

exports.salla_booking_field = SallaBookingField;
exports.salla_color_picker = SallaColorPicker;
exports.salla_conditional_fields = SallaConditionalFields;
exports.salla_datetime_picker = SallaDatetimePicker;
exports.salla_file_upload = SallaFileUpload;
exports.salla_quantity_input = SallaQuantityInput;
exports.salla_tel_input = SallaTelInput;
